// @salt-ds/lab/dist-es/query-input/QueryInput.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { makePrefixer } from '@salt-ds/core';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import './internal/CategoryList.js';
import './internal/CategoryListContext.js';
import './internal/CategoryListItem.js';
import { QueryInputBody } from './internal/QueryInputBody.js';
import './internal/SearchList.js';
import './internal/ValueList.js';
import { ValueSelector } from './internal/ValueSelector.js';
import { useQueryInput } from './useQueryInput.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './QueryInput.css.js';

const withBaseName = makePrefixer("saltQueryInput");
const QueryInput = forwardRef(
  function QueryInput2(props, externalRef) {
    const {
      disabled,
      categories,
      selectedItems,
      onChange,
      defaultSelectedItems,
      showCategory,
      autoClose,
      displayedItemCount,
      booleanOperator,
      onBooleanOperatorChange,
      defaultBooleanOperator,
      className,
      ...restProps
    } = props;
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-query-input",
      css: css_248z,
      window: targetWindow
    });
    const {
      queryInputProps,
      queryInputBodyRef,
      queryInputBodyProps,
      valueSelectorProps
    } = useQueryInput(props, externalRef);
    return /* @__PURE__ */ jsxs("div", {
      className: clsx(withBaseName(), className),
      onBlur: queryInputProps.onBlur,
      onFocus: queryInputProps.onFocus,
      "data-testid": "query-input",
      ...restProps,
      children: [
        /* @__PURE__ */ jsx(QueryInputBody, {
          ref: queryInputBodyRef,
          ...queryInputBodyProps,
          showCategory
        }),
        /* @__PURE__ */ jsx(ValueSelector, {
          ...valueSelectorProps
        })
      ]
    });
  }
);

export { QueryInput };
