// @salt-ds/lab/dist-es/toolbar/toolbar-field/useToolbarField.js
import { clsx } from 'clsx';
import { useCallback, isValidElement } from 'react';
import { ToolbarButton } from '../ToolbarButton.js';

const NullActivationIndicator = () => null;
const primaryControls = ["Dropdown", "Input"];
const getChildElementName = (element) => {
  if (isValidElement(element)) {
    const { type } = element;
    if (typeof type === "string") {
      return type;
    } else if (typeof type.name === "string") {
      return type.name;
    } else if ("displayName" in type) {
      return type.displayName;
    } else {
      return "";
    }
  } else {
    throw Error(
      "useToolbarField, child of ToolbarField is not valid ReactElememnt"
    );
  }
};
const isToolbarButton = (element) => isValidElement(element) && element.type === ToolbarButton;
const getToolbarFormFieldProps = (child, isOverflowPanel = false) => {
  let activationIndicator = NullActivationIndicator;
  let variant = "tertiary";
  const element = child;
  if (isValidElement(element)) {
    const name = getChildElementName(element);
    if (primaryControls.includes(name)) {
      activationIndicator = void 0;
      variant = "primary";
    }
  }
  return {
    ActivationIndicatorComponent: activationIndicator,
    className: clsx({
      "saltFormFieldLegacy-toolbarButton": isOverflowPanel && isToolbarButton(element),
      [`saltFormFieldLegacy-${variant}`]: variant
    }),
    fullWidth: false
  };
};
const InteractiveComponents = ["Input", "Dropdown"];
const useToolbarField = (props) => {
  const {
    ActivationIndicatorComponent: ActivationIndicatorComponentProp,
    className: classNameProp,
    inOverflowPanel,
    labelPlacement: labelPlacementProp,
    onClick,
    ...rest
  } = props;
  const childElementName = getChildElementName(props.children);
  const { className, ActivationIndicatorComponent, ...calculatedProps } = getToolbarFormFieldProps(props.children, inOverflowPanel);
  const labelPlacement = labelPlacementProp != null ? labelPlacementProp : inOverflowPanel ? "top" : "left";
  const handleClick = useCallback(
    (e) => {
      if (onClick && !InteractiveComponents.includes(childElementName)) {
        onClick(e);
      }
    },
    [childElementName, onClick]
  );
  return {
    ActivationIndicatorComponent: ActivationIndicatorComponentProp != null ? ActivationIndicatorComponentProp : ActivationIndicatorComponent,
    className: clsx("saltToolbarField", classNameProp, className),
    labelPlacement,
    onClick: handleClick,
    ...rest,
    ...calculatedProps
  };
};

export { getToolbarFormFieldProps, useToolbarField };
