// @salt-ds/core/dist-cjs/utils/useFloatingUI/useFloatingUI.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var react = require('@floating-ui/react');
var React = require('react');
var SaltProvider = require('../../salt-provider/SaltProvider.js');

const DefaultFloatingComponent = React.forwardRef(function DefaultFloatingComponent2(props, ref) {
  const {
    open,
    top,
    left,
    position,
    width,
    height,
    focusManagerProps,
    ...rest
  } = props;
  const style = {
    top,
    left,
    position
  };
  const { themeNext } = SaltProvider.useTheme();
  const ChosenSaltProvider = themeNext ? SaltProvider.UNSTABLE_SaltProviderNext : SaltProvider.SaltProvider;
  if (focusManagerProps && open) {
    return /* @__PURE__ */ jsxRuntime.jsx(react.FloatingPortal, {
      children: /* @__PURE__ */ jsxRuntime.jsx(ChosenSaltProvider, {
        children: /* @__PURE__ */ jsxRuntime.jsx(react.FloatingFocusManager, {
          ...focusManagerProps,
          children: /* @__PURE__ */ jsxRuntime.jsx("div", {
            style,
            ...rest,
            ref
          })
        })
      })
    });
  }
  return open ? /* @__PURE__ */ jsxRuntime.jsx(react.FloatingPortal, {
    children: /* @__PURE__ */ jsxRuntime.jsx(ChosenSaltProvider, {
      children: /* @__PURE__ */ jsxRuntime.jsx("div", {
        style,
        ...rest,
        ref
      })
    })
  }) : null;
});
const FloatingComponentContext = React.createContext({
  Component: DefaultFloatingComponent
});
if (process.env.NODE_ENV !== "production") {
  FloatingComponentContext.displayName = "FloatingComponentContext";
}
function FloatingComponentProvider(props) {
  const { Component, children } = props;
  const value = React.useMemo(() => ({ Component }), [Component]);
  return /* @__PURE__ */ jsxRuntime.jsx(FloatingComponentContext.Provider, {
    value,
    children
  });
}
function useFloatingComponent() {
  return React.useContext(FloatingComponentContext);
}
const defaultGetMiddleware = (defaultMiddleware) => defaultMiddleware;
const defaultFloatingPlaform = {
  platform: react.platform,
  middleware: defaultGetMiddleware,
  animationFrame: false
};
const FloatingPlatformContext = React.createContext(
  defaultFloatingPlaform
);
function FloatingPlatformProvider(props) {
  const {
    platform: platformProp,
    middleware,
    animationFrame,
    children
  } = props;
  const floatingPlatformContextValue = React.useMemo(
    () => ({
      platform: platformProp != null ? platformProp : react.platform,
      middleware: middleware != null ? middleware : defaultGetMiddleware,
      animationFrame: animationFrame || false
    }),
    [platformProp, middleware, animationFrame]
  );
  return /* @__PURE__ */ jsxRuntime.jsx(FloatingPlatformContext.Provider, {
    value: floatingPlatformContextValue,
    children
  });
}
function useFloatingPlatform() {
  return React.useContext(FloatingPlatformContext);
}
const DEFAULT_FLOATING_UI_MIDDLEWARE = [
  react.flip(),
  react.shift({ limiter: react.limitShift() })
];
function useFloatingUI(props) {
  const {
    placement,
    strategy,
    middleware = DEFAULT_FLOATING_UI_MIDDLEWARE,
    open = false,
    onOpenChange
  } = props;
  const handleOpenChange = (open2, boolean, reason) => {
    update();
    onOpenChange == null ? void 0 : onOpenChange(open2, boolean, reason);
  };
  const {
    platform: contextPlatform,
    middleware: contextMiddleware,
    animationFrame
  } = useFloatingPlatform();
  const { refs, update, ...rest } = react.useFloating({
    placement,
    strategy,
    middleware: contextMiddleware(middleware),
    open,
    onOpenChange: handleOpenChange,
    whileElementsMounted: (...args) => {
      const cleanup = react.autoUpdate(...args, { animationFrame });
      return cleanup;
    },
    platform: contextPlatform
  });
  return {
    reference: refs.setReference,
    floating: refs.setFloating,
    refs,
    update,
    ...rest
  };
}

exports.DEFAULT_FLOATING_UI_MIDDLEWARE = DEFAULT_FLOATING_UI_MIDDLEWARE;
exports.FloatingComponentProvider = FloatingComponentProvider;
exports.FloatingPlatformProvider = FloatingPlatformProvider;
exports.useFloatingComponent = useFloatingComponent;
exports.useFloatingPlatform = useFloatingPlatform;
exports.useFloatingUI = useFloatingUI;
