// @salt-ds/lab/dist-es/contact-details/ContactSecondaryInfo.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { makePrefixer, useId, Text } from '@salt-ds/core';
import { clsx } from 'clsx';
import { forwardRef, useEffect } from 'react';
import './internal/FavoriteToggleWithTooltip.js';
import { useContactDetailsContext } from './internal/ContactDetailsContext.js';

const withBaseName = makePrefixer("saltContactSecondaryInfo");
const ContactSecondaryInfo = forwardRef(function ContactSecondaryInfo2(props, ref) {
  const {
    id: idProp,
    text,
    icon: Icon,
    className,
    ValueComponent,
    ...restProps
  } = props;
  const { setSecondary, setSecondaryId, variant } = useContactDetailsContext();
  const id = useId(idProp);
  useEffect(() => {
    setSecondary(text || "");
    setSecondaryId(id);
    return () => {
      setSecondary(void 0);
      setSecondaryId(void 0);
    };
  }, [id, text, setSecondary, setSecondaryId]);
  return /* @__PURE__ */ jsxs(Text, {
    styleAs: variant === "default" ? "h4" : void 0,
    maxRows: 1,
    ...restProps,
    id,
    ref,
    className: clsx(withBaseName(), className),
    "data-testid": "secondary",
    children: [
      Icon ? /* @__PURE__ */ jsx(Icon, {
        className: withBaseName("icon")
      }) : null,
      text
    ]
  });
});

export { ContactSecondaryInfo };
