// @salt-ds/lab/dist-es/contact-details/ContactFavoriteToggle.js
import { jsx } from 'react/jsx-runtime';
import { useId, useControlled } from '@salt-ds/core';
import { forwardRef } from 'react';
import { FavoriteToggleWithTooltip } from './internal/FavoriteToggleWithTooltip.js';
import './internal/ContactDetailsContext.js';

const ContactFavoriteToggle = forwardRef(function ContactFavoriteToggle2(props, ref) {
  const {
    id: idProp,
    isFavorite: isFavoriteProp,
    defaultIsFavorite = false,
    onChange: onChangeProp,
    tooltipProps,
    ...restProps
  } = props;
  const id = useId(idProp);
  const [isFavorite, setIsFavorite] = useControlled({
    controlled: isFavoriteProp,
    default: defaultIsFavorite,
    name: "ContactDetails",
    state: "isFavorite"
  });
  const onChange = (isSelected) => {
    setIsFavorite(isSelected);
    if (onChangeProp) {
      onChangeProp(isSelected);
    }
  };
  const tooltipTitle = `${isFavorite ? "Remove" : "Mark"} as favorite`;
  return /* @__PURE__ */ jsx(FavoriteToggleWithTooltip, {
    ...restProps,
    id,
    ref,
    tooltipProps,
    "aria-label": "Favorite",
    isSelected: isFavorite,
    onChange,
    tooltipTitle
  });
});

export { ContactFavoriteToggle };
