// @salt-ds/lab/dist-es/contact-details/ContactMetadataItem.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { makePrefixer, Text } from '@salt-ds/core';
import { forwardRef } from 'react';
import { MailLinkComponent } from './MailLinkComponent.js';

const withBaseName = makePrefixer("saltContactMetadataItem");
const ContactMetadataItem = forwardRef(function ContactMetadataItem2(props, ref) {
  const { label, value, icon: Icon, ValueComponent, ...restProps } = props;
  return /* @__PURE__ */ jsxs("div", {
    ...restProps,
    ref,
    className: withBaseName(),
    children: [
      !Icon ? /* @__PURE__ */ jsx(Text, {
        maxRows: 1,
        className: withBaseName("label"),
        variant: "secondary",
        children: label
      }) : null,
      Icon ? /* @__PURE__ */ jsx(Icon, {
        className: withBaseName("icon")
      }) : null,
      ValueComponent ? /* @__PURE__ */ jsx(ValueComponent, {
        value
      }) : /* @__PURE__ */ jsx(MailLinkComponent, {
        value
      })
    ]
  });
});

export { ContactMetadataItem };
