// @salt-ds/lab/dist-es/contact-details/ContactTertiaryInfo.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { makePrefixer, useId, Text } from '@salt-ds/core';
import { clsx } from 'clsx';
import { forwardRef, useEffect } from 'react';
import './internal/FavoriteToggleWithTooltip.js';
import { useContactDetailsContext } from './internal/ContactDetailsContext.js';

const withBaseName = makePrefixer("saltContactTertiaryInfo");
const ContactTertiaryInfo = forwardRef(function ContactTertiaryInfo2(props, ref) {
  const { id: idProp, text, icon: Icon, className, ...restProps } = props;
  const { variant, setTertiary, setTertiaryId } = useContactDetailsContext();
  const id = useId(idProp);
  useEffect(() => {
    setTertiary(text);
    setTertiaryId(id);
    return () => {
      setTertiary(void 0);
      setTertiaryId(void 0);
    };
  }, [id, text, setTertiary, setTertiaryId]);
  if (variant === "mini") {
    return null;
  }
  return /* @__PURE__ */ jsxs(Text, {
    ...restProps,
    maxRows: 1,
    styleAs: variant === "default" ? "h4" : void 0,
    id,
    ref,
    className: clsx(withBaseName(), className),
    "data-testid": "tertiary",
    children: [
      Icon ? /* @__PURE__ */ jsx(Icon, {
        className: withBaseName("icon")
      }) : null,
      text
    ]
  });
});

export { ContactTertiaryInfo };
