// @salt-ds/lab/dist-cjs/slider/internal/SliderMarkLabels.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var styles$1 = require('./styles.js');
var utils = require('./utils.js');
var React = require('react');
var window = require('@salt-ds/window');
var styles = require('@salt-ds/styles');
var Slider = require('../Slider.css.js');

const withBaseName = core.makePrefixer("saltSliderMarkLabels");
function SliderMarkLabels(props) {
  const { min, max, marks } = props;
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-slider",
    css: Slider,
    window: targetWindow
  });
  const style = React.useMemo(
    () => styles$1.createSliderMarkLabelsStyle(min, max, marks),
    [min, max, marks]
  );
  const labelStyles = React.useMemo(
    () => styles$1.createSliderMarkLabelStyles(marks),
    [marks]
  );
  return /* @__PURE__ */ jsxRuntime.jsx("div", {
    className: withBaseName(),
    style,
    children: marks.map((mark, i) => {
      return /* @__PURE__ */ jsxRuntime.jsx("div", {
        className: withBaseName("label"),
        style: labelStyles[i],
        children: utils.isLabeledMark(mark) ? mark.label : `${mark}`
      }, i);
    })
  });
}

exports.SliderMarkLabels = SliderMarkLabels;
