// @salt-ds/core/dist-es/card/useInteractableCard.js
import { useState, useEffect } from 'react';

const useInteractableCard = ({
  disabled,
  onKeyUp,
  onKeyDown,
  onClick,
  onBlur
}) => {
  const [keyIsDown, setkeyIsDown] = useState("");
  const [active, setActive] = useState(false);
  const enter = "Enter";
  useEffect(() => {
    const t = setTimeout(() => {
      if (keyIsDown !== enter) {
        setActive(false);
      }
    }, 0);
    return () => {
      clearTimeout(t);
    };
  }, [active, keyIsDown]);
  const handleKeyUp = (event) => {
    setkeyIsDown("");
    setActive(false);
    onKeyUp == null ? void 0 : onKeyUp(event);
  };
  const handleClick = (event) => {
    setActive(true);
    onClick == null ? void 0 : onClick(event);
  };
  const handleBlur = (event) => {
    setActive(false);
    onBlur == null ? void 0 : onBlur(event);
  };
  const handleKeyDown = (event) => {
    if (event.key === enter) {
      setkeyIsDown(event.key);
      setActive(true);
    }
    onKeyDown == null ? void 0 : onKeyDown(event);
  };
  const cardProps = {
    "aria-disabled": disabled ? true : void 0,
    disabled,
    tabIndex: disabled ? -1 : 0,
    onBlur: handleBlur,
    onClick: !disabled ? handleClick : void 0,
    onKeyDown: handleKeyDown,
    onKeyUp: handleKeyUp
  };
  return {
    active,
    cardProps
  };
};

export { useInteractableCard };
