// @salt-ds/lab/dist-es/carousel/CarouselSlide.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { forwardRef, useRef } from 'react';
import { makePrefixer } from '@salt-ds/core';
import { clsx } from 'clsx';

const withBaseName = makePrefixer("saltCarouselSlide");
const CarouselSlide = forwardRef(
  function CarouselSlide2({ ButtonBar, Media, description, title, contentAlignment }, ref) {
    const buttonBarRef = useRef(null);
    return /* @__PURE__ */ jsxs("div", {
      ref,
      children: [
        Media && /* @__PURE__ */ jsx("div", {
          className: withBaseName("mediaContainer"),
          children: Media
        }),
        /* @__PURE__ */ jsxs("div", {
          className: withBaseName("fixedContainer"),
          ref: buttonBarRef,
          children: [
            /* @__PURE__ */ jsxs("div", {
              className: clsx({
                [withBaseName("textContainer")]: contentAlignment === "center",
                [withBaseName("textContainerLeft")]: contentAlignment === "left"
              }),
              children: [
                title && /* @__PURE__ */ jsx("div", {
                  "aria-level": 1,
                  className: withBaseName("titleContainer"),
                  role: "heading",
                  children: title
                }),
                description && /* @__PURE__ */ jsx("div", {
                  className: withBaseName("descriptionContainer"),
                  children: description
                })
              ]
            }),
            ButtonBar && /* @__PURE__ */ jsx("div", {
              className: clsx({
                [withBaseName("buttonBarOverride")]: contentAlignment === "center",
                [withBaseName("buttonBarOverrideLeft")]: contentAlignment === "left"
              }),
              children: /* @__PURE__ */ jsx(ButtonBar, {
                className: clsx({
                  [withBaseName("buttonBarContainer")]: contentAlignment === "center",
                  [withBaseName("buttonBarContainerLeft")]: contentAlignment === "left"
                }),
                stackAtBreakpoint: 0
              })
            })
          ]
        })
      ]
    });
  }
);

export { CarouselSlide };
