// @salt-ds/lab/dist-cjs/input-legacy/InputLegacy.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var clsx = require('clsx');
var React = require('react');
var core = require('@salt-ds/core');
require('../form-field-context-legacy/FormFieldLegacyContext.js');
var useFormFieldLegacyProps = require('../form-field-context-legacy/useFormFieldLegacyProps.js');
var useCursorOnFocus = require('./useCursorOnFocus.js');
var window = require('@salt-ds/window');
var styles = require('@salt-ds/styles');
var InputLegacy$1 = require('./InputLegacy.css.js');

const withBaseName = core.makePrefixer("saltInputLegacy");
function mergeA11yProps(a11yProps = {}, inputProps = {}, misplacedAriaProps) {
  const ariaLabelledBy = clsx.clsx(
    a11yProps["aria-labelledby"],
    inputProps["aria-labelledby"]
  );
  return {
    ...misplacedAriaProps,
    ...a11yProps,
    ...inputProps,
    "aria-labelledby": ariaLabelledBy ? Array.from(new Set(ariaLabelledBy.split(" "))).join(" ") : null
  };
}
const InputLegacy = React.forwardRef(
  function Input({
    "aria-activedescendant": ariaActiveDescendant,
    "aria-expanded": ariaExpanded,
    "aria-owns": ariaOwns,
    className: classNameProp,
    cursorPositionOnFocus,
    disabled,
    emptyReadOnlyMarker = "\u2014",
    endAdornment,
    highlightOnFocus,
    id,
    inputComponent: InputComponent = "input",
    inputProps: inputPropsProp,
    role,
    style,
    value: valueProp,
    defaultValue: defaultValueProp = valueProp === void 0 ? "" : void 0,
    onBlur,
    onChange,
    onFocus,
    onKeyDown,
    onKeyUp,
    onMouseUp,
    onMouseMove,
    onMouseDown,
    readOnly: readOnlyProp,
    renderSuffix,
    startAdornment,
    textAlign = "left",
    type = "text",
    ...other
  }, ref) {
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-input-legacy",
      css: InputLegacy$1,
      window: targetWindow
    });
    const {
      a11yProps: {
        readOnly: a11yReadOnly,
        disabled: a11yDisabled,
        ...restA11y
      } = {},
      setFocused: setFormFieldFocused,
      inFormField
    } = useFormFieldLegacyProps.useFormFieldLegacyProps();
    const [focused, setFocused] = React.useState(false);
    const inputRef = React.useRef(null);
    const handleRef = core.useForkRef(inputRef, ref);
    const cursorOnFocusHelpers = useCursorOnFocus.useCursorOnFocus(inputRef, {
      cursorPositionOnFocus,
      highlightOnFocus
    });
    const isDisabled = disabled || a11yDisabled;
    const isReadOnly = readOnlyProp || a11yReadOnly;
    const misplacedAriaProps = {
      "aria-activedescendant": ariaActiveDescendant,
      "aria-expanded": ariaExpanded,
      "aria-owns": ariaOwns,
      role
    };
    const inputProps = mergeA11yProps(
      restA11y,
      inputPropsProp,
      misplacedAriaProps
    );
    const isEmptyReadOnly = isReadOnly && !defaultValueProp && !valueProp;
    const defaultValue = isEmptyReadOnly ? emptyReadOnlyMarker : defaultValueProp;
    const [value, setValue] = core.useControlled({
      controlled: valueProp,
      default: defaultValue,
      name: "Input",
      state: "value"
    });
    const handleChange = (event) => {
      const value2 = event.target.value;
      setValue(value2);
      onChange == null ? void 0 : onChange(event, value2);
    };
    const handleFocus = (event) => {
      onFocus == null ? void 0 : onFocus(event);
      setFormFieldFocused == null ? void 0 : setFormFieldFocused(true);
      setFocused(true);
    };
    const handleBlur = (event) => {
      onBlur == null ? void 0 : onBlur(event);
      setFormFieldFocused == null ? void 0 : setFormFieldFocused(false);
      setFocused(false);
    };
    const handleMouseMove = (event) => {
      cursorOnFocusHelpers.handleMouseMove(event);
      onMouseMove == null ? void 0 : onMouseMove(event);
    };
    const handleMouseUp = (event) => {
      cursorOnFocusHelpers.handleMouseUp();
      onMouseUp == null ? void 0 : onMouseUp(event);
    };
    const handleMouseDown = (event) => {
      cursorOnFocusHelpers.handleMouseDown();
      onMouseDown == null ? void 0 : onMouseDown(event);
    };
    return /* @__PURE__ */ jsxRuntime.jsxs("div", {
      className: clsx.clsx(
        withBaseName(),
        {
          [withBaseName(`${textAlign}TextAlign`)]: textAlign,
          [withBaseName("formField")]: inFormField,
          [withBaseName("focused")]: focused && !inFormField,
          [withBaseName("disabled")]: isDisabled,
          [withBaseName("inputAdornedStart")]: startAdornment,
          [withBaseName("inputAdornedEnd")]: endAdornment
        },
        classNameProp
      ),
      style,
      ...other,
      children: [
        startAdornment && /* @__PURE__ */ jsxRuntime.jsx("div", {
          className: withBaseName("prefixContainer"),
          children: startAdornment
        }),
        /* @__PURE__ */ jsxRuntime.jsx(InputComponent, {
          type,
          id,
          ...inputProps,
          className: clsx.clsx(withBaseName("input"), inputProps == null ? void 0 : inputProps.className),
          disabled: isDisabled,
          ref: handleRef,
          value,
          onBlur: handleBlur,
          onChange: handleChange,
          onKeyDown,
          onKeyUp,
          onFocus: handleFocus,
          onMouseDown: handleMouseDown,
          onMouseUp: handleMouseUp,
          onMouseMove: handleMouseMove,
          readOnly: isReadOnly
        }),
        endAdornment && /* @__PURE__ */ jsxRuntime.jsx("div", {
          className: withBaseName("suffixContainer"),
          children: endAdornment
        }),
        renderSuffix == null ? void 0 : renderSuffix({ disabled, focused })
      ]
    });
  }
);

exports.InputLegacy = InputLegacy;
