// @salt-ds/lab/dist-es/menu-button/MenuButtonTrigger.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { forwardRef } from 'react';
import { clsx } from 'clsx';
import { makePrefixer, Button } from '@salt-ds/core';
import { ChevronDownIcon } from '@salt-ds/icons';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './MenuButtonTrigger.css.js';

const withBaseName = makePrefixer("saltMenuButtonTrigger");
const MenuButtonTrigger = forwardRef(function MenuButtonTrigger2({ className, hideCaret, isMenuOpen, children, ...rest }, ref) {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-menu-button-trigger",
    css: css_248z,
    window: targetWindow
  });
  return /* @__PURE__ */ jsxs(Button, {
    ref,
    className: clsx(withBaseName(), className, {
      [withBaseName("buttonOpen")]: isMenuOpen
    }),
    "data-testid": "menu-button-trigger",
    ...rest,
    children: [
      children,
      !hideCaret && /* @__PURE__ */ jsx(ChevronDownIcon, {
        className: withBaseName("caretIcon"),
        "data-testid": "menu-button-trigger-caret"
      })
    ]
  });
});

export { MenuButtonTrigger };
