// @salt-ds/lab/dist-es/button-bar/OrderedButton.js
import { jsx } from 'react/jsx-runtime';
import { forwardRef, useContext } from 'react';
import { clsx } from 'clsx';
import { makePrefixer, Button, capitalize } from '@salt-ds/core';
import { ButtonBarContext } from './internal/ButtonBarContext.js';
import { useDescendant } from './internal/useDescendant.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './OrderedButton.css.js';

const withBasename = makePrefixer("saltOrderedButton");
const OrderedButton = forwardRef(
  function OrderedButton2({
    className,
    align: alignProp,
    order,
    alignLeftOrder,
    stackOrder,
    variant,
    ...restProps
  }, ref) {
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-ordered-button",
      css: css_248z,
      window: targetWindow
    });
    const index = useDescendant({ order, stackOrder, alignLeftOrder, variant });
    const {
      matches,
      align: alignContext,
      alignedIndex
    } = useContext(ButtonBarContext);
    const alignFromParent = index === alignedIndex ? alignContext : void 0;
    const align = alignProp || alignFromParent;
    return /* @__PURE__ */ jsx(Button, {
      className: clsx(
        withBasename(),
        {
          [withBasename(`align${align ? capitalize(align) : ""}`)]: align && !matches,
          [withBasename("stacked")]: matches
        },
        className
      ),
      ref,
      variant,
      ...restProps
    });
  }
);

export { OrderedButton };
