// @salt-ds/lab/dist-cjs/common-hooks/keyUtils.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function union(set1, ...sets) {
  const result = new Set(set1);
  for (let set of sets) {
    for (let element of set) {
      result.add(element);
    }
  }
  return result;
}
const ArrowUp = "ArrowUp";
const ArrowDown = "ArrowDown";
const ArrowLeft = "ArrowLeft";
const ArrowRight = "ArrowRight";
const Enter = "Enter";
const Home = "Home";
const End = "End";
const PageUp = "PageUp";
const PageDown = "PageDown";
const Space = " ";
const actionKeys = /* @__PURE__ */ new Set(["Enter", "Delete", " "]);
const focusKeys = /* @__PURE__ */ new Set(["Tab"]);
const arrowLeftRightKeys = /* @__PURE__ */ new Set(["ArrowRight", "ArrowLeft"]);
const navigationKeys = /* @__PURE__ */ new Set([
  Home,
  End,
  PageUp,
  PageDown,
  ArrowDown,
  ArrowUp
]);
const functionKeys = /* @__PURE__ */ new Set([
  "F1",
  "F2",
  "F3",
  "F4",
  "F5",
  "F6",
  "F7",
  "F8",
  "F9",
  "F10",
  "F11",
  "F12"
]);
const specialKeys = union(
  actionKeys,
  navigationKeys,
  arrowLeftRightKeys,
  functionKeys,
  focusKeys
);
const isCharacterKey = (evt) => {
  if (specialKeys.has(evt.key)) {
    return false;
  }
  return evt.key.length === 1 && !evt.ctrlKey && !evt.metaKey && !evt.altKey;
};
const isNavigationKey = ({ key }) => {
  return navigationKeys.has(key);
};

exports.ArrowDown = ArrowDown;
exports.ArrowLeft = ArrowLeft;
exports.ArrowRight = ArrowRight;
exports.ArrowUp = ArrowUp;
exports.End = End;
exports.Enter = Enter;
exports.Home = Home;
exports.PageDown = PageDown;
exports.PageUp = PageUp;
exports.Space = Space;
exports.isCharacterKey = isCharacterKey;
exports.isNavigationKey = isNavigationKey;
