// @salt-ds/lab/dist-cjs/deck-layout/DeckLayout.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var core = require('@salt-ds/core');
var DeckItem = require('../deck-item/DeckItem.js');
require('../responsive/useResizeObserver.js');
var useWidth = require('../responsive/useWidth.js');
var clsx = require('clsx');
var window = require('@salt-ds/window');
var styles = require('@salt-ds/styles');
var DeckLayout$1 = require('./DeckLayout.css.js');

const withBaseName = core.makePrefixer("saltDeckLayout");
const DeckLayout = React.forwardRef(
  function DeckLayout2({
    activeIndex = 0,
    animation,
    className,
    children,
    direction = "horizontal",
    style,
    deckItemProps,
    ...rest
  }, ref) {
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-deck-layout",
      css: DeckLayout$1,
      window: targetWindow
    });
    const [deckItemRef, deckItemWidth] = useWidth.useWidth(true);
    const [deckItemHeight, setDeckItemHeight] = React.useState(0);
    const handleResize = React.useCallback(
      function handleResize2(contentRect) {
        setDeckItemHeight(contentRect.height);
      },
      []
    );
    core.useIsomorphicLayoutEffect(() => {
      if (!deckItemRef.current) {
        return void 0;
      }
      handleResize(deckItemRef.current.getBoundingClientRect());
      const observer = new ResizeObserver(
        ([{ contentRect }]) => {
          handleResize(contentRect);
        }
      );
      observer.observe(deckItemRef.current);
      return () => {
        observer.disconnect();
      };
    }, [deckItemRef, handleResize]);
    const deckLayoutStyles = {
      ...style,
      "--deckLayout-width": `${deckItemWidth}px`,
      "--deckLayout-height": `${deckItemHeight}px`
    };
    const innerStyles = {
      "--deckLayout-transform-value": `-${activeIndex * 100}%`
    };
    return /* @__PURE__ */ jsxRuntime.jsx("div", {
      className: clsx.clsx(withBaseName(), className),
      style: deckLayoutStyles,
      ref,
      ...rest,
      children: /* @__PURE__ */ jsxRuntime.jsx("div", {
        className: clsx.clsx(
          {
            [withBaseName("animate")]: animation
          },
          {
            [withBaseName(`${animation || "slide"}-${direction}`)]: animation
          }
        ),
        style: innerStyles,
        children: React.Children.map(children, (child, index) => {
          return /* @__PURE__ */ jsxRuntime.jsx(DeckItem.DeckItem, {
            ref: deckItemRef,
            index,
            activeIndex,
            animation,
            ...deckItemProps,
            children: child
          });
        })
      })
    });
  }
);

exports.DeckLayout = DeckLayout;
