// @salt-ds/core/dist-cjs/text/Display.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var Text = require('./Text.js');

const Display1 = React.forwardRef(function Display12({ children, className, styleAs = "display1", ...rest }, ref) {
  return /* @__PURE__ */ jsxRuntime.jsx(Text.Text, {
    as: "span",
    styleAs,
    ref,
    ...rest,
    children
  });
});
const Display2 = React.forwardRef(function Display22({ children, className, styleAs = "display2", ...rest }, ref) {
  return /* @__PURE__ */ jsxRuntime.jsx(Text.Text, {
    as: "span",
    styleAs,
    ref,
    ...rest,
    children
  });
});
const Display3 = React.forwardRef(function Display32({ children, className, styleAs = "display3", ...rest }, ref) {
  return /* @__PURE__ */ jsxRuntime.jsx(Text.Text, {
    as: "span",
    styleAs,
    ref,
    ...rest,
    children
  });
});

exports.Display1 = Display1;
exports.Display2 = Display2;
exports.Display3 = Display3;
