// @salt-ds/core/dist-cjs/text/Headings.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var Text = require('./Text.js');

const H1 = React.forwardRef(
  function H12({ children, ...rest }, ref) {
    return /* @__PURE__ */ jsxRuntime.jsx(Text.Text, {
      as: "h1",
      ref,
      ...rest,
      children
    });
  }
);
const H2 = React.forwardRef(
  function H22({ children, ...rest }, ref) {
    return /* @__PURE__ */ jsxRuntime.jsx(Text.Text, {
      as: "h2",
      ref,
      ...rest,
      children
    });
  }
);
const H3 = React.forwardRef(
  function H32({ children, ...rest }, ref) {
    return /* @__PURE__ */ jsxRuntime.jsx(Text.Text, {
      as: "h3",
      ref,
      ...rest,
      children
    });
  }
);
const H4 = React.forwardRef(
  function H42({ children, ...rest }, ref) {
    return /* @__PURE__ */ jsxRuntime.jsx(Text.Text, {
      as: "h4",
      ref,
      ...rest,
      children
    });
  }
);

exports.H1 = H1;
exports.H2 = H2;
exports.H3 = H3;
exports.H4 = H4;
