// @salt-ds/lab/dist-es/stepped-tracker/TrackerStep/TrackerStep.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { forwardRef, useEffect } from 'react';
import { clsx } from 'clsx';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { makePrefixer } from '@salt-ds/core';
import { StepActiveIcon, StepDefaultIcon, StepSuccessIcon } from '@salt-ds/icons';
import { TrackerConnector } from '../TrackerConnector/TrackerConnector.js';
import { useSteppedTrackerContext, useTrackerStepContext } from '../SteppedTrackerContext.js';
import css_248z from './TrackerStep.css.js';

const withBaseName = makePrefixer("saltTrackerStep");
const iconMap = {
  default: StepDefaultIcon,
  completed: StepSuccessIcon
};
const getStateIcon = ({
  isActive,
  state
}) => {
  if (state === "default" && isActive) {
    return StepActiveIcon;
  }
  return iconMap[state];
};
const getState = ({
  isActive,
  state
}) => {
  if (state === "default" && isActive) {
    return "active";
  }
  return state;
};
const useCheckWithinSteppedTracker = (isWithinSteppedTracker) => {
  useEffect(() => {
    if (process.env.NODE_ENV !== "production") {
      if (!isWithinSteppedTracker) {
        console.error(
          "The TrackerStep component must be placed within a SteppedTracker component"
        );
      }
    }
  }, [isWithinSteppedTracker]);
};
const TrackerStep = forwardRef(
  function TrackerStep2(props, ref) {
    const {
      state = "default",
      style,
      className,
      children,
      ...restProps
    } = props;
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-tracker-step",
      css: css_248z,
      window: targetWindow
    });
    const { activeStep, totalSteps, isWithinSteppedTracker } = useSteppedTrackerContext();
    const stepNumber = useTrackerStepContext();
    useCheckWithinSteppedTracker(isWithinSteppedTracker);
    const isActive = activeStep === stepNumber;
    const Icon = getStateIcon({ isActive, state });
    const resolvedState = getState({ isActive, state });
    const connectorState = activeStep > stepNumber ? "active" : "default";
    const hasConnector = stepNumber < totalSteps - 1;
    const innerStyle = {
      ...style,
      "--saltTrackerStep-width": `${100 / totalSteps}%`
    };
    return /* @__PURE__ */ jsxs("li", {
      className: clsx(withBaseName(), withBaseName(resolvedState), className),
      style: innerStyle,
      "aria-current": isActive ? "step" : void 0,
      "data-state": state,
      ref,
      ...restProps,
      children: [
        /* @__PURE__ */ jsx(Icon, {}),
        hasConnector && /* @__PURE__ */ jsx(TrackerConnector, {
          state: connectorState
        }),
        /* @__PURE__ */ jsx("div", {
          className: withBaseName("body"),
          children
        })
      ]
    });
  }
);

export { TrackerStep };
