// @salt-ds/lab/dist-es/content-status/ContentStatus.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { makePrefixer, useId, useAriaAnnouncer, Text, Button } from '@salt-ds/core';
import { clsx } from 'clsx';
import { forwardRef, useEffect } from 'react';
import { StatusIndicator } from './internal/StatusIndicator.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './ContentStatus.css.js';

const withBaseName = makePrefixer("saltContentStatus");
const ContentStatus = forwardRef(
  function ContentStatus2({
    CircularProgressProps,
    SpinnerProps,
    actionLabel,
    buttonRef,
    className,
    children,
    disableAnnouncer,
    message,
    onActionClick,
    status = "info",
    title,
    value,
    ...rest
  }, ref) {
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-content-status",
      css: css_248z,
      window: targetWindow
    });
    const id = useId();
    const hasDefaultActionContent = actionLabel && onActionClick;
    const hasActions = children || hasDefaultActionContent;
    const hasContent = title || message || hasActions;
    const { announce } = useAriaAnnouncer();
    useEffect(() => {
      if (disableAnnouncer)
        return;
      const toBeAnnounced = [];
      if (title) {
        toBeAnnounced.push(title);
      }
      if (message) {
        toBeAnnounced.push(message);
      }
      if (status !== "loading") {
        toBeAnnounced.push(status);
      }
      if (toBeAnnounced.length > 0) {
        announce(toBeAnnounced.join(" "));
      }
    }, [announce, disableAnnouncer, message, status, title]);
    return /* @__PURE__ */ jsxs("div", {
      className: clsx(withBaseName(), className),
      ref,
      ...rest,
      children: [
        /* @__PURE__ */ jsx(StatusIndicator, {
          CircularProgressProps,
          SpinnerProps,
          disableAnnouncer,
          id,
          message,
          status,
          title,
          value,
          ...rest
        }),
        hasContent && /* @__PURE__ */ jsxs("div", {
          "aria-labelledby": id,
          className: clsx(withBaseName("content")),
          id,
          role: "region",
          children: [
            title && /* @__PURE__ */ jsx(Text, {
              className: clsx(withBaseName("title")),
              children: title
            }),
            message && /* @__PURE__ */ jsx(Text, {
              className: clsx(withBaseName("message")),
              children: message
            }),
            hasActions && /* @__PURE__ */ jsx("div", {
              className: clsx(withBaseName("actions")),
              children: children || /* @__PURE__ */ jsx(Button, {
                onClick: onActionClick,
                ref: buttonRef,
                children: actionLabel
              })
            })
          ]
        })
      ]
    });
  }
);

export { ContentStatus };
