// @salt-ds/lab/dist-es/list-deprecated/ListBase.js
import { jsx } from 'react/jsx-runtime';
import { makePrefixer, useForkRef, useId, useIsomorphicLayoutEffect } from '@salt-ds/core';
import { clsx } from 'clsx';
import { createContext, memo, forwardRef, useContext, useRef, useMemo, useImperativeHandle, Children } from 'react';
import { areEqual, VariableSizeList, FixedSizeList } from 'react-window';
import { calcPreferredListHeight } from './internal/calcPreferredListHeight.js';
import { scrollIntoView } from './internal/scrollIntoView.js';
import { useListAutoSizer } from './internal/useListAutoSizer.js';
import { itemToString } from './itemToString.js';
import { ListItemBase } from './ListItemBase.js';
import { ListItemContext } from './ListItemContext.js';
import { useListStateContext } from './ListStateContext.js';
import { useListItem, useVirtualizedListItem } from './useListItem.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './List.css.js';

const withBaseName = makePrefixer("saltListDeprecated");
const ListboxContext = createContext({});
const DefaultItem = memo(function DefaultItem2(props) {
  const { item, itemToString, itemProps } = useListItem(props);
  return /* @__PURE__ */ jsx(ListItemBase, {
    ...itemProps,
    children: itemToString(item)
  });
}, areEqual);
const DefaultVirtualizedItem = memo(
  function DefaultVirtualizedItem2(props) {
    const { item, itemToString, itemProps } = useVirtualizedListItem(props);
    return /* @__PURE__ */ jsx(ListItemBase, {
      ...itemProps,
      children: itemToString(item)
    });
  },
  areEqual
);
const Listbox = forwardRef(function Listbox2(props, ref) {
  const { style, onScroll, children } = props;
  const {
    className,
    borderless,
    disabled,
    disableFocus,
    listRef,
    style: styleProp,
    onScroll: onScrollProp,
    ...restListProps
  } = useContext(ListboxContext);
  const setListRef = useForkRef(ref, listRef);
  const handleScroll = (event) => {
    if (onScroll) {
      onScroll(event);
    }
    if (onScrollProp) {
      onScrollProp(event);
    }
  };
  return /* @__PURE__ */ jsx("div", {
    className: clsx(
      withBaseName(),
      {
        [withBaseName("disabled")]: disabled
      },
      className
    ),
    onScroll: handleScroll,
    ref: setListRef,
    style: { ...style, ...styleProp },
    tabIndex: disabled || disableFocus ? void 0 : 0,
    ...restListProps,
    children
  });
});
const noScrolling = {
  scrollToIndex: (itemIndex) => void 0,
  scrollToItem: (item) => void 0,
  scrollTo: (scrollOffset) => void 0
};
const ListBase = forwardRef(function ListBase2(props, ref) {
  var _a;
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-list-deprecated",
    css: css_248z,
    window: targetWindow
  });
  const { state } = useListStateContext();
  const generatedId = useId(props.id);
  const defaultId = (_a = state.id) != null ? _a : generatedId;
  const sizeStackable = "36px";
  const defaultItemHeight = parseInt(sizeStackable, 10);
  const hasIndexer = typeof props.getItemAtIndex === "function";
  const hasVariableHeight = typeof props.getItemHeight === "function";
  const {
    id = defaultId,
    source = [],
    borderless,
    children,
    disableMouseDown,
    itemTextHighlightPattern,
    itemCount = source.length,
    itemToString: itemToString$1 = itemToString,
    itemGapSize = 1,
    itemHeight = defaultItemHeight,
    getItemHeight = () => itemHeight,
    getItemId = (index) => `${id}-item-${index}`,
    getItemIndex = (item) => source.indexOf(item),
    getItemAtIndex,
    overscanCount = 10,
    displayedItemCount = 10,
    virtualized,
    width,
    height,
    maxWidth,
    maxHeight,
    minWidth,
    minHeight,
    ListPlaceholder,
    ListItem = virtualized ? DefaultVirtualizedItem : DefaultItem,
    listRef: listRefProp,
    ...restProps
  } = props;
  const { highlightedIndex } = state;
  const preferredHeight = height != null ? height : calcPreferredListHeight({
    borderless,
    displayedItemCount,
    itemCount,
    itemHeight,
    getItemHeight,
    itemGapSize
  });
  const [containerRef, autoSize] = useListAutoSizer({
    responsive: width === void 0 || height === void 0,
    height: preferredHeight,
    width
  });
  const virtualizedListRef = useRef(null);
  const listRef = useRef(null);
  const setListRef = useForkRef(listRef, listRefProp);
  const scrollToIndex = (itemIndex) => {
    var _a2;
    scrollIntoView(
      (_a2 = listRef.current) == null ? void 0 : _a2.querySelector(`[data-option-index="${itemIndex}"]`),
      listRef
    );
  };
  const scrollHandles = useMemo(
    () => ({
      scrollToIndex,
      scrollToItem: (item) => {
        scrollToIndex(getItemIndex(item));
      },
      scrollTo: (scrollOffset) => {
        if (listRef.current) {
          listRef.current.scrollTop = scrollOffset;
        }
      }
    }),
    [getItemIndex]
  );
  const virtualizedScrollHandles = useMemo(
    () => ({
      scrollToIndex: (itemIndex) => {
        if (virtualizedListRef.current) {
          virtualizedListRef.current.scrollToItem(itemIndex);
        }
      },
      scrollToItem: (item) => {
        virtualizedListRef.current.scrollToItem(getItemIndex(item));
      },
      scrollTo: (scrollOffset) => {
        virtualizedListRef.current.scrollTo(scrollOffset);
      }
    }),
    [getItemIndex]
  );
  useImperativeHandle(
    ref,
    () => {
      if (virtualized && virtualizedListRef.current) {
        return virtualizedScrollHandles;
      } else if (listRef.current) {
        return scrollHandles;
      } else {
        return noScrolling;
      }
    },
    [virtualized, scrollHandles, virtualizedScrollHandles]
  );
  useIsomorphicLayoutEffect(() => {
    if (highlightedIndex == null) {
      return;
    }
    if (virtualized && virtualizedListRef.current) {
      virtualizedListRef.current.scrollToItem(highlightedIndex);
    } else if (listRef.current) {
      scrollToIndex(highlightedIndex);
    }
  }, [highlightedIndex, virtualized]);
  const renderList = () => {
    if (Children.count(children)) {
      return /* @__PURE__ */ jsx(Listbox, {
        style: autoSize,
        children: /* @__PURE__ */ jsx(ListItemContext.Provider, {
          value: {
            disableMouseDown,
            getItemId,
            getItemHeight,
            itemToString: itemToString$1,
            itemTextHighlightPattern
          },
          children
        })
      });
    }
    if (virtualized) {
      const VirtualizedList = hasVariableHeight ? VariableSizeList : FixedSizeList;
      return /* @__PURE__ */ jsx(ListItemContext.Provider, {
        value: {
          disableMouseDown,
          getItemId,
          itemToString: itemToString$1,
          itemTextHighlightPattern
        },
        children: /* @__PURE__ */ jsx(VirtualizedList, {
          height: autoSize.height,
          itemCount,
          itemData: source,
          itemSize: hasVariableHeight ? getItemHeight : itemHeight,
          outerElementType: Listbox,
          overscanCount,
          ref: virtualizedListRef,
          width: autoSize.width,
          children: ListItem
        })
      });
    }
    return /* @__PURE__ */ jsx(Listbox, {
      style: autoSize,
      children: /* @__PURE__ */ jsx(ListItemContext.Provider, {
        value: {
          disableMouseDown,
          getItemId,
          getItemHeight,
          itemToString: itemToString$1,
          itemTextHighlightPattern
        },
        children: (hasIndexer ? Array.from({ length: itemCount }) : source).map(
          (item, index) => /* @__PURE__ */ jsx(ListItem, {
            index,
            item: hasIndexer ? getItemAtIndex(index) : item
          }, getItemId(index))
        )
      })
    });
  };
  return /* @__PURE__ */ jsx("div", {
    className: clsx(withBaseName("wrapper"), {
      [withBaseName("borderless")]: borderless
    }),
    ref: containerRef,
    style: {
      minWidth,
      minHeight,
      width: width != null ? width : "100%",
      height: height != null ? height : "100%",
      maxWidth: maxWidth != null ? maxWidth : width,
      maxHeight: maxHeight != null ? maxHeight : preferredHeight
    },
    children: itemCount === 0 && ListPlaceholder !== void 0 ? /* @__PURE__ */ jsx(ListPlaceholder, {
      style: autoSize
    }) : /* @__PURE__ */ jsx(ListboxContext.Provider, {
      value: {
        ...restProps,
        listRef: setListRef,
        id,
        borderless
      },
      children: renderList()
    })
  });
});

export { ListBase };
