// @salt-ds/lab/dist-es/skip-link/SkipLink.js
import { jsx } from 'react/jsx-runtime';
import { forwardRef } from 'react';
import { makePrefixer, Link } from '@salt-ds/core';
import { clsx } from 'clsx';
import { useManageFocusOnTarget } from './internal/useManageFocusOnTarget.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './SkipLink.css.js';

const withBaseName = makePrefixer("saltSkipLink");
const SkipLink = forwardRef(
  function SkipLink2({ className, targetRef, ...rest }, ref) {
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-skip-link",
      css: css_248z,
      window: targetWindow
    });
    const targetClass = clsx(withBaseName("target"), className);
    const eventHandlers = useManageFocusOnTarget({ targetRef, targetClass });
    return /* @__PURE__ */ jsx("li", {
      children: /* @__PURE__ */ jsx(Link, {
        ...eventHandlers,
        ...rest,
        className: clsx(withBaseName(), className),
        ref
      })
    });
  }
);

export { SkipLink };
