// @salt-ds/lab/dist-cjs/query-input/internal/CategoryList.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var core = require('@salt-ds/core');
require('../../list/Highlighter.js');
require('../../list/ListItem.js');
var List = require('../../list/List.js');
require('../../common-hooks/collectionProvider.js');
require('../../common-hooks/keyUtils.js');
require('../../responsive/useResizeObserver.js');
require('../../list/VirtualizedList.js');
var CategoryListItem = require('./CategoryListItem.js');
var CategoryListContext = require('./CategoryListContext.js');

const withBaseName = core.makePrefixer("saltCategoryList");
const getCategoryLabel = (category) => category.name;
function CategoryList(props) {
  const {
    categories,
    rootWidth,
    onCategorySelect,
    highlightedCategoryIndex,
    setHighlightedCategoryIndex
  } = props;
  const measureRef = React.useRef(null);
  const [measuredCategories, setMeasuredCategories] = React.useState([]);
  const [contextValue, setContextValue] = React.useState({
    width: 0
  });
  core.useIsomorphicLayoutEffect(() => {
    setTimeout(() => {
      if (measureRef.current) {
        const width = measureRef.current.offsetWidth;
        setMeasuredCategories(categories);
        setContextValue({
          width
        });
      }
    }, 0);
  }, [categories, setMeasuredCategories, setContextValue]);
  const onSelect = React.useCallback(
    (_, item) => {
      onCategorySelect(item);
    },
    [onCategorySelect]
  );
  const isMeasuring = measuredCategories !== categories;
  if (isMeasuring) {
    return /* @__PURE__ */ jsxRuntime.jsx("div", {
      ref: measureRef,
      className: withBaseName("categoryWidthMeasure"),
      children: categories.map((c, i) => {
        return /* @__PURE__ */ jsxRuntime.jsx("div", {
          className: withBaseName("category"),
          children: c.name
        }, i);
      })
    });
  }
  return /* @__PURE__ */ jsxRuntime.jsx(CategoryListContext.CategoryListContext.Provider, {
    value: contextValue,
    children: /* @__PURE__ */ jsxRuntime.jsx(List.List, {
      ListItem: CategoryListItem.CategoryListItem,
      "data-testid": "category-list",
      highlightedIndex: highlightedCategoryIndex,
      itemToString: getCategoryLabel,
      onHighlight: setHighlightedCategoryIndex,
      onSelect,
      source: categories,
      width: rootWidth
    })
  });
}

exports.CategoryList = CategoryList;
