// @salt-ds/lab/dist-cjs/query-input/internal/QueryInputBody.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var icons = require('@salt-ds/icons');
var React = require('react');
var TokenizedInput = require('../../tokenized-input/TokenizedInput.js');
require('../../tokenized-input/TokenizedInputBase.js');
require('clipboard-copy');
require('../../form-field-context-legacy/FormFieldLegacyContext.js');

const withBaseName = core.makePrefixer("saltQueryInputBody");
const QueryInputBody = React.forwardRef(
  function QueryInputBody2(props, ref) {
    const {
      inputRef,
      disabled,
      selectedItems,
      onInputFocus,
      onInputBlur,
      showCategory,
      inputValue,
      onInputChange,
      isFocused,
      onInputClear,
      onSelectedItemsChange,
      onKeyDown,
      onInputClick,
      booleanOperator,
      onBooleanOperatorChange
    } = props;
    const itemToString = React.useMemo(() => {
      if (showCategory) {
        return (item) => [item.category, item.value].join(": ");
      }
      return (item) => item.value;
    }, [showCategory]);
    const stringToItem = (selectedItems2, value) => {
      return {
        category: null,
        value
      };
    };
    const onChange = (event) => {
      const newBooleanOperator = event.currentTarget.value;
      onBooleanOperatorChange(newBooleanOperator);
    };
    return /* @__PURE__ */ jsxRuntime.jsxs("div", {
      ref,
      className: withBaseName(),
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(icons.FilterIcon, {}),
        /* @__PURE__ */ jsxRuntime.jsx(TokenizedInput.TokenizedInput, {
          inputRef,
          disabled,
          className: withBaseName("input"),
          selectedItems,
          itemToString,
          stringToItem,
          onInputFocus,
          onInputBlur,
          value: inputValue,
          onClick: onInputClick,
          onInputChange,
          expanded: isFocused,
          onClear: onInputClear,
          onChange: onSelectedItemsChange,
          onKeyDown
        }),
        /* @__PURE__ */ jsxRuntime.jsx("div", {
          className: withBaseName("separator")
        }),
        /* @__PURE__ */ jsxRuntime.jsxs(core.ToggleButtonGroup, {
          className: withBaseName("buttonGroup"),
          "data-testid": "boolean-selector",
          value: booleanOperator,
          onChange,
          children: [
            /* @__PURE__ */ jsxRuntime.jsx(core.ToggleButton, {
              value: "and",
              children: "And"
            }),
            /* @__PURE__ */ jsxRuntime.jsx(core.ToggleButton, {
              value: "or",
              children: "Or"
            })
          ]
        })
      ]
    });
  }
);

exports.QueryInputBody = QueryInputBody;
