// @salt-ds/lab/dist-cjs/query-input/internal/ValueList.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var icons = require('@salt-ds/icons');
require('../../list/Highlighter.js');
var ListItem = require('../../list/ListItem.js');
var List = require('../../list/List.js');
require('react');
require('../../common-hooks/collectionProvider.js');
require('../../common-hooks/keyUtils.js');
require('../../responsive/useResizeObserver.js');
require('../../list/VirtualizedList.js');

const withBaseName = core.makePrefixer("saltQueryInputValueList");
function ValueList(props) {
  const {
    category,
    rootWidth,
    onBack,
    selectedValues,
    onValueToggle,
    highlightedValueIndex,
    setHighlightedValueIndex
  } = props;
  const handleSelect = (_, value) => {
    onValueToggle(category, value);
  };
  const items = category ? category.values.map((value) => /* @__PURE__ */ jsxRuntime.jsx(ListItem.ListItem, {
    item: value,
    children: value
  }, value)) : [];
  return /* @__PURE__ */ jsxRuntime.jsx("div", {
    className: withBaseName(),
    tabIndex: 0,
    "data-testid": "value-list",
    children: /* @__PURE__ */ jsxRuntime.jsxs(List.List, {
      borderless: true,
      checkable: false,
      highlightedIndex: highlightedValueIndex,
      onHighlight: setHighlightedValueIndex,
      onSelect: handleSelect,
      selectionStrategy: "multiple",
      selected: selectedValues,
      width: rootWidth,
      children: [
        /* @__PURE__ */ jsxRuntime.jsxs(ListItem.ListItem, {
          className: withBaseName("back"),
          onClick: onBack,
          selectable: false,
          children: [
            /* @__PURE__ */ jsxRuntime.jsx(icons.ChevronLeftIcon, {}),
            /* @__PURE__ */ jsxRuntime.jsx("div", {
              className: withBaseName("category"),
              children: category ? category.name : ""
            })
          ]
        }),
        items
      ]
    })
  });
}

exports.ValueList = ValueList;
