// @salt-ds/lab/dist-cjs/list/Highlighter.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var escapeRegExp = require('../utils/escapeRegExp.js');
require('react');
var window = require('@salt-ds/window');
var styles = require('@salt-ds/styles');
var Highlighter$1 = require('./Highlighter.css.js');

const withBaseName = core.makePrefixer("saltHighlighter");
const Highlighter = (props) => {
  const { matchPattern, text = "" } = props;
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-hightligher",
    css: Highlighter$1,
    window: targetWindow
  });
  const matchRegex = typeof matchPattern === "string" ? new RegExp(`(${escapeRegExp.escapeRegExp(matchPattern)})`, "gi") : matchPattern;
  if (matchRegex === void 0 || matchPattern === "") {
    return /* @__PURE__ */ jsxRuntime.jsx(jsxRuntime.Fragment, {
      children: text
    });
  }
  return /* @__PURE__ */ jsxRuntime.jsx("span", {
    children: text.split(matchRegex).map(
      (part, index) => part.match(matchRegex) ? /* @__PURE__ */ jsxRuntime.jsx("strong", {
        className: withBaseName("highlight"),
        children: part
      }, `${index}-${part}`) : part
    )
  });
};

exports.Highlighter = Highlighter;
