// @salt-ds/lab/dist-cjs/list/List.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var core = require('@salt-ds/core');
var clsx = require('clsx');
require('../common-hooks/collectionProvider.js');
var itemToString = require('../common-hooks/itemToString.js');
require('../common-hooks/keyUtils.js');
var useCollectionItems = require('../common-hooks/useCollectionItems.js');
var useImperativeScrollingAPI = require('../common-hooks/useImperativeScrollingAPI.js');
var useKeyboardNavigation = require('../common-hooks/useKeyboardNavigation.js');
require('../responsive/useResizeObserver.js');
var isSelected = require('../common-hooks/utils/isSelected.js');
var ListItem = require('./ListItem.js');
var useList = require('./useList.js');
var useListHeight = require('./useListHeight.js');
var window = require('@salt-ds/window');
var styles = require('@salt-ds/styles');
var List$1 = require('./List.css.js');

const defaultEmptyMessage = "No data to display";
const withBaseName = core.makePrefixer("saltList");
const List = React.forwardRef(function List2({
  ListItem: ListItem$1 = ListItem.ListItem,
  ListPlaceholder,
  borderless,
  children,
  className,
  collapsibleHeaders = false,
  defaultHighlightedIndex,
  defaultSelected,
  disabled: listDisabled = false,
  disableFocus = false,
  disableTypeToSelect,
  displayedItemCount = 10,
  emptyMessage,
  focusVisible: focusVisibleProp,
  getItemHeight: getItemHeightProp,
  getItemId,
  height,
  highlightedIndex: highlightedIndexProp,
  id: idProp,
  itemGapSize = 0,
  itemHeight: itemHeightProp,
  itemTextHighlightPattern,
  itemToString: itemToString$1 = itemToString.itemToString,
  listHandlers: listHandlersProp,
  maxHeight,
  maxWidth,
  minHeight,
  minWidth,
  onSelect,
  onSelectionChange,
  onHighlight,
  restoreLastFocus,
  selected: selectedProp,
  selectionStrategy,
  checkable = selectionStrategy === "multiple",
  scrollingApiRef,
  selectionKeys,
  showEmptyMessage = false,
  source,
  style: styleProp,
  stickyHeaders,
  tabToSelect,
  width,
  ...htmlAttributes
}, forwardedRef) {
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-list",
    css: List$1,
    window: targetWindow
  });
  const id = core.useIdMemo(idProp);
  const rootRef = React.useRef(null);
  const contentRef = React.useRef(null);
  const rowHeightProxyRef = React.useRef(null);
  const collectionHook = useCollectionItems.useCollectionItems({
    id,
    label: "List",
    source,
    children,
    options: {
      collapsibleHeaders,
      getItemId,
      itemToString: itemToString$1
    }
  });
  const { preferredHeight } = useListHeight.useListHeight({
    borderless,
    displayedItemCount,
    getItemHeight: getItemHeightProp,
    height,
    itemCount: collectionHook.data.length,
    itemGapSize,
    itemHeight: itemHeightProp,
    rowHeightRef: rowHeightProxyRef
  });
  const {
    focusVisible,
    highlightedIndex,
    listControlProps,
    listHandlers,
    listItemHeaderHandlers,
    scrollIntoView,
    selected
  } = useList.useList({
    collapsibleHeaders,
    collectionHook,
    containerRef: rootRef,
    contentRef,
    defaultHighlightedIndex,
    defaultSelected: collectionHook.itemToCollectionItem(defaultSelected),
    disabled: listDisabled,
    disableTypeToSelect,
    highlightedIndex: highlightedIndexProp,
    label: id,
    listHandlers: listHandlersProp,
    onSelect,
    onSelectionChange,
    onHighlight,
    restoreLastFocus,
    selected: collectionHook.itemToCollectionItem(selectedProp),
    selectionStrategy,
    selectionKeys,
    stickyHeaders,
    tabToSelect
  });
  useImperativeScrollingAPI.useImperativeScrollingAPI({
    collectionHook,
    forwardedRef: scrollingApiRef,
    scrollableRef: rootRef,
    scrollIntoView
  });
  const appliedFocusVisible = focusVisibleProp != null ? focusVisibleProp : focusVisible;
  const createHeader = function createHeader2(idx, headerId, title, expanded) {
    const header = /* @__PURE__ */ React.createElement(ListItem$1, {
      ...listItemHeaderHandlers,
      className: clsx.clsx("saltListItemHeader", {
        focusVisible: collapsibleHeaders && appliedFocusVisible === idx.value
      }),
      "aria-expanded": expanded,
      "data-idx": collapsibleHeaders ? idx.value : void 0,
      "data-highlighted": idx.value === highlightedIndex || void 0,
      "data-sticky": stickyHeaders,
      "data-selectable": false,
      id: headerId,
      key: `header-${idx.value}`,
      label: title
    });
    idx.value += 1;
    return header;
  };
  const getItemHeight = getItemHeightProp === void 0 ? () => itemHeightProp : getItemHeightProp;
  function renderCollectionItem(list, item, idx) {
    var _a;
    const { disabled, value, id: itemId, label } = item;
    const isChildItem = React.isValidElement(value);
    const listItemProps = {
      className: clsx.clsx({
        saltHighlighted: idx.value === highlightedIndex,
        saltFocusVisible: appliedFocusVisible === idx.value
      }),
      disabled: disabled || listDisabled,
      id: itemId,
      item: isChildItem ? void 0 : (_a = item == null ? void 0 : item.value) != null ? _a : void 0,
      itemHeight: getItemHeight(idx.value),
      itemTextHighlightPattern,
      key: itemId,
      "data-idx": idx.value,
      label,
      role: "option",
      selected: isSelected.isSelected(selected, item),
      showCheckbox: checkable
    };
    list.push(
      isChildItem ? React.cloneElement(value, listItemProps) : /* @__PURE__ */ jsxRuntime.jsx(ListItem$1, {
        ...listItemProps
      })
    );
    idx.value += 1;
  }
  const addGroup = function addGroup2(list, items, idx) {
    const { count = 0, id: id2, expanded, label = "" } = items[idx.value];
    const header = createHeader(idx, id2, label, expanded);
    const childItems = expanded !== false ? [header].concat(
      renderCollectionItems(items, idx, idx.value + count) || []
    ) : header;
    list.push(
      /* @__PURE__ */ jsxRuntime.jsx("div", {
        role: "group",
        children: childItems
      }, id2)
    );
  };
  const renderCollectionItems = (items, idx = { value: 0 }, end = items.length) => {
    const listItems = [];
    while (idx.value < end) {
      const item = items[idx.value];
      if (item.header) {
        listItems.push(
          createHeader(idx, item.id, item.label, item.expanded === false)
        );
      } else if (item.childNodes) {
        addGroup(listItems, items, idx);
      } else {
        renderCollectionItem(listItems, item, idx);
      }
    }
    return listItems;
  };
  function renderEmpty() {
    if (emptyMessage || showEmptyMessage) {
      return /* @__PURE__ */ jsxRuntime.jsx("span", {
        className: withBaseName("empty-message"),
        children: emptyMessage != null ? emptyMessage : defaultEmptyMessage
      });
    } else {
      return null;
    }
  }
  const renderContent = () => {
    if (collectionHook.data.length) {
      return renderCollectionItems(collectionHook.data);
    } else {
      renderEmpty();
    }
  };
  const contentHeight = "auto";
  const sizeStyles = {
    "--list-item-gap": itemGapSize ? `${itemGapSize}px` : void 0,
    minWidth,
    minHeight,
    width: width != null ? width : "100%",
    height: height != null ? height : "100%",
    maxWidth: maxWidth != null ? maxWidth : width,
    maxHeight: maxHeight != null ? maxHeight : preferredHeight
  };
  return /* @__PURE__ */ jsxRuntime.jsxs("div", {
    "aria-multiselectable": selectionStrategy === "multiple" || selectionStrategy === "extended" || selectionStrategy === "extended-multi-range" || void 0,
    ...htmlAttributes,
    ...listHandlers,
    ...listControlProps,
    className: clsx.clsx(withBaseName(), className, {
      [withBaseName("borderless")]: borderless,
      saltDisabled: listDisabled,
      [withBaseName("collapsible")]: collapsibleHeaders,
      saltFocusVisible: highlightedIndex === useKeyboardNavigation.LIST_FOCUS_VISIBLE
    }),
    id: `${id}`,
    ref: core.useForkRef(rootRef, forwardedRef),
    role: "listbox",
    style: { ...styleProp, ...sizeStyles },
    tabIndex: listDisabled || disableFocus ? void 0 : 0,
    children: [
      /* @__PURE__ */ jsxRuntime.jsx(ListItem.ListItemProxy, {
        ref: rowHeightProxyRef
      }),
      collectionHook.data.length === 0 && ListPlaceholder !== void 0 ? /* @__PURE__ */ jsxRuntime.jsx(jsxRuntime.Fragment, {
        children: /* @__PURE__ */ jsxRuntime.jsx(ListPlaceholder, {})
      }) : /* @__PURE__ */ jsxRuntime.jsx("div", {
        className: withBaseName("scrollingContentContainer"),
        ref: contentRef,
        style: { height: contentHeight },
        children: renderContent()
      })
    ]
  });
});

exports.List = List;
