// @salt-ds/lab/dist-cjs/list/VirtualizedList.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var clsx = require('clsx');
var React = require('react');
require('../common-hooks/collectionProvider.js');
require('../common-hooks/keyUtils.js');
var useCollectionItems = require('../common-hooks/useCollectionItems.js');
var useImperativeScrollingAPI = require('../common-hooks/useImperativeScrollingAPI.js');
require('../responsive/useResizeObserver.js');
var isSelected = require('../common-hooks/utils/isSelected.js');
var useListHeight = require('./useListHeight.js');
var ListItem = require('./ListItem.js');
var useList = require('./useList.js');
var useVirtualization = require('./useVirtualization.js');
var window = require('@salt-ds/window');
var styles = require('@salt-ds/styles');
var List = require('./List.css.js');

const defaultEmptyMessage = "No data to display";
const withBaseName = core.makePrefixer("saltList");
const VirtualizedList = React.forwardRef(function List$1({
  borderless,
  children,
  className,
  collapsibleHeaders = false,
  defaultHighlightedIndex: defaultHighlightedIdx,
  defaultSelected,
  disabled: listDisabled = false,
  disableFocus = false,
  disableTypeToSelect,
  displayedItemCount = 10,
  emptyMessage,
  getItemHeight,
  getItemId,
  height,
  highlightedIndex: highlightedIdxProp,
  id: idProp,
  itemGapSize = 0,
  itemHeight: itemHeightProp,
  itemTextHighlightPattern,
  itemToString,
  maxHeight,
  maxWidth,
  minHeight,
  minWidth,
  onSelect,
  onSelectionChange,
  onHighlight,
  restoreLastFocus,
  selected: selectedProp,
  selectionStrategy,
  scrollingApiRef,
  selectionKeys,
  showEmptyMessage = false,
  source,
  style: styleProp,
  stickyHeaders,
  tabToSelect,
  width,
  ...htmlAttributes
}, forwardedRef) {
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-list",
    css: List,
    window: targetWindow
  });
  const id = core.useIdMemo(idProp);
  const rootRef = React.useRef(null);
  const rowHeightProxyRef = React.useRef(null);
  const collectionHook = useCollectionItems.useCollectionItems({
    id,
    label: "List",
    source,
    children,
    options: {
      collapsibleHeaders,
      getItemId,
      itemToString
    }
  });
  const { preferredHeight } = useListHeight.useListHeight({
    borderless,
    displayedItemCount,
    height,
    itemCount: collectionHook.data.length,
    itemGapSize,
    itemHeight: itemHeightProp,
    rowHeightRef: rowHeightProxyRef
  });
  const {
    focusVisible,
    highlightedIndex,
    listControlProps,
    listHandlers,
    scrollIntoView,
    selected
  } = useList.useList({
    collapsibleHeaders,
    collectionHook,
    containerRef: rootRef,
    defaultHighlightedIndex: defaultHighlightedIdx,
    defaultSelected: collectionHook.itemToCollectionItem(defaultSelected),
    disabled: listDisabled,
    disableTypeToSelect,
    highlightedIndex: highlightedIdxProp,
    label: id,
    onSelect,
    onSelectionChange,
    onHighlight,
    restoreLastFocus,
    selected: collectionHook.itemToCollectionItem(selectedProp),
    selectionStrategy,
    selectionKeys,
    stickyHeaders,
    tabToSelect
  });
  const {
    rows: data,
    contentHeight,
    onVerticalScroll: onScroll
  } = useVirtualization.useVirtualization({
    viewportRef: rootRef,
    data: collectionHook.data,
    itemGapSize
  });
  useImperativeScrollingAPI.useImperativeScrollingAPI({
    collectionHook,
    forwardedRef: scrollingApiRef,
    scrollableRef: rootRef,
    scrollIntoView
  });
  function addItem(list, row, idx) {
    const [key, offset, pos, item] = row;
    const index = pos - 1;
    list.push(
      /* @__PURE__ */ jsxRuntime.jsx(ListItem.ListItem, {
        "aria-setsize": collectionHook.data.length,
        "aria-posinset": pos,
        className: clsx.clsx(className, {
          saltHighlighted: index === highlightedIndex,
          saltFocusVisible: focusVisible === index
        }),
        "data-idx": index,
        "data-offset": offset,
        role: "option",
        selected: isSelected.isSelected(selected, item),
        id: item.id,
        style: {
          transform: `translate3d(0px, ${offset}px, 0px)`
        },
        children: item.label
      }, key)
    );
    idx.value += 1;
  }
  function renderItems(data2, idx = { value: 0 }, end = data2.length) {
    const listItems = [];
    while (idx.value < end) {
      const item = data2[idx.value];
      addItem(listItems, item, idx);
    }
    return listItems;
  }
  function renderEmpty() {
    if (emptyMessage || showEmptyMessage) {
      return /* @__PURE__ */ jsxRuntime.jsx("span", {
        className: withBaseName("empty-message"),
        children: emptyMessage != null ? emptyMessage : defaultEmptyMessage
      });
    } else {
      return null;
    }
  }
  const renderContent = () => {
    if (data.length) {
      return renderItems(data);
    } else {
      renderEmpty();
    }
  };
  const sizeStyles = {
    "--list-item-gap": itemGapSize ? `${itemGapSize}px` : void 0,
    minWidth,
    minHeight,
    width: width != null ? width : "100%",
    height: height != null ? height : "100%",
    maxWidth: maxWidth != null ? maxWidth : width,
    maxHeight: maxHeight != null ? maxHeight : preferredHeight
  };
  return /* @__PURE__ */ jsxRuntime.jsx("div", {
    ...htmlAttributes,
    ...listHandlers,
    ...listControlProps,
    className: clsx.clsx(withBaseName(), className, withBaseName("virtualized")),
    id: `${id}`,
    ref: core.useForkRef(rootRef, forwardedRef),
    role: "listbox",
    onScroll,
    style: { ...styleProp, ...sizeStyles },
    tabIndex: listDisabled || disableFocus ? void 0 : 0,
    children: /* @__PURE__ */ jsxRuntime.jsxs("div", {
      className: withBaseName("scrollingContentContainer"),
      style: { height: contentHeight },
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(ListItem.ListItemProxy, {
          ref: rowHeightProxyRef
        }),
        renderContent()
      ]
    })
  });
});

exports.VirtualizedList = VirtualizedList;
