// @salt-ds/lab/dist-cjs/tabs/Tabstrip.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var core = require('@salt-ds/core');
var icons = require('@salt-ds/icons');
var clsx = require('clsx');
require('../dropdown/DropdownBase.js');
require('../dropdown/DropdownButton.js');
var Dropdown = require('../dropdown/Dropdown.js');
require('../responsive/useResizeObserver.js');
var useOverflowCollectionItems = require('../responsive/useOverflowCollectionItems.js');
var useOverflowLayout = require('../responsive/useOverflowLayout.js');
require('../form-field-context-legacy/FormFieldLegacyContext.js');
var Tab = require('./Tab.js');
var TabActivationIndicator = require('./TabActivationIndicator.js');
var useTabstrip = require('./useTabstrip.js');
var window = require('@salt-ds/window');
var styles = require('@salt-ds/styles');
var ThemeTabstrip = require('./ThemeTabstrip.css.js');
var Tabstrip$1 = require('./Tabstrip.css.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

const withBaseName = core.makePrefixer("saltTabstrip");
const ADD_TAB_LABEL = "Create Tab";
const tabDescriptors = (tabs) => tabs && tabs.map(
  (tab) => typeof tab === "string" ? { label: tab } : tab
);
const Tabstrip = React.forwardRef(function Tabstrip2({
  activeTabIndex: activeTabIndexProp,
  allowDragDrop = false,
  centered = false,
  children,
  className: classNameProp,
  defaultSource,
  defaultActiveTabIndex,
  editing,
  variant,
  enableAddTab = false,
  enableCloseTab,
  enableRenameTab,
  id: idProp,
  keyBoardActivation = "manual",
  onAddTab,
  onActiveChange,
  onCloseTab,
  onEnterEditMode,
  onExitEditMode,
  onMouseDown,
  onMoveTab,
  orientation = "horizontal",
  overflowMenu: overflowMenuProp = true,
  promptForNewTabName = true,
  showActivationIndicator = true,
  source,
  title,
  ...htmlAttributes
}, forwardedRef) {
  var _a;
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-tab-strip",
    css: Tabstrip$1,
    window: targetWindow
  });
  styles.useComponentCssInjection({
    testId: "salt-theme-tab-strip",
    css: ThemeTabstrip,
    window: targetWindow
  });
  const root = React.useRef(null);
  const activeRef = React.useRef(
    activeTabIndexProp || defaultActiveTabIndex || 0
  );
  const overflowItemsRef = React.useRef([]);
  const [showOverflowMenu, _setShowOverflowMenu] = React.useState(false);
  const setShowOverflowMenu = React.useCallback((value) => {
    _setShowOverflowMenu(value);
  }, []);
  const tabstripId = core.useIdMemo(idProp);
  const injectedItems = enableAddTab ? [
    {
      source: { label: "Add Tab", position: -1, priority: 1 }
    }
  ] : void 0;
  const collectionHook = useOverflowCollectionItems.useOverflowCollectionItems({
    children,
    defaultSource: tabDescriptors(defaultSource),
    id: tabstripId,
    injectedItems,
    label: "Tabstrip",
    orientation,
    source: tabDescriptors(source),
    options: {
      closeable: enableCloseTab,
      editable: enableRenameTab,
      getPriority: (item, index) => {
        return index === activeRef.current ? 1 : void 0;
      }
    }
  });
  const childCount = React.useRef(React__default["default"].Children.count(children));
  const getChildren = () => {
    if (React__default["default"].Children.count(children) === 0) {
      return void 0;
    } else {
      return React__default["default"].Children.toArray(children);
    }
  };
  const [innerContainerRef, switchOverflowPriorities] = useOverflowLayout.useOverflowLayout({
    collectionHook,
    disableOverflow: overflowMenuProp === false,
    id: tabstripId,
    orientation,
    label: "Tabstrip"
  });
  overflowItemsRef.current = collectionHook.data;
  const overflowedItems = collectionHook.data.filter((item) => item.overflowed);
  const tabMovedHandler = React.useCallback(
    (fromIndex, toIndex) => {
      onMoveTab == null ? void 0 : onMoveTab(fromIndex, toIndex);
      setTimeout(() => {
        collectionHook.dispatch({
          type: "reset"
        });
      }, 50);
    },
    [collectionHook, onMoveTab]
  );
  const handleTabSelectionChange = React.useCallback(
    (tabIndex) => {
      const selectedItem = collectionHook.data[tabIndex];
      const prevSelectedItem = collectionHook.data.find(
        (item) => item.priority === 1 && !item.isOverflowIndicator
      );
      if (selectedItem && prevSelectedItem && overflowMenuProp) {
        switchOverflowPriorities(selectedItem, prevSelectedItem);
      }
      onActiveChange == null ? void 0 : onActiveChange(tabIndex);
      setShowOverflowMenu(false);
    },
    [
      collectionHook.data,
      onActiveChange,
      overflowMenuProp,
      setShowOverflowMenu,
      switchOverflowPriorities
    ]
  );
  const {
    activeTabIndex,
    activateTab,
    addTab,
    onMouseDown: tabstripHookMouseDown,
    ...tabstripHook
  } = useTabstrip.useTabstrip({
    activeTabIndex: activeTabIndexProp,
    allowDragDrop,
    collectionHook,
    defaultTabs: tabDescriptors(defaultSource),
    defaultActiveTabIndex,
    editing,
    enableAddTab,
    idRoot: tabstripId,
    innerContainerRef,
    keyBoardActivation,
    onActiveChange: handleTabSelectionChange,
    onCloseTab,
    onEnterEditMode,
    onExitEditMode,
    onMoveTab: tabMovedHandler,
    orientation,
    promptForNewTabName,
    tabs: (_a = tabDescriptors(source)) != null ? _a : getChildren()
  });
  activeRef.current = activeTabIndex;
  React.useImperativeHandle(
    forwardedRef,
    () => ({
      focus: () => {
        const { current: tabstrip } = root;
        if (tabstrip) {
          const selectedTab = tabstrip.querySelector(
            '.saltTab[aria-selected="true"]'
          );
          if (selectedTab) {
            selectedTab.focus();
          }
        }
      }
    }),
    []
  );
  const handleAddTabClick = React.useCallback(() => {
    if (!collectionHook.isControlled) {
      addTab();
    }
    onAddTab == null ? void 0 : onAddTab();
  }, [collectionHook.isControlled, onAddTab, addTab]);
  const selectedIndex = React.useRef(activeTabIndex);
  const focusedTabIndex = tabstripHook.highlightedIdx;
  const handleOverflowSelectionChange = React.useCallback(
    (e, tab) => {
      if (tab !== null) {
        activateTab(tab.index);
      }
    },
    [activateTab]
  );
  const handleKeydownOverflowMenu = React.useCallback(
    (e) => {
      var _a2, _b;
      (_b = (_a2 = tabstripHook.navigationProps) == null ? void 0 : _a2.onKeyDown) == null ? void 0 : _b.call(_a2, e);
    },
    [tabstripHook.navigationProps]
  );
  const handleOverflowMenuOpen = React.useCallback(
    (open) => {
      setShowOverflowMenu(open);
    },
    [setShowOverflowMenu]
  );
  const handleMouseDown = React.useCallback(
    (evt) => {
      onMouseDown == null ? void 0 : onMouseDown(evt);
      tabstripHookMouseDown == null ? void 0 : tabstripHookMouseDown(evt);
    },
    [onMouseDown, tabstripHookMouseDown]
  );
  core.useIsomorphicLayoutEffect(() => {
    if (selectedIndex.current !== activeTabIndex && overflowMenuProp) {
      selectedIndex.current = activeTabIndex;
    }
  }, [overflowMenuProp, activeTabIndex]);
  core.useIsomorphicLayoutEffect(() => {
    if (React__default["default"].Children.count(children) !== childCount.current) {
      childCount.current = React__default["default"].Children.count(children);
    }
  }, [children]);
  core.useIsomorphicLayoutEffect(() => {
    if (activeTabIndex !== null && focusedTabIndex !== activeTabIndex && focusedTabIndex !== -1) {
      tabstripHook.focusTab(activeTabIndex);
    }
  }, [activeTabIndex]);
  const renderContent = () => {
    const content = collectionHook.data.filter((item) => !item.isOverflowIndicator && !item.isInjectedItem).map((item, index) => {
      const tab = item.source;
      const element = item.element;
      const selected = index === activeTabIndex;
      const focusVisible = tabstripHook.focusVisible === index;
      const overflowed = overflowedItems.findIndex(
        (item2) => item2.index === index
      ) !== -1;
      const tabIsBeingEdited = tabstripHook.editing && selected;
      const tabIndex = tabIsBeingEdited ? void 0 : selected && !tabstripHook.focusIsWithinComponent ? 0 : -1;
      const baseProps = {
        "data-index": index,
        "data-priority": item.priority,
        "data-overflowed": overflowed ? true : void 0,
        ...tabstripHook.navigationProps,
        id: item.id,
        key: index,
        onMouseDown: handleMouseDown,
        tabIndex
      };
      const tabProps = {
        ...tabstripHook.tabProps,
        closeable: item.closeable,
        dragging: tabstripHook.draggedItemIndex === index,
        editable: item.editable,
        editing: tabIsBeingEdited,
        focusVisible,
        index,
        onClose: item.closeable ? tabstripHook.closeTab : void 0,
        orientation,
        selected
      };
      if (React__default["default"].isValidElement(element)) {
        if (element.type === Tab.Tab) {
          return React__default["default"].cloneElement(element, { ...baseProps, ...tabProps });
        } else {
          return React__default["default"].cloneElement(element, baseProps);
        }
      } else {
        return React__default["default"].createElement(Tab.Tab, {
          ...baseProps,
          ...tabProps,
          label: tab.label
        });
      }
    });
    const overflowCount = overflowedItems.length;
    const draggingActiveTab = tabstripHook.draggedItemIndex === activeTabIndex;
    const showOverflow = tabstripHook.revealOverflowedItems && !draggingActiveTab || showOverflowMenu;
    const showTooltip = tabstripHook.revealOverflowedItems && draggingActiveTab;
    const overflowIndicator = collectionHook.data.find(
      (i) => i.isOverflowIndicator
    );
    const [injectedItem] = collectionHook.data.filter((i) => i.isInjectedItem);
    if (overflowIndicator) {
      content.push(
        /* @__PURE__ */ jsxRuntime.jsx(core.Tooltip, {
          content: "Active Tab cannot be moved into overflow list",
          open: true,
          disabled: !showTooltip,
          status: "warning",
          hideArrow: true,
          children: /* @__PURE__ */ jsxRuntime.jsx(Dropdown.Dropdown, {
            className: clsx.clsx(withBaseName("overflowMenu"), {
              [withBaseName("overflowMenu-open")]: showOverflow
            }),
            ListProps: {
              className: clsx.clsx({
                [withBaseName("overflowMenu-dropTarget")]: tabstripHook.revealOverflowedItems
              })
            },
            "data-overflow-indicator": true,
            "data-priority": 0,
            id: overflowIndicator.id,
            isOpen: showOverflow,
            onOpenChange: handleOverflowMenuOpen,
            onKeyDown: handleKeydownOverflowMenu,
            onSelectionChange: handleOverflowSelectionChange,
            placement: "bottom-end",
            source: overflowedItems,
            selected: null,
            triggerComponent: /* @__PURE__ */ jsxRuntime.jsx(core.Button, {
              "aria-label": `Tabs overflow menu ${overflowCount} item${overflowCount === 1 ? "" : "s"}`,
              variant: "secondary",
              tabIndex: -1,
              children: /* @__PURE__ */ jsxRuntime.jsx(icons.OverflowMenuIcon, {})
            }),
            width: "auto"
          }, "overflow")
        }, "tooltip")
      );
    }
    if (injectedItem) {
      content.push(
        /* @__PURE__ */ React.createElement(core.Button, {
          ...tabstripHook.navigationProps,
          "aria-label": ADD_TAB_LABEL,
          "data-priority": injectedItem.priority,
          "data-overflowed": injectedItem.overflowed,
          id: injectedItem.id,
          key: "addButton",
          onClick: handleAddTabClick,
          variant: "secondary",
          tabIndex: -1
        }, /* @__PURE__ */ jsxRuntime.jsx(icons.AddIcon, {}))
      );
    }
    return content;
  };
  const selectedTabOverflowed = overflowedItems.some(
    (item) => item.index === activeTabIndex
  );
  const className = clsx.clsx(
    withBaseName(),
    withBaseName(orientation),
    classNameProp,
    {
      [withBaseName("centered")]: centered,
      [withBaseName("draggingTab")]: tabstripHook.isDragging,
      [withBaseName("tertiary")]: variant === "tertiary"
    }
  );
  const selectedTabId = activeTabIndex !== null ? collectionHook.data[activeTabIndex].id : null;
  return /* @__PURE__ */ jsxRuntime.jsxs("div", {
    ...htmlAttributes,
    ...tabstripHook.containerProps,
    className,
    id: tabstripId,
    ref: root,
    role: "tablist",
    children: [
      /* @__PURE__ */ jsxRuntime.jsx("div", {
        className: withBaseName("inner"),
        ref: innerContainerRef,
        children: renderContent()
      }),
      showActivationIndicator ? /* @__PURE__ */ jsxRuntime.jsx(TabActivationIndicator.TabActivationIndicator, {
        hideThumb: selectedTabOverflowed || tabstripHook.isDragging,
        orientation,
        tabId: selectedTabId
      }) : null,
      tabstripHook.draggable
    ]
  });
});
Tabstrip.displayName = "Tabstrip";

exports.Tabstrip = Tabstrip;
