// @salt-ds/lab/dist-cjs/tabs/useActivationIndicator.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
require('@salt-ds/core');
require('react/jsx-runtime');
var useResizeObserver = require('../responsive/useResizeObserver.js');

const MEASUREMENTS = {
  horizontal: {
    pos: "left",
    size: "width"
  },
  vertical: {
    pos: "top",
    size: "height"
  }
};
function useActivationIndicator({
  rootRef,
  tabId,
  orientation
}) {
  const [style, setStyle] = React.useState({
    left: 0,
    width: 0
  });
  const styleRef = React.useRef(style);
  const getTabPos = React.useCallback(() => {
    const { pos, size } = MEASUREMENTS[orientation];
    return [pos, size];
  }, [orientation]);
  const createIndicatorStyle = React.useCallback(
    (tabElement) => {
      if (tabElement) {
        const tabRect = tabElement.getBoundingClientRect();
        if (rootRef.current && tabRect) {
          const rootRect = rootRef.current.getBoundingClientRect();
          const [pos, size] = getTabPos();
          const { [pos]: existingPos, [size]: existingSize } = styleRef.current;
          const newPos = tabRect[pos] - rootRect[pos];
          const newSize = tabRect[size];
          return {
            [pos]: newPos,
            [size]: newSize,
            hasChanged: newPos !== existingPos || newSize !== existingSize
          };
        }
      }
      return {};
    },
    [orientation, rootRef]
  );
  const onResize = React.useCallback(() => {
    requestAnimationFrame(() => {
      if (tabId) {
        const tabEl = document.getElementById(tabId);
        const { hasChanged, ...newStyle } = createIndicatorStyle(tabEl);
        if (hasChanged) {
          setStyle(newStyle);
        }
      }
    });
  }, [createIndicatorStyle, tabId]);
  React.useEffect(() => {
    if (tabId) {
      setTimeout(() => {
        const tabEl = document.getElementById(tabId);
        const { hasChanged, ...newStyle } = createIndicatorStyle(tabEl);
        if (hasChanged) {
          setStyle(styleRef.current = newStyle);
        }
      }, 50);
    }
  }, [createIndicatorStyle, tabId]);
  useResizeObserver.useResizeObserver(rootRef, useResizeObserver.WidthOnly, onResize);
  return style;
}

exports.useActivationIndicator = useActivationIndicator;
