// @salt-ds/lab/dist-cjs/tabs/useSelection.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@salt-ds/core');
var React = require('react');

const defaultSelectionKeys = ["Enter", " "];
const isTabElement = (el) => el && el.matches('[class*="saltTab "]');
const useSelection = ({
  defaultSelected,
  highlightedIdx,
  onSelectionChange,
  selected: selectedProp
}) => {
  const [selected, setSelected, isControlled] = core.useControlled({
    controlled: selectedProp,
    default: defaultSelected != null ? defaultSelected : 0,
    name: "Tabstrip",
    state: "value"
  });
  const isSelectionEvent = React.useCallback(
    (evt) => defaultSelectionKeys.includes(evt.key),
    []
  );
  const selectItem = React.useCallback(
    (tabIndex) => {
      setSelected(tabIndex);
      onSelectionChange == null ? void 0 : onSelectionChange(tabIndex);
    },
    [onSelectionChange, setSelected]
  );
  const handleKeyDown = React.useCallback(
    (e) => {
      const targetElement = e.target;
      if (isSelectionEvent(e) && highlightedIdx !== selected && isTabElement(targetElement)) {
        e.stopPropagation();
        e.preventDefault();
        selectItem(highlightedIdx);
      }
    },
    [isSelectionEvent, highlightedIdx, selected, selectItem]
  );
  const onClick = React.useCallback(
    (e, tabIndex) => {
      if (tabIndex !== selected) {
        selectItem(tabIndex);
      }
    },
    [selectItem, selected]
  );
  return {
    activateTab: selectItem,
    isControlled,
    onClick,
    onKeyDown: handleKeyDown,
    selected
  };
};

exports.isTabElement = isTabElement;
exports.useSelection = useSelection;
