// @salt-ds/lab/dist-cjs/tabs/useTabs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@salt-ds/core');
var React = require('react');
var useItemsWithIds = require('./useItemsWithIds.js');

const useTabs = ({
  activeTabIndex: activeTabIndexProp,
  children,
  defaultActiveTabIndex,
  id: idProp,
  onActiveChange
}) => {
  var _a, _b;
  const id = core.useIdMemo(idProp);
  const [itemsWithIds] = useItemsWithIds.useItemsWithIds(children, id);
  const [activeTabIndex, setActiveTabIndex] = core.useControlled({
    controlled: activeTabIndexProp,
    default: defaultActiveTabIndex != null ? defaultActiveTabIndex : 0,
    name: "useTabs",
    state: "activeTabIndex"
  });
  const handleTabActivated = React.useCallback(
    (tabIndex) => {
      setActiveTabIndex(tabIndex);
      onActiveChange == null ? void 0 : onActiveChange(tabIndex);
    },
    [onActiveChange, setActiveTabIndex]
  );
  return {
    id,
    onActiveChange: handleTabActivated,
    activeTabIndex,
    tabs: itemsWithIds,
    tabPanel: activeTabIndex === null ? null : (_b = (_a = itemsWithIds[activeTabIndex]) == null ? void 0 : _a.element) != null ? _b : null
  };
};

exports.useTabs = useTabs;
