// @salt-ds/lab/dist-cjs/content-status/internal/StatusIndicator.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var clsx = require('clsx');

const withBaseName = core.makePrefixer("saltContentStatus");
function StatusIndicator({
  status = "info",
  disableAnnouncer,
  value,
  title,
  message,
  CircularProgressProps: {
    className: circularProgressClassName,
    ...restCircularProgressProps
  } = {},
  SpinnerProps: { className: spinnerClassName, ...restSpinnerProps } = {},
  id
}) {
  if (status === "loading") {
    if (value !== void 0) {
      return /* @__PURE__ */ jsxRuntime.jsx(core.CircularProgress, {
        "aria-label": title || message,
        className: clsx.clsx(
          withBaseName("determinateLoading"),
          circularProgressClassName
        ),
        value,
        ...restCircularProgressProps
      });
    }
    return /* @__PURE__ */ jsxRuntime.jsx(core.Spinner, {
      className: clsx.clsx(withBaseName("indeterminateLoading"), spinnerClassName),
      completionAnnouncement: null,
      "data-testid": `spinner-${id}`,
      disableAnnouncer,
      ...restSpinnerProps
    });
  }
  return /* @__PURE__ */ jsxRuntime.jsx(core.StatusIndicator, {
    status,
    "data-jpmui-test": `icon-${status}-${id}`,
    size: 2
  });
}

exports.StatusIndicator = StatusIndicator;
