// @salt-ds/lab/dist-es/combo-box/ComboBox.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { useIdMemo } from '@salt-ds/core';
import { forwardRef, useCallback } from 'react';
import { DropdownBase } from '../dropdown/DropdownBase.js';
import '../dropdown/DropdownButton.js';
import '../dropdown/Dropdown.js';
import '../responsive/useResizeObserver.js';
import '../form-field-context-legacy/FormFieldLegacyContext.js';
import '../list/Highlighter.js';
import '../list/ListItem.js';
import { List } from '../list/List.js';
import { CollectionProvider } from '../common-hooks/collectionProvider.js';
import '../common-hooks/keyUtils.js';
import { useCollectionItems } from '../common-hooks/useCollectionItems.js';
import '../list/VirtualizedList.js';
import { InputLegacy } from '../input-legacy/InputLegacy.js';
import '../input-legacy/StaticInputAdornment.js';
import { useCombobox } from './useCombobox.js';

const ComboBox = forwardRef(function Combobox({
  InputProps: InputProps2,
  ListProps: ListProps2,
  ListItem,
  "aria-label": ariaLabel,
  allowFreeText,
  children,
  defaultIsOpen,
  defaultValue,
  disabled,
  onBlur,
  onFocus,
  onChange,
  onSelect,
  getFilterRegex,
  id: idProp,
  isOpen: isOpenProp,
  itemToString,
  onOpenChange: onOpenChangeProp,
  onSelectionChange,
  selectionStrategy,
  source,
  stringToItem,
  value: valueProp,
  width = 180,
  ...props
}, forwardedRef) {
  const id = useIdMemo(idProp);
  const collectionHook = useCollectionItems({
    id,
    source,
    children,
    options: {
      filterPattern: valueProp != null ? valueProp : defaultValue,
      getFilterRegex,
      itemToString
    }
  });
  const {
    focusVisible,
    highlightedIndex,
    inputProps,
    isOpen,
    listHandlers,
    listControlProps: controlProps,
    onOpenChange,
    selected
  } = useCombobox({
    InputProps: InputProps2,
    allowFreeText,
    ariaLabel,
    collectionHook,
    defaultIsOpen,
    defaultValue,
    disabled,
    onBlur,
    onFocus,
    onChange,
    onSelect,
    id,
    isOpen: isOpenProp,
    itemToString,
    label: "ComboBox",
    onOpenChange: onOpenChangeProp,
    onSelectionChange,
    selectionStrategy,
    stringToItem,
    value: valueProp
  });
  const collectionItemsToItem = useCallback(
    (sel) => {
      if (Array.isArray(sel)) {
        return sel.map((i) => i.value);
      } else if (sel) {
        return sel.value;
      } else {
        return sel;
      }
    },
    []
  );
  return /* @__PURE__ */ jsx(CollectionProvider, {
    collectionHook,
    children: /* @__PURE__ */ jsxs(DropdownBase, {
      ...props,
      fullWidth: true,
      id,
      isOpen,
      onOpenChange,
      openOnFocus: true,
      ref: forwardedRef,
      width,
      children: [
        /* @__PURE__ */ jsx(InputLegacy, {
          ...inputProps,
          disabled,
          ...controlProps
        }),
        /* @__PURE__ */ jsx(List, {
          ...ListProps2,
          ListItem,
          focusVisible,
          highlightedIndex,
          itemTextHighlightPattern: inputProps.value || void 0,
          id: `${id}-list`,
          listHandlers,
          onSelectionChange,
          selected: collectionItemsToItem(selected),
          selectionStrategy
        })
      ]
    })
  });
});

export { ComboBox };
