// @salt-ds/lab/dist-es/formatted-input/internal/InputWithMask.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { useId } from '@salt-ds/core';
import { clsx } from 'clsx';
import { forwardRef } from 'react';

const baseName = "saltFormattedInput";
function getClippedMask(mask, value, textAlign) {
  if (mask) {
    if (textAlign === "right") {
      return mask.substring(
        0,
        value ? mask.length - value.length : mask.length
      );
    }
    return mask.substring(value ? value.length : 0);
  }
  return "";
}
const InputWithMask = forwardRef(
  function InputWithMask2({
    "aria-labelledby": ariaLabelledBy,
    className,
    mask = "",
    value,
    id: idProp,
    style,
    textAlign,
    ...other
  }, ref) {
    const id = useId(idProp);
    const spacerAndMask = [
      /* @__PURE__ */ jsx("i", {
        children: value
      }, value),
      getClippedMask(mask, value, textAlign)
    ];
    return /* @__PURE__ */ jsxs("div", {
      className: `${baseName}-inputWrapper`,
      children: [
        /* @__PURE__ */ jsx("input", {
          className,
          ...other,
          "aria-label": mask,
          "aria-labelledby": [ariaLabelledBy, id].filter((label) => label != null).join(" "),
          id,
          ref,
          style,
          value
        }),
        /* @__PURE__ */ jsx("span", {
          "aria-hidden": "true",
          className: clsx(`${baseName}-mask`, className),
          style,
          children: textAlign !== "right" ? spacerAndMask : spacerAndMask.reverse()
        })
      ]
    });
  }
);

export { InputWithMask, baseName };
