// @salt-ds/lab/dist-es/content-status/internal/StatusIndicator.js
import { jsx } from 'react/jsx-runtime';
import { makePrefixer, CircularProgress, Spinner, StatusIndicator as StatusIndicator$1 } from '@salt-ds/core';
import { clsx } from 'clsx';

const withBaseName = makePrefixer("saltContentStatus");
function StatusIndicator({
  status = "info",
  disableAnnouncer,
  value,
  title,
  message,
  CircularProgressProps: {
    className: circularProgressClassName,
    ...restCircularProgressProps
  } = {},
  SpinnerProps: { className: spinnerClassName, ...restSpinnerProps } = {},
  id
}) {
  if (status === "loading") {
    if (value !== void 0) {
      return /* @__PURE__ */ jsx(CircularProgress, {
        "aria-label": title || message,
        className: clsx(
          withBaseName("determinateLoading"),
          circularProgressClassName
        ),
        value,
        ...restCircularProgressProps
      });
    }
    return /* @__PURE__ */ jsx(Spinner, {
      className: clsx(withBaseName("indeterminateLoading"), spinnerClassName),
      completionAnnouncement: null,
      "data-testid": `spinner-${id}`,
      disableAnnouncer,
      ...restSpinnerProps
    });
  }
  return /* @__PURE__ */ jsx(StatusIndicator$1, {
    status,
    "data-jpmui-test": `icon-${status}-${id}`,
    size: 2
  });
}

export { StatusIndicator };
