// @salt-ds/lab/dist-es/overlay/OverlayPanelBase.js
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import { FloatingArrow } from '@floating-ui/react';
import css_248z from './Overlay.css.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import { makePrefixer, Button } from '@salt-ds/core';
import { CloseIcon } from '@salt-ds/icons';
import { useOverlayContext } from './OverlayContext.js';
import clsx from 'clsx';

const withBaseName = makePrefixer("saltOverlayPanelBase");
const OverlayPanelBase = (props) => {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-overlay",
    css: css_248z,
    window: targetWindow
  });
  const { arrowProps, handleCloseButton } = useOverlayContext();
  const { children, className, ...rest } = props;
  return /* @__PURE__ */ jsxs(Fragment, {
    children: [
      /* @__PURE__ */ jsxs("div", {
        className: clsx(withBaseName("container"), className),
        ...rest,
        children: [
          /* @__PURE__ */ jsx(Button, {
            onClick: handleCloseButton,
            variant: "secondary",
            className: withBaseName("closeButton"),
            "aria-label": "Close Overlay",
            children: /* @__PURE__ */ jsx(CloseIcon, {
              "aria-hidden": true
            })
          }),
          /* @__PURE__ */ jsx("div", {
            className: withBaseName("content"),
            children
          })
        ]
      }),
      /* @__PURE__ */ jsx(FloatingArrow, {
        ...arrowProps,
        strokeWidth: 1,
        fill: "var(--overlay-background)",
        stroke: "var(--overlay-borderColor)",
        height: 5,
        width: 10
      })
    ]
  });
};

export { OverlayPanelBase };
