// @salt-ds/lab/dist-es/overlay/OverlayPanel.js
import { jsx } from 'react/jsx-runtime';
import { forwardRef } from 'react';
import { makePrefixer, useFloatingComponent, useForkRef } from '@salt-ds/core';
import { clsx } from 'clsx';
import { useOverlayContext } from './OverlayContext.js';
import { OverlayPanelBase } from './OverlayPanelBase.js';

const withBaseName = makePrefixer("saltOverlayPanel");
const OverlayPanel = forwardRef(
  function OverlayPanel2(props, ref) {
    const { className, ["aria-labelledby"]: ariaLabelledby, ...rest } = props;
    const { Component: FloatingComponent } = useFloatingComponent();
    const {
      id,
      openState,
      floatingStyles,
      placement,
      context,
      getFloatingProps,
      floating
    } = useOverlayContext();
    const handleRef = useForkRef(floating, ref);
    const { top, left, width, height, position } = floatingStyles;
    const getOverlayProps = () => {
      return getFloatingProps({
        "data-placement": placement,
        ref: floating,
        id: `${id}-panel`
      });
    };
    return /* @__PURE__ */ jsx(FloatingComponent, {
      open: openState,
      className: clsx(withBaseName(), className),
      "aria-modal": "true",
      ...getOverlayProps(),
      ref: handleRef,
      width,
      height,
      top,
      left,
      position,
      focusManagerProps: {
        context
      },
      "aria-labelledby": ariaLabelledby,
      children: /* @__PURE__ */ jsx(OverlayPanelBase, {
        ...rest
      })
    });
  }
);

export { OverlayPanel };
