// @salt-ds/lab/dist-es/tokenized-input-next/TokenizedInputNext.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { makePrefixer, useId, useForkRef, StatusAdornment, Button } from '@salt-ds/core';
import { forwardRef, useRef } from 'react';
import { useTokenizedInputNext } from './useTokenizedInputNext.js';
import { clsx } from 'clsx';
import { InputPill } from './internal/InputPill.js';
import { CloseIcon, OverflowMenuIcon } from '@salt-ds/icons';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './TokenizedInputNext.css.js';

const withBaseName = makePrefixer("saltTokenizedInputNext");
const getItemsAriaLabel = (itemCount) => itemCount === 0 ? "no item selected" : `${itemCount} ${itemCount > 1 ? "items" : "item"}`;
const TokenizedInputNext = forwardRef(function TokenizedInputNext2({
  textAreaRef: textAreaRefProp,
  textAreaProps = {},
  variant = "primary",
  ...rest
}, ref) {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-tokenized-input-next",
    css: css_248z,
    window: targetWindow
  });
  const {
    "aria-describedby": textAreaDescribedBy,
    "aria-labelledby": textAreaLabelledBy,
    required: textAreaRequired,
    ...restTextAreaProps
  } = textAreaProps;
  const { refs, helpers, inputProps, firstHiddenIndex } = useTokenizedInputNext(rest);
  const {
    textAreaRef: textAreaHookRef,
    pillsRef,
    clearButtonRef,
    expandButtonRef,
    statusAdornmentRef,
    containerRef: containerHookRef
  } = refs;
  const {
    ExpandButtonProps = {
      "aria-label": "expand edit"
    },
    className,
    activeIndices = [],
    selectedItems = [],
    highlightedIndex,
    value,
    expanded,
    disabled,
    onBlur,
    onKeyDown,
    onRemoveItem,
    onInputChange,
    focused,
    validationStatus,
    readOnly,
    onInputFocus,
    onInputBlur,
    onClear,
    onClick,
    onKeyUp,
    id: idProp,
    "aria-label": ariaLabel,
    "aria-labelledby": ariaLabelledBy,
    "aria-describedby": ariaDescribedBy,
    ...restProps
  } = inputProps;
  const id = useId(idProp);
  const inputId = `${id}-input`;
  const expandButtonId = `${id}-expand-button`;
  const clearButtonId = `${id}-clear-button`;
  const keydownExpandButton = useRef(false);
  const containerRef = useForkRef(ref, containerHookRef);
  const textAreaRef = useForkRef(textAreaHookRef, textAreaRefProp);
  const showExpandButton = !expanded && firstHiddenIndex != null;
  const hasHelpers = (helpers2) => {
    if (process.env.NODE_ENV !== "production") {
      if (helpers2 == null) {
        console.warn(
          'TokenizedInputNext is used without helpers. You should pass in "helpers" from "useTokenizedInputNext".'
        );
      }
    }
    return helpers2 != null;
  };
  const handleExpandButtonKeyDown = (event) => {
    const singleChar = event.key.length === 1;
    const triggerExpand = [
      "CONTROL",
      "META",
      "ENTER",
      "BACKSPACE",
      "ARROWDOWN",
      "ARROWLEFT",
      "ARROWRIGHT"
    ].indexOf(event.key.toUpperCase()) !== -1;
    if ((singleChar || triggerExpand) && hasHelpers(helpers)) {
      if (event.key === "Enter" || event.key === " ") {
        event.preventDefault();
        event.stopPropagation();
      }
      helpers.updateExpanded(event, true);
      keydownExpandButton.current = true;
    }
  };
  const handleInputKeyUp = (event) => {
    if (keydownExpandButton.current && "Enter" !== event.key) {
      keydownExpandButton.current = false;
      onKeyDown == null ? void 0 : onKeyDown(event);
    }
    onKeyUp == null ? void 0 : onKeyUp(event);
  };
  const handleExpand = (event) => {
    event.stopPropagation();
    if (hasHelpers(helpers)) {
      helpers.updateExpanded(event, true);
    }
  };
  const handleClearButtonFocus = (event) => {
    event.stopPropagation();
  };
  const selectedItemIds = selectedItems.map(
    (_, index) => `${id}-pill-${index}`
  );
  const inputAriaLabelledBy = disabled ? [ariaLabelledBy, inputId, ...selectedItemIds] : [ariaLabelledBy, inputId];
  const expandedWithItems = expanded && !showExpandButton && selectedItems.length > 0;
  const { "aria-label": expandButtonAccessibleText, ...restExpandButtonProps } = ExpandButtonProps;
  return /* @__PURE__ */ jsxs("div", {
    className: withBaseName("container"),
    children: [
      /* @__PURE__ */ jsx("span", {
        "aria-owns": selectedItemIds.join(" "),
        className: withBaseName("hidden"),
        role: "listbox"
      }),
      /* @__PURE__ */ jsxs("div", {
        className: clsx(
          withBaseName(),
          withBaseName(variant),
          {
            [withBaseName("expanded")]: expanded,
            [withBaseName("focused")]: !disabled && focused,
            [withBaseName("disabled")]: disabled,
            [withBaseName("readOnly")]: readOnly,
            [withBaseName(validationStatus)]: validationStatus
          },
          className
        ),
        ref: containerRef,
        onClick,
        tabIndex: -1,
        ...restProps,
        children: [
          selectedItems.length > 0 && selectedItems.map((item, index) => {
            const label = String(item);
            return /* @__PURE__ */ jsx(InputPill, {
              disabled,
              hidden: showExpandButton && index >= firstHiddenIndex,
              highlighted: index === highlightedIndex || activeIndices.indexOf(index) !== -1,
              id: `${id}-pill-${index}`,
              index,
              label,
              onClose: expanded && !readOnly ? (event) => onRemoveItem == null ? void 0 : onRemoveItem(event, index) : void 0,
              pillsRef
            }, `${index}-${label}`);
          }),
          /* @__PURE__ */ jsx("textarea", {
            "aria-labelledby": clsx(inputAriaLabelledBy, textAreaLabelledBy),
            "aria-describedby": clsx(ariaDescribedBy, textAreaDescribedBy),
            "aria-label": clsx(ariaLabel, getItemsAriaLabel(selectedItems.length)),
            "aria-activedescendant": highlightedIndex && highlightedIndex >= 0 ? `${id}-pill-${highlightedIndex}` : void 0,
            disabled,
            id: inputId,
            readOnly,
            ref: textAreaRef,
            required: textAreaRequired,
            rows: 1,
            tabIndex: disabled ? -1 : 0,
            value,
            className: clsx(withBaseName("textarea"), textAreaProps == null ? void 0 : textAreaProps.className),
            onChange: onInputChange,
            onBlur: onInputBlur,
            onFocus: !disabled ? onInputFocus : void 0,
            onKeyDown,
            ...restTextAreaProps
          }),
          /* @__PURE__ */ jsxs("div", {
            className: withBaseName("endAdornmentContainer"),
            children: [
              !disabled && !readOnly && validationStatus && /* @__PURE__ */ jsx(StatusAdornment, {
                className: withBaseName("statusAdornment"),
                ref: statusAdornmentRef,
                status: validationStatus
              }),
              expandedWithItems && !readOnly && /* @__PURE__ */ jsx(Button, {
                className: withBaseName("endAdornment"),
                disabled,
                id: clearButtonId,
                onBlur,
                onClick: onClear,
                onFocus: handleClearButtonFocus,
                ref: clearButtonRef,
                variant: "secondary",
                "data-testid": "clear-button",
                "aria-label": "clear input",
                children: /* @__PURE__ */ jsx(CloseIcon, {
                  "aria-hidden": true
                })
              }),
              showExpandButton && /* @__PURE__ */ jsx(Button, {
                className: withBaseName("endAdornment"),
                "aria-label": expandButtonAccessibleText,
                "aria-labelledby": clsx(ariaLabelledBy, inputId, expandButtonId),
                disabled,
                id: expandButtonId,
                onBlur,
                onClick: handleExpand,
                onKeyDown: handleExpandButtonKeyDown,
                onKeyUp: handleInputKeyUp,
                ref: expandButtonRef,
                variant: "secondary",
                "data-testid": "expand-button",
                ...restExpandButtonProps,
                children: /* @__PURE__ */ jsx(OverflowMenuIcon, {})
              })
            ]
          }),
          /* @__PURE__ */ jsx("div", {
            className: withBaseName("activationIndicator")
          })
        ]
      })
    ]
  });
});

export { TokenizedInputNext };
