// @salt-ds/lab/dist-es/color-chooser/ColorChooser.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { useState } from 'react';
import { clsx } from 'clsx';
import { Overlay } from '../overlay/Overlay.js';
import { OverlayTrigger } from '../overlay/OverlayTrigger.js';
import { OverlayPanel } from '../overlay/OverlayPanel.js';
import '../overlay/OverlayPanelCloseButton.js';
import { OverlayPanelContent } from '../overlay/OverlayPanelContent.js';
import { makePrefixer, Button } from '@salt-ds/core';
import { RefreshIcon } from '@salt-ds/icons';
import { isTransparent } from './color-utils.js';
import { getHexValue, getColorNameByHexValue, hexValueWithoutAlpha, convertColorMapValueToHex } from './ColorHelpers.js';
import { saltColorMap } from './colorMap.js';
import { DictTabs } from './DictTabs.js';
import { getColorPalettes } from './GetColorPalettes.js';
import { createTabsMapping } from './createTabsMapping.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './ColorChooser.css.js';

const withBaseName = makePrefixer("saltColorChooser");
function getActiveTab(hexValue, tabs, saltColorOverrides) {
  if (tabs.Swatches && tabs["Color Picker"]) {
    const hexNoAlpha = hexValueWithoutAlpha(hexValue);
    const colors = saltColorOverrides != null ? saltColorOverrides : saltColorMap;
    if (hexNoAlpha && !Object.keys(colors).find(
      (key) => {
        var _a;
        return ((_a = convertColorMapValueToHex(colors[key])) == null ? void 0 : _a.toLowerCase()) === (hexNoAlpha == null ? void 0 : hexNoAlpha.toLowerCase());
      }
    )) {
      return 1;
    }
  }
  return 0;
}
const ColorChooser = ({
  onClear,
  onSelect,
  color,
  showSwatches = true,
  showColorPicker = true,
  defaultAlpha = 1,
  disableAlphaChooser = false,
  hideLabel = false,
  placeholder,
  buttonProps,
  saltColorOverrides,
  readOnly = false,
  displayHexOnly = false
}) => {
  var _a, _b, _c;
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-color-chooser",
    css: css_248z,
    window: targetWindow
  });
  const [open, setOpen] = useState(false);
  const allColors = saltColorOverrides ? getColorPalettes(saltColorOverrides) : getColorPalettes();
  const displayColorName = displayHexOnly ? getHexValue(color == null ? void 0 : color.hex, disableAlphaChooser) : getColorNameByHexValue(
    color == null ? void 0 : color.hex,
    disableAlphaChooser,
    saltColorOverrides
  );
  const handleOpenChange = (open2) => setOpen(open2);
  const alphaForTabs = isTransparent(color == null ? void 0 : color.hex) ? defaultAlpha : (_b = (_a = color == null ? void 0 : color.rgba) == null ? void 0 : _a.a) != null ? _b : defaultAlpha;
  const tabsMapping = createTabsMapping({
    swatches: showSwatches,
    colorPicker: showColorPicker,
    disableAlphaChooser,
    allColors,
    color,
    alpha: alphaForTabs,
    handleColorChange: onSelect,
    displayColorName,
    placeholder,
    onDialogClosed: () => {
      setOpen(false);
    }
  });
  const [activeTab, setActiveTab] = useState(
    getActiveTab(color == null ? void 0 : color.hex, tabsMapping, saltColorOverrides)
  );
  const onDefaultSelected = () => {
    if (activeTab === 0 && showSwatches) {
      onClear();
      handleOpenChange(false);
    } else {
      onClear();
    }
  };
  const onTabClick = (index) => {
    setActiveTab(index);
  };
  const OverlayData = () => {
    return /* @__PURE__ */ jsxs("div", {
      className: clsx(withBaseName("overlayContent")),
      "data-testid": "overlay-content",
      children: [
        /* @__PURE__ */ jsxs(Button, {
          "data-testid": "default-button",
          variant: "secondary",
          className: clsx(withBaseName("defaultButton")),
          onClick: onDefaultSelected,
          children: [
            /* @__PURE__ */ jsx(RefreshIcon, {
              className: clsx(withBaseName("refreshIcon"))
            }),
            "Default"
          ]
        }),
        /* @__PURE__ */ jsx(DictTabs, {
          tabs: tabsMapping,
          hexValue: color == null ? void 0 : color.hex,
          onTabClick,
          activeTab
        })
      ]
    });
  };
  return /* @__PURE__ */ jsxs(Overlay, {
    placement: "bottom",
    "data-testid": "color-chooser-overlay",
    children: [
      /* @__PURE__ */ jsx(OverlayTrigger, {
        children: /* @__PURE__ */ jsxs(Button, {
          className: clsx(withBaseName("overlayButton"), {
            [withBaseName("overlayButtonHiddenLabel")]: hideLabel
          }),
          "data-testid": "color-chooser-overlay-button",
          disabled: readOnly,
          ...buttonProps,
          children: [
            color && /* @__PURE__ */ jsx("div", {
              className: clsx(withBaseName("overlayButtonSwatch"), {
                [withBaseName("overlayButtonSwatchWithBorder")]: color == null ? void 0 : color.hex.startsWith("#ffffff"),
                [withBaseName("overlayButtonSwatchTransparent")]: isTransparent(
                  color == null ? void 0 : color.hex
                )
              }),
              style: {
                backgroundColor: color == null ? void 0 : color.hex
              }
            }),
            !hideLabel && /* @__PURE__ */ jsx("div", {
              className: withBaseName("overlayButtonText"),
              children: (_c = displayColorName != null ? displayColorName : placeholder) != null ? _c : "No color selected"
            })
          ]
        })
      }),
      /* @__PURE__ */ jsx(OverlayPanel, {
        children: /* @__PURE__ */ jsx(OverlayPanelContent, {
          children: /* @__PURE__ */ jsx(OverlayData, {})
        })
      })
    ]
  });
};

export { ColorChooser };
