// @salt-ds/lab/dist-es/color-chooser/DictTabs.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { clsx } from 'clsx';
import { makePrefixer } from '@salt-ds/core';
import '../tabs/TabActivationIndicator.js';
import { Tabstrip } from '../tabs/Tabstrip.js';
import { Tab } from '../tabs/Tab.js';
import '../tabs/TabPanel.js';
import '../tabs/Tabs.js';

const withBaseName = makePrefixer("saltColorChooserDictTabs");
const DictTabs = ({
  tabs,
  hexValue,
  onTabClick,
  activeTab,
  ...props
}) => {
  return /* @__PURE__ */ jsxs("div", {
    children: [
      /* @__PURE__ */ jsx(Tabstrip, {
        ...props,
        "data-testid": "color-chooser-tabstrip",
        className: clsx(withBaseName("wrapper")),
        activeTabIndex: activeTab,
        onActiveChange: (tabIndex) => onTabClick(tabIndex),
        children: [...Object.keys(tabs)].map((label, i) => /* @__PURE__ */ jsx(Tab, {
          className: clsx(withBaseName("text")),
          label
        }, i))
      }),
      [...Object.values(tabs)].map((tab, idx) => {
        if (!tab) {
          return null;
        }
        const TabComponent = tab == null ? void 0 : tab.Component;
        return /* @__PURE__ */ jsx("div", {
          "aria-hidden": activeTab !== idx,
          hidden: activeTab !== idx,
          children: (tab == null ? void 0 : tab.props) ? /* @__PURE__ */ jsx(TabComponent, {
            ...tab == null ? void 0 : tab.props
          }) : null
        }, idx);
      })
    ]
  });
};

export { DictTabs };
