// @salt-ds/lab/dist-es/color-chooser/RGBAInput.js
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import { clsx } from 'clsx';
import { makePrefixer } from '@salt-ds/core';
import { AlphaInput } from './AlphaInputField.js';
import { RGBInput } from './RGBAInputField.js';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import css_248z from './RGBAInput.css.js';

const withBaseName = makePrefixer("saltColorChooser");
const RGBAInput = ({
  disableAlphaChooser,
  rgbaText,
  rgbaValue,
  onSubmitAlpha,
  onSubmitRgb
}) => {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-rgba-input",
    css: css_248z,
    window: targetWindow
  });
  return /* @__PURE__ */ jsxs(Fragment, {
    children: [
      /* @__PURE__ */ jsx("span", {
        className: clsx(withBaseName("textDivOverrides")),
        children: rgbaText
      }),
      ["r", "g", "b"].map((value) => /* @__PURE__ */ jsxs("div", {
        className: withBaseName("rgbaInputDiv"),
        children: [
          /* @__PURE__ */ jsx("span", {
            className: withBaseName("rgbaText"),
            children: value.toUpperCase()
          }),
          /* @__PURE__ */ jsx(RGBInput, {
            rgbaValue,
            value,
            onSubmit: onSubmitRgb
          })
        ]
      })),
      !disableAlphaChooser ? /* @__PURE__ */ jsxs("div", {
        className: clsx(withBaseName("rgbaInputDiv")),
        children: [
          /* @__PURE__ */ jsx("span", {
            className: clsx(withBaseName("rgbaText")),
            children: "A"
          }),
          /* @__PURE__ */ jsx(AlphaInput, {
            alphaValue: parseFloat(rgbaValue.a.toFixed(2)),
            onSubmit: onSubmitAlpha
          })
        ]
      }) : /* @__PURE__ */ jsx("div", {
        className: clsx(withBaseName("alphaSpacerDiv"))
      })
    ]
  });
};

export { RGBAInput };
