// @salt-ds/lab/dist-es/list/Highlighter.js
import { jsx, Fragment } from 'react/jsx-runtime';
import { makePrefixer } from '@salt-ds/core';
import { escapeRegExp } from '../utils/escapeRegExp.js';
import 'react';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './Highlighter.css.js';

const withBaseName = makePrefixer("saltHighlighter");
const Highlighter = (props) => {
  const { matchPattern, text = "" } = props;
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-hightligher",
    css: css_248z,
    window: targetWindow
  });
  const matchRegex = typeof matchPattern === "string" ? new RegExp(`(${escapeRegExp(matchPattern)})`, "gi") : matchPattern;
  if (matchRegex === void 0 || matchPattern === "") {
    return /* @__PURE__ */ jsx(Fragment, {
      children: text
    });
  }
  return /* @__PURE__ */ jsx("span", {
    children: text.split(matchRegex).map(
      (part, index) => part.match(matchRegex) ? /* @__PURE__ */ jsx("strong", {
        className: withBaseName("highlight"),
        children: part
      }, `${index}-${part}`) : part
    )
  });
};

export { Highlighter };
