// @salt-ds/lab/dist-es/stepper-input/StepperInput.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { clsx } from 'clsx';
import { forwardRef, useRef } from 'react';
import { makePrefixer, Button } from '@salt-ds/core';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import { RefreshIcon, TriangleUpIcon, TriangleDownIcon } from '@salt-ds/icons';
import { InputLegacy } from '../input-legacy/InputLegacy.js';
import '../input-legacy/StaticInputAdornment.js';
import { useActivationIndicatorPosition } from './internal/useActivationIndicatorPosition.js';
import { useStepperInput } from './useStepperInput.js';
import css_248z from './StepperInput.css.js';

const withBaseName = makePrefixer("saltStepperInput");
const StepperInput = forwardRef(
  function StepperInput2(props, ref) {
    const {
      ButtonProps: ButtonPropsProp,
      InputProps: InputPropsProp,
      textAlign = "left",
      className,
      showRefreshButton = false
    } = props;
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-stepper-input",
      css: css_248z,
      window: targetWindow
    });
    const adornmentRef = useRef(null);
    const inputRef = useRef(null);
    const {
      decrementButtonDown,
      getButtonIcon,
      getButtonProps,
      getInputProps,
      incrementButtonDown,
      isAtMax,
      isAtMin,
      refreshCurrentValue,
      stepperDirection,
      valuesHaveDiverged
    } = useStepperInput(props, inputRef);
    useActivationIndicatorPosition(
      adornmentRef,
      valuesHaveDiverged() || showRefreshButton
    );
    const endAdornment = /* @__PURE__ */ jsxs("div", {
      className: withBaseName("adornmentContainer"),
      ref: adornmentRef,
      children: [
        /* @__PURE__ */ jsx(Button, {
          "aria-label": "Refresh default value",
          className: clsx(withBaseName("secondaryButton"), {
            [withBaseName("hideSecondaryButton")]: !(showRefreshButton || valuesHaveDiverged())
          }),
          onClick: refreshCurrentValue,
          variant: "secondary",
          children: /* @__PURE__ */ jsx(RefreshIcon, {
            "aria-label": "refresh"
          })
        }),
        /* @__PURE__ */ jsxs("div", {
          className: withBaseName("buttonContainer"),
          children: [
            /* @__PURE__ */ jsx(Button, {
              className: clsx(
                withBaseName("stepperButton"),
                withBaseName("increment"),
                {
                  active: incrementButtonDown
                }
              ),
              disabled: isAtMax(),
              ...getButtonProps(stepperDirection.INCREMENT, ButtonPropsProp),
              children: /* @__PURE__ */ jsx(TriangleUpIcon, {
                className: withBaseName("stepperButtonIcon"),
                "aria-label": getButtonIcon(stepperDirection.INCREMENT)
              })
            }),
            /* @__PURE__ */ jsx(Button, {
              className: clsx(
                withBaseName("stepperButton"),
                withBaseName("decrement"),
                {
                  active: decrementButtonDown
                }
              ),
              disabled: isAtMin(),
              ...getButtonProps(stepperDirection.DECREMENT, ButtonPropsProp),
              children: /* @__PURE__ */ jsx(TriangleDownIcon, {
                className: withBaseName("stepperButtonIcon"),
                "aria-label": getButtonIcon(stepperDirection.DECREMENT)
              })
            })
          ]
        })
      ]
    });
    return /* @__PURE__ */ jsx("div", {
      className: clsx(withBaseName(), className),
      onBlur: props.onBlur,
      onFocus: props.onFocus,
      ref,
      children: /* @__PURE__ */ jsx(InputLegacy, {
        className: withBaseName("input"),
        endAdornment,
        highlightOnFocus: true,
        ref: inputRef,
        textAlign,
        ...getInputProps(InputPropsProp)
      })
    });
  }
);

export { StepperInput };
