// @salt-ds/lab/dist-es/dropdown/DropdownBase.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { flip, shift, limitShift, size } from '@floating-ui/react';
import { makePrefixer, useIdMemo, useFloatingUI, useForkRef } from '@salt-ds/core';
import { clsx } from 'clsx';
import { forwardRef, useRef, Children, useState, cloneElement } from 'react';
import { forwardCallbackProps } from '../utils/forwardCallbackProps.js';
import { useDropdownBase } from './useDropdownBase.js';
import { Portal } from '../portal/Portal.js';
import { useWindow as useWindow$1, isDesktop } from '../window/WindowContext.js';
import '../window/ElectronWindow.js';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import css_248z from './Dropdown.css.js';

const withBaseName = makePrefixer("saltDropdown");
const DropdownBase = forwardRef(
  function Dropdown({
    "aria-labelledby": ariaLabelledByProp,
    children,
    className: classNameProp,
    container,
    defaultIsOpen,
    disabled,
    disablePortal,
    fullWidth,
    id: idProp,
    isOpen: isOpenProp,
    onKeyDown,
    onOpenChange,
    openOnFocus,
    placement = "bottom-start",
    popupWidth,
    width,
    ...htmlAttributes
  }, forwardedRef) {
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-dropdown",
      css: css_248z,
      window: targetWindow
    });
    const rootRef = useRef(null);
    const className = clsx(withBaseName(), classNameProp, {
      [withBaseName("fullWidth")]: fullWidth,
      [withBaseName("disabled")]: disabled
    });
    const [trigger, popupComponent] = Children.toArray(
      children
    );
    const id = useIdMemo(idProp);
    const Window = useWindow$1();
    const { componentProps, popperRef, isOpen, triggerProps } = useDropdownBase(
      {
        ariaLabelledBy: ariaLabelledByProp,
        defaultIsOpen,
        disabled,
        fullWidth,
        id,
        isOpen: isOpenProp,
        onOpenChange,
        onKeyDown,
        openOnFocus,
        popupComponent,
        popupWidth,
        rootRef,
        width
      }
    );
    const [maxPopupHeight, setMaxPopupHeight] = useState(
      void 0
    );
    const middleware = isDesktop ? [] : [
      flip({
        fallbackPlacements: ["bottom-start", "top-start"]
      }),
      shift({ limiter: limitShift() }),
      size({
        apply({ availableHeight }) {
          setMaxPopupHeight(availableHeight);
        }
      })
    ];
    const { reference, floating, x, y, strategy } = useFloatingUI({
      placement,
      middleware
    });
    const handlePopperListAdapterRef = useForkRef(
      reference,
      forwardedRef
    );
    const handleRootRef = useForkRef(rootRef, handlePopperListAdapterRef);
    const handleFloatingRef = useForkRef(floating, popperRef);
    const getTriggerComponent = () => {
      const {
        id: defaultId,
        role: defaultRole,
        ...restTriggerProps
      } = triggerProps;
      const {
        id: id2 = defaultId,
        role = defaultRole,
        ...ownProps
      } = trigger.props;
      return cloneElement(
        trigger,
        forwardCallbackProps(ownProps, {
          ...restTriggerProps,
          id: id2,
          role
        })
      );
    };
    const getPopupComponent = () => {
      const { id: defaultId, width: width2, ...restComponentProps } = componentProps;
      const {
        className: className2,
        id: id2 = defaultId,
        width: ownWidth,
        ...ownProps
      } = popupComponent.props;
      return cloneElement(popupComponent, {
        ...ownProps,
        ...restComponentProps,
        className: clsx(className2, withBaseName("popup-component")),
        id: id2,
        width: ownWidth != null ? ownWidth : width2
      });
    };
    return /* @__PURE__ */ jsxs("div", {
      ...htmlAttributes,
      className,
      "data-testid": "dropdown",
      id: idProp,
      ref: handleRootRef,
      children: [
        getTriggerComponent(),
        isOpen && /* @__PURE__ */ jsx(Portal, {
          disablePortal,
          container,
          children: /* @__PURE__ */ jsx(Window, {
            className: clsx(withBaseName("popup"), classNameProp),
            id: `${id}-popup`,
            style: {
              top: y != null ? y : 0,
              left: x != null ? x : 0,
              position: strategy,
              maxHeight: maxPopupHeight
            },
            ref: handleFloatingRef,
            children: getPopupComponent()
          })
        })
      ]
    });
  }
);

export { DropdownBase };
