// @salt-ds/lab/dist-es/dropdown/Dropdown.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { useIdMemo, useForkRef } from '@salt-ds/core';
import { forwardRef, useRef, useCallback, cloneElement } from 'react';
import { CollectionProvider } from '../common-hooks/collectionProvider.js';
import { itemToString } from '../common-hooks/itemToString.js';
import '../common-hooks/keyUtils.js';
import { useCollectionItems } from '../common-hooks/useCollectionItems.js';
import '../responsive/useResizeObserver.js';
import { forwardCallbackProps } from '../utils/forwardCallbackProps.js';
import { List } from '../list/List.js';
import { DropdownBase } from './DropdownBase.js';
import { DropdownButton } from './DropdownButton.js';
import { useDropdown } from './useDropdown.js';

const Dropdown = forwardRef(function Dropdown2({
  "aria-label": ariaLabel,
  children,
  defaultIsOpen,
  defaultSelected,
  id: idProp,
  isOpen: isOpenProp,
  itemToString: itemToString$1 = itemToString,
  onOpenChange,
  onSelectionChange,
  onSelect,
  selected: selectedProp,
  selectionStrategy,
  source,
  triggerComponent,
  ListItem,
  ListProps: ListProps2,
  width = 180,
  ...props
}, forwardedRef) {
  const id = useIdMemo(idProp);
  const rootRef = useRef(null);
  const forkedRef = useForkRef(rootRef, forwardedRef);
  const collectionHook = useCollectionItems({
    id,
    source,
    children,
    options: {
      itemToString: itemToString$1
    }
  });
  const {
    highlightedIndex,
    triggerLabel,
    listHandlers,
    listControlProps,
    selected,
    ...dropdownListHook
  } = useDropdown({
    collectionHook,
    defaultHighlightedIndex: ListProps2 == null ? void 0 : ListProps2.defaultHighlightedIndex,
    defaultIsOpen,
    defaultSelected: collectionHook.itemToCollectionItem(defaultSelected),
    highlightedIndex: ListProps2 == null ? void 0 : ListProps2.highlightedIndex,
    isOpen: isOpenProp,
    itemToString: itemToString$1,
    label: "Dropdown",
    onHighlight: ListProps2 == null ? void 0 : ListProps2.onHighlight,
    onOpenChange,
    onSelectionChange,
    onSelect,
    selected: collectionHook.itemToCollectionItem(selectedProp),
    selectionStrategy
  });
  const collectionItemsToItem = useCallback(
    (itemOrItems) => {
      if (Array.isArray(itemOrItems)) {
        return itemOrItems.map((i) => i.value);
      } else if (itemOrItems) {
        return itemOrItems.value;
      }
    },
    []
  );
  const getTriggerComponent = () => {
    const ariaProps = {
      "aria-activedescendant": dropdownListHook.isOpen ? listControlProps == null ? void 0 : listControlProps["aria-activedescendant"] : void 0,
      "aria-label": ariaLabel
    };
    if (triggerComponent) {
      const ownProps = triggerComponent.props;
      return cloneElement(
        triggerComponent,
        forwardCallbackProps(ownProps, {
          ...dropdownListHook.isOpen ? listControlProps : {},
          ...ariaProps
        })
      );
    } else {
      return /* @__PURE__ */ jsx(DropdownButton, {
        label: triggerLabel,
        ...dropdownListHook.isOpen ? listControlProps : {},
        ...ariaProps
      });
    }
  };
  return /* @__PURE__ */ jsx(CollectionProvider, {
    collectionHook,
    children: /* @__PURE__ */ jsxs(DropdownBase, {
      ...props,
      id,
      isOpen: dropdownListHook.isOpen,
      onOpenChange: dropdownListHook.onOpenChange,
      ref: forkedRef,
      width,
      children: [
        getTriggerComponent(),
        /* @__PURE__ */ jsx(List, {
          ListItem,
          itemToString: itemToString$1,
          ...ListProps2,
          highlightedIndex,
          listHandlers,
          onSelectionChange,
          onSelect,
          selected: collectionItemsToItem(selected),
          selectionStrategy,
          "data-testid": "dropdown-list"
        })
      ]
    })
  });
});

export { Dropdown };
