// @salt-ds/lab/dist-es/dropdown/DropdownButton.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { makePrefixer, Button } from '@salt-ds/core';
import { ChevronDownIcon, DEFAULT_ICON_SIZE } from '@salt-ds/icons';
import '../form-field-context-legacy/FormFieldLegacyContext.js';
import { useFormFieldLegacyProps } from '../form-field-context-legacy/useFormFieldLegacyProps.js';
import { clsx } from 'clsx';
import { forwardRef } from 'react';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './DropdownButton.css.js';

const withBaseName = makePrefixer("saltDropdownButton");
const DropdownButton = forwardRef(function DropdownButton2({
  IconComponent = ChevronDownIcon,
  ariaHideOptionRole,
  className,
  disabled,
  iconSize = DEFAULT_ICON_SIZE,
  isOpen,
  label,
  labelId,
  fullWidth,
  posInSet,
  setSize,
  labelAriaAttributes,
  ...rest
}, ref) {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-dropdown-button",
    css: css_248z,
    window: targetWindow
  });
  const { inFormField } = useFormFieldLegacyProps();
  return /* @__PURE__ */ jsx(Button, {
    className: clsx(
      withBaseName(),
      {
        [withBaseName("fullwidth")]: fullWidth,
        [withBaseName("formField")]: inFormField
      },
      className
    ),
    "data-testid": "dropdown-button",
    disabled,
    variant: "secondary",
    ...rest,
    ref,
    children: /* @__PURE__ */ jsxs("div", {
      className: withBaseName("content"),
      children: [
        /* @__PURE__ */ jsx("span", {
          "aria-hidden": ariaHideOptionRole ? "true" : void 0,
          ...labelAriaAttributes,
          className: withBaseName("buttonLabel"),
          id: labelId,
          role: "option",
          children: label
        }),
        /* @__PURE__ */ jsx(IconComponent, {
          className: withBaseName("icon"),
          size: iconSize,
          "aria-label": null,
          "aria-hidden": "true"
        })
      ]
    })
  });
});

export { DropdownButton };
