// @salt-ds/lab/dist-es/dropdown/useDropdownBase.js
import { useControlled } from '@salt-ds/core';
import { useRef, useCallback, useState } from 'react';
import { useClickAway } from './useClickAway.js';
import 'react/jsx-runtime';
import { useResizeObserver, WidthOnly } from '../responsive/useResizeObserver.js';
import '../form-field-context-legacy/FormFieldLegacyContext.js';
import { useFormFieldLegacyProps } from '../form-field-context-legacy/useFormFieldLegacyProps.js';

const NO_OBSERVER = [];
const useDropdownBase = ({
  ariaLabelledBy: ariaLabelledByProp,
  defaultIsOpen,
  disabled,
  fullWidth: fullWidthProp,
  id,
  isOpen: isOpenProp,
  onOpenChange,
  onKeyDown: onKeyDownProp,
  openOnFocus,
  popupComponent: { props: componentProps },
  popupWidth: popupWidthProp,
  rootRef,
  width
}) => {
  var _a;
  const justFocused = useRef(null);
  const popperRef = useRef(null);
  const popperCallbackRef = useCallback((element) => {
    popperRef.current = element;
  }, []);
  const [isOpen, setIsOpen] = useControlled({
    controlled: isOpenProp,
    default: Boolean(defaultIsOpen),
    name: "useDropdown",
    state: "isOpen"
  });
  const {
    inFormField,
    setFocused: setFormFieldFocused,
    a11yProps: { "aria-labelledby": ariaLabelledBy, ...restA11yProps } = {}
  } = useFormFieldLegacyProps();
  const [popup, setPopup] = useState({
    width: (_a = popupWidthProp != null ? popupWidthProp : width) != null ? _a : 0
  });
  const showDropdown = useCallback(() => {
    setIsOpen(true);
    onOpenChange == null ? void 0 : onOpenChange(true);
  }, [onOpenChange, setIsOpen]);
  const hideDropdown = useCallback(() => {
    setIsOpen(false);
    onOpenChange == null ? void 0 : onOpenChange(false);
  }, [onOpenChange, setIsOpen]);
  useClickAway({
    popperRef,
    rootRef,
    isOpen,
    onClose: hideDropdown
  });
  const handleTriggerFocus = useCallback(() => {
    if (!disabled) {
      setFormFieldFocused == null ? void 0 : setFormFieldFocused(true);
      if (openOnFocus) {
        setIsOpen(true);
        onOpenChange == null ? void 0 : onOpenChange(true);
        justFocused.current = window.setTimeout(() => {
          justFocused.current = null;
        }, 1e3);
      }
    }
  }, [disabled, onOpenChange, openOnFocus, setFormFieldFocused, setIsOpen]);
  const handleTriggerBlur = useCallback(() => {
    setFormFieldFocused == null ? void 0 : setFormFieldFocused(false);
  }, [setFormFieldFocused]);
  const handleTriggerToggle = useCallback(
    (e) => {
      if (["Enter", " "].indexOf(
        e.key
      ) === -1) {
        const newIsOpen = !isOpen;
        setIsOpen(newIsOpen);
        onOpenChange == null ? void 0 : onOpenChange(newIsOpen);
      }
    },
    [isOpen, setIsOpen, onOpenChange]
  );
  const handleKeydown = useCallback(
    (evt) => {
      if ((evt.key === "Tab" || evt.key === "Escape") && isOpen) {
        hideDropdown();
      } else if ((evt.key === "Enter" || evt.key === "ArrowDown" || evt.key === " ") && !isOpen) {
        evt.preventDefault();
        showDropdown();
      } else {
        onKeyDownProp == null ? void 0 : onKeyDownProp(evt);
      }
    },
    [hideDropdown, isOpen, onKeyDownProp, showDropdown]
  );
  const fullWidth = fullWidthProp != null ? fullWidthProp : inFormField;
  const measurements2 = fullWidth ? WidthOnly : NO_OBSERVER;
  useResizeObserver(rootRef, measurements2, setPopup, fullWidth);
  const componentId = `${id}-dropdown`;
  const getAriaLabelledBy = (labelledBy, labelledByProp) => {
    if (labelledBy === void 0 && labelledByProp === void 0) {
      return void 0;
    } else {
      return [labelledBy, labelledByProp].filter((x) => !!x).join(" ");
    }
  };
  const triggerProps = {
    ...restA11yProps,
    "aria-expanded": isOpen,
    "aria-labelledby": getAriaLabelledBy(ariaLabelledBy, ariaLabelledByProp),
    "aria-owns": isOpen ? componentId : void 0,
    id: `${id}-control`,
    onClick: disabled || openOnFocus ? void 0 : handleTriggerToggle,
    onFocus: handleTriggerFocus,
    onBlur: handleTriggerBlur,
    role: "listbox",
    onKeyDown: disabled ? void 0 : handleKeydown,
    style: { width: fullWidth ? void 0 : width }
  };
  const dropdownComponentProps = {
    "aria-labelledby": ariaLabelledBy,
    id: componentId,
    width: popup.width
  };
  return {
    componentProps: dropdownComponentProps,
    popperRef: popperCallbackRef,
    isOpen,
    label: "Dropdown Button",
    triggerProps
  };
};

export { useDropdownBase };
