// @salt-ds/lab/dist-cjs/button-bar/OrderedButton.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var clsx = require('clsx');
var core = require('@salt-ds/core');
var ButtonBarContext = require('./internal/ButtonBarContext.js');
var useDescendant = require('./internal/useDescendant.js');
var window = require('@salt-ds/window');
var styles = require('@salt-ds/styles');
var OrderedButton$1 = require('./OrderedButton.css.js');

const withBasename = core.makePrefixer("saltOrderedButton");
const OrderedButton = React.forwardRef(
  function OrderedButton2({
    className,
    align: alignProp,
    order,
    alignLeftOrder,
    stackOrder,
    variant,
    ...restProps
  }, ref) {
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-ordered-button",
      css: OrderedButton$1,
      window: targetWindow
    });
    const index = useDescendant.useDescendant({ order, stackOrder, alignLeftOrder, variant });
    const {
      matches,
      align: alignContext,
      alignedIndex
    } = React.useContext(ButtonBarContext.ButtonBarContext);
    const alignFromParent = index === alignedIndex ? alignContext : void 0;
    const align = alignProp || alignFromParent;
    return /* @__PURE__ */ jsxRuntime.jsx(core.Button, {
      className: clsx.clsx(
        withBasename(),
        {
          [withBasename(`align${align ? core.capitalize(align) : ""}`)]: align && !matches,
          [withBasename("stacked")]: matches
        },
        className
      ),
      ref,
      variant,
      ...restProps
    });
  }
);

exports.OrderedButton = OrderedButton;
