// @salt-ds/lab/dist-cjs/contact-details/internal/FavoriteToggle.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var React = require('react');
var StarIconContainer = require('./StarIconContainer.js');

const FavoriteToggle = React.forwardRef(
  function FavoriteToggle2(props, ref) {
    const {
      isSelected: isSelectedProp,
      onBlur: onBlurProp,
      onFocus: onFocusProp,
      onClick: onClickProp,
      onChange: onChangeProp,
      onMouseEnter: onMouseEnterProp,
      onMouseLeave: onMouseLeaveProp,
      onKeyDown: onKeyDownProp,
      ...restProps
    } = props;
    const {
      isFocusVisibleRef,
      ref: focusVisibleRef,
      onFocus,
      onBlur
    } = core.useIsFocusVisible();
    const [isFocused, setIsFocused] = React.useState(false);
    const [isHighlighted, setIsHighlighted] = React.useState(false);
    const [isFocusVisible, setIsFocusVisible] = React.useState(false);
    const [isSelected, setIsSelected] = core.useControlled({
      controlled: isSelectedProp,
      default: false,
      name: "FavoriteToggle",
      state: "isSelected"
    });
    const toggleSelected = () => {
      setIsSelected((s) => !s);
      if (onChangeProp) {
        onChangeProp(!isSelected);
      }
    };
    const handleFocus = (event) => {
      if (onFocusProp) {
        onFocusProp(event);
      }
      onFocus(event);
      setIsFocused(true);
      setIsFocusVisible(isFocusVisibleRef.current);
    };
    const handleBlur = (event) => {
      if (onBlurProp) {
        onBlurProp(event);
      }
      onBlur();
      setIsFocused(false);
      setIsFocusVisible(false);
    };
    const handleClick = (event) => {
      if (onClickProp) {
        onClickProp(event);
      }
      setIsFocusVisible(isFocusVisibleRef.current);
      toggleSelected();
    };
    const handleMouseEnter = (event) => {
      if (onMouseEnterProp) {
        onMouseEnterProp(event);
      }
      setIsHighlighted(true);
    };
    const handleMouseLeave = (event) => {
      if (onMouseLeaveProp) {
        onMouseLeaveProp(event);
      }
      setIsHighlighted(false);
    };
    const handleKeyDown = (event) => {
      if (onKeyDownProp) {
        onKeyDownProp(event);
      }
      setIsFocusVisible(isFocusVisibleRef.current);
      if (["Enter", " "].includes(event.key)) {
        toggleSelected();
      }
    };
    return /* @__PURE__ */ jsxRuntime.jsx(StarIconContainer.StarIconContainer, {
      isFocusVisible,
      isFocused,
      isHighlighted,
      isSelected,
      onBlur: handleBlur,
      onClick: handleClick,
      onFocus: handleFocus,
      onKeyDown: handleKeyDown,
      onMouseEnter: handleMouseEnter,
      onMouseLeave: handleMouseLeave,
      ref: core.useForkRef(focusVisibleRef, ref),
      tabIndex: 0,
      ...restProps
    });
  }
);

exports.FavoriteToggle = FavoriteToggle;
