// @salt-ds/core/dist-es/dropdown/Dropdown.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { forwardRef, useRef, useEffect } from 'react';
import { ChevronUpIcon, ChevronDownIcon } from '@salt-ds/icons';
import { useListControl, defaultValueToString } from '../list-control/ListControlState.js';
import { makePrefixer } from '../utils/makePrefixer.js';
import { useFloatingComponent, useFloatingUI } from '../utils/useFloatingUI/useFloatingUI.js';
import { useForkRef } from '../utils/useForkRef.js';
import { useId } from '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import { clsx } from 'clsx';
import '../accordion/AccordionGroup.js';
import '../accordion/AccordionPanel.js';
import '../accordion/Accordion.js';
import '../accordion/AccordionHeader.js';
import '../aria-announcer/AriaAnnouncerContext.js';
import '../avatar/Avatar.js';
import '../badge/Badge.js';
import '../banner/Banner.js';
import '../banner/BannerActions.js';
import '../banner/BannerContent.js';
import '../border-item/BorderItem.js';
import '../border-layout/BorderLayout.js';
import '../button/Button.js';
import '../card/Card.js';
import '../checkbox/Checkbox.js';
import '../checkbox/CheckboxGroup.js';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import '../combo-box/ComboBox.js';
import '../dialog/Dialog.js';
import '../dialog/DialogHeader.js';
import '../dialog/DialogActions.js';
import '../dialog/DialogContent.js';
import '../dialog/DialogContext.js';
import '../dialog/DialogCloseButton.js';
import '../drawer/Drawer.js';
import '../drawer/DrawerCloseButton.js';
import '../file-drop-zone/FileDropZone.js';
import '../file-drop-zone/FileDropZoneIcon.js';
import '../file-drop-zone/FileDropZoneTrigger.js';
import '../flex-item/FlexItem.js';
import '../flex-layout/FlexLayout.js';
import '../flow-layout/FlowLayout.js';
import '../form-field/FormField.js';
import '../status-indicator/StatusIndicator.js';
import '../text/Text.js';
import '../text/Code.js';
import '../text/Display.js';
import '../text/Headings.js';
import '../text/Label.js';
import '../text/TextAction.js';
import '../text/TextNotation.js';
import '../form-field-context/FormFieldContext.js';
import { useFormFieldProps } from '../form-field-context/useFormFieldProps.js';
import '../grid-item/GridItem.js';
import '../grid-layout/GridLayout.js';
import '../input/Input.js';
import '../interactable-card/InteractableCard.js';
import '../interactable-card/InteractableCardGroup.js';
import '../interactable-card/InteractableCardGroupContext.js';
import '../link/Link.js';
import '../link-card/LinkCard.js';
import { ListControlContext } from '../list-control/ListControlContext.js';
import '../multiline-input/MultilineInput.js';
import '../navigation-item/NavigationItem.js';
import '../option/Option.js';
import '../option/OptionGroup.js';
import '../pagination/Pagination.js';
import '../pagination/Paginator.js';
import '../pagination/CompactInput.js';
import '../pagination/CompactPaginator.js';
import '../pagination/GoToInput.js';
import '../panel/Panel.js';
import '../pill/Pill.js';
import '../progress/CircularProgress/CircularProgress.js';
import '../progress/LinearProgress/LinearProgress.js';
import '../radio-button/RadioButton.js';
import '../radio-button/RadioButtonGroup.js';
import '../scrim/Scrim.js';
import '../segmented-button-group/SegmentedButtonGroup.js';
import '../spinner/Spinner.js';
import '../stack-layout/StackLayout.js';
import { StatusAdornment } from '../status-adornment/StatusAdornment.js';
import '../toast/Toast.js';
import '../toast/ToastContent.js';
import '../split-layout/SplitLayout.js';
import '../switch/Switch.js';
import '../toggle-button/ToggleButton.js';
import '../toggle-button-group/ToggleButtonGroup.js';
import '../toggle-button-group/ToggleButtonGroupContext.js';
import '../tooltip/Tooltip.js';
import { size, flip, useInteractions, useDismiss, useFocus, useClick } from '@floating-ui/react';
import css_248z from './Dropdown.css.js';
import { OptionList } from '../option/OptionList.js';

function ExpandIcon({ open }) {
  return open ? /* @__PURE__ */ jsx(ChevronUpIcon, {
    "aria-hidden": true
  }) : /* @__PURE__ */ jsx(ChevronDownIcon, {
    "aria-hidden": true
  });
}
const withBaseName = makePrefixer("saltDropdown");
const Dropdown = forwardRef(function Dropdown2(props, ref) {
  var _a, _b;
  const {
    "aria-labelledby": ariaLabelledBy,
    "aria-describedby": ariaDescribedBy,
    children,
    className,
    disabled: disabledProp,
    emptyReadOnlyMarker = "\u2014",
    readOnly: readOnlyProp,
    multiselect,
    onSelectionChange,
    selected,
    defaultSelected,
    defaultOpen,
    value,
    onOpenChange,
    open,
    placeholder,
    startAdornment,
    required: requiredProp,
    variant = "primary",
    validationStatus: validationStatusProp,
    onKeyDown,
    onFocus,
    onBlur,
    valueToString = defaultValueToString,
    ...rest
  } = props;
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-dropdown",
    css: css_248z,
    window: targetWindow
  });
  const {
    a11yProps: {
      "aria-describedby": formFieldDescribedBy,
      "aria-labelledby": formFieldLabelledBy
    } = {},
    disabled: formFieldDisabled,
    readOnly: formFieldReadOnly,
    necessity: formFieldRequired,
    validationStatus: formFieldValidationStatus
  } = useFormFieldProps();
  const disabled = Boolean(disabledProp) || formFieldDisabled;
  const readOnly = Boolean(readOnlyProp) || formFieldReadOnly;
  const validationStatus = validationStatusProp != null ? validationStatusProp : formFieldValidationStatus;
  const required = formFieldRequired ? ["required", "asterisk"].includes(formFieldRequired) : requiredProp;
  const listControl = useListControl({
    open,
    defaultOpen,
    onOpenChange,
    multiselect,
    defaultSelected,
    selected,
    onSelectionChange,
    valueToString
  });
  const {
    activeState,
    setActive,
    openState,
    setOpen,
    openKey,
    getOptionAtIndex,
    getIndexOfOption,
    getOptionsMatching,
    getOptionFromSearch,
    options,
    selectedState,
    select,
    setFocusVisibleState,
    focusedState,
    setFocusedState,
    listRef
  } = listControl;
  const selectedValue = selectedState.map((item) => valueToString(item)).join(", ");
  const isEmptyReadOnly = readOnly && selectedValue === "";
  const valueText = isEmptyReadOnly ? emptyReadOnlyMarker : value != null ? value : selectedValue;
  useFloatingComponent();
  const handleOpenChange = (newOpen, _event, reason) => {
    const focusNotBlur = reason === "focus" && newOpen;
    if (readOnly || focusNotBlur)
      return;
    setOpen(newOpen);
  };
  const { x, y, strategy, elements, floating, reference, context } = useFloatingUI({
    open: openState && !readOnly && children != void 0,
    onOpenChange: handleOpenChange,
    placement: "bottom-start",
    middleware: [
      size({
        apply({ rects, elements: elements2, availableHeight }) {
          Object.assign(elements2.floating.style, {
            minWidth: `${rects.reference.width}px`,
            maxHeight: `max(calc((var(--salt-size-base) + var(--salt-spacing-100)) * 5), calc(${availableHeight}px - var(--salt-spacing-100)))`
          });
        }
      }),
      flip({ fallbackStrategy: "initialPlacement" })
    ]
  });
  const { getReferenceProps, getFloatingProps } = useInteractions([
    useDismiss(context),
    useFocus(context),
    useClick(context)
  ]);
  const buttonRef = useRef(null);
  const handleTriggerRef = useForkRef(reference, buttonRef);
  const handleButtonRef = useForkRef(handleTriggerRef, ref);
  const typeaheadString = useRef("");
  const typeaheadTimeout = useRef();
  const handleTypeahead = (event) => {
    if (typeaheadTimeout.current) {
      clearTimeout(typeaheadTimeout.current);
    }
    typeaheadString.current += event.key;
    typeaheadTimeout.current = window.setTimeout(() => {
      typeaheadString.current = "";
    }, 500);
    if (!openState) {
      setOpen(true, "input");
    }
    let newOption = getOptionFromSearch(typeaheadString.current, activeState);
    if (!newOption) {
      newOption = getOptionFromSearch(typeaheadString.current);
    }
    if (newOption) {
      setActive(newOption);
      setFocusVisibleState(true);
    }
  };
  const handleKeyDown = (event) => {
    const currentIndex = activeState ? getIndexOfOption(activeState) : -1;
    const count = options.length - 1;
    if (readOnly) {
      return;
    }
    if (!openState) {
      if (event.key === "ArrowDown" || event.key === "ArrowUp") {
        setOpen(true, void 0, event.key);
        return;
      }
    }
    if (event.key.length === 1 && !event.ctrlKey && !event.metaKey && !event.altKey) {
      event.preventDefault();
      event.stopPropagation();
      handleTypeahead(event);
    }
    let newActive;
    switch (event.key) {
      case "ArrowDown":
        newActive = getOptionAtIndex(Math.min(count, currentIndex + 1));
        break;
      case "ArrowUp":
        newActive = getOptionAtIndex(Math.max(0, currentIndex - 1));
        break;
      case "Home":
        newActive = getOptionAtIndex(0);
        break;
      case "End":
        newActive = getOptionAtIndex(count);
        break;
      case "PageUp":
        newActive = getOptionAtIndex(Math.max(0, currentIndex - 10));
        break;
      case "PageDown":
        newActive = getOptionAtIndex(Math.min(count, currentIndex + 10));
        break;
      case "Enter":
      case " ":
        if (openState && Boolean(activeState == null ? void 0 : activeState.disabled) || typeaheadString.current.trim().length > 0 && event.key === " ") {
          event.preventDefault();
          return;
        }
        if (!openState || !activeState) {
          return;
        }
        event.preventDefault();
        select(event, activeState);
        break;
      case "Tab":
        if (!multiselect && activeState) {
          select(event, activeState);
        }
        break;
    }
    if (newActive && (newActive == null ? void 0 : newActive.id) != (activeState == null ? void 0 : activeState.id)) {
      event.preventDefault();
      setActive(newActive);
      setFocusVisibleState(true);
    }
    onKeyDown == null ? void 0 : onKeyDown(event);
  };
  const handleFocus = (event) => {
    setFocusedState(true);
    onFocus == null ? void 0 : onFocus(event);
  };
  const handleBlur = (event) => {
    setFocusedState(false);
    onBlur == null ? void 0 : onBlur(event);
  };
  const handleListMouseOver = () => {
    setFocusVisibleState(false);
  };
  const handleFocusButton = () => {
    var _a2;
    (_a2 = buttonRef.current) == null ? void 0 : _a2.focus();
  };
  useEffect(() => {
    const activeIndex = activeState ? getIndexOfOption(activeState) : -1;
    let newActive = void 0;
    if (activeIndex > 0) {
      return;
    }
    if (!openState) {
      setActive(void 0);
      return;
    }
    if (selectedState.length > 0) {
      newActive = getOptionsMatching(
        (option) => option.value === selectedState[0]
      ).pop();
    }
    if (!newActive) {
      if (openKey.current === "ArrowDown") {
        newActive = getOptionAtIndex(0);
      } else if (openKey.current === "ArrowUp") {
        newActive = getOptionAtIndex(options.length - 1);
      }
    }
    if (!newActive) {
      newActive = getOptionAtIndex(0);
    }
    setActive(newActive);
  }, [openState, children]);
  const listId = useId();
  const handleListRef = useForkRef(listRef, floating);
  return /* @__PURE__ */ jsxs(ListControlContext.Provider, {
    value: listControl,
    children: [
      /* @__PURE__ */ jsxs("button", {
        className: clsx(
          withBaseName(),
          withBaseName(variant),
          {
            [withBaseName("disabled")]: disabled,
            [withBaseName(validationStatus != null ? validationStatus : "")]: validationStatus
          },
          className
        ),
        ref: handleButtonRef,
        role: "combobox",
        type: "button",
        disabled,
        "aria-readonly": readOnly ? "true" : void 0,
        "aria-required": required ? "true" : void 0,
        "aria-expanded": openState,
        "aria-activedescendant": activeState == null ? void 0 : activeState.id,
        "aria-labelledby": clsx(formFieldLabelledBy, ariaLabelledBy) || void 0,
        "aria-describedby": clsx(formFieldDescribedBy, ariaDescribedBy) || void 0,
        "aria-multiselectable": multiselect,
        "aria-controls": openState ? listId : void 0,
        ...getReferenceProps({
          onKeyDown: handleKeyDown,
          onFocus: handleFocus,
          onBlur: handleBlur,
          ...rest
        }),
        children: [
          startAdornment,
          /* @__PURE__ */ jsx("span", {
            className: clsx(withBaseName("content"), {
              [withBaseName("placeholder")]: !valueText
            }),
            children: !valueText ? placeholder : valueText
          }),
          validationStatus && /* @__PURE__ */ jsx(StatusAdornment, {
            status: validationStatus
          }),
          !readOnly && /* @__PURE__ */ jsx(ExpandIcon, {
            open: openState
          })
        ]
      }),
      /* @__PURE__ */ jsx(OptionList, {
        open: (openState || focusedState) && !readOnly && children != void 0,
        ...getFloatingProps({
          onMouseOver: handleListMouseOver,
          onFocus: handleFocusButton,
          onClick: handleFocusButton
        }),
        left: x != null ? x : 0,
        top: y != null ? y : 0,
        position: strategy,
        width: (_a = elements.floating) == null ? void 0 : _a.offsetWidth,
        height: (_b = elements.floating) == null ? void 0 : _b.offsetHeight,
        ref: handleListRef,
        id: listId,
        collapsed: !openState,
        children
      })
    ]
  });
});

export { Dropdown };
