// @salt-ds/lab/dist-es/tokenized-input-next/internal/InputPill.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { clsx } from 'clsx';
import { memo, useRef, useState } from 'react';
import { makePrefixer, useIsomorphicLayoutEffect, Tooltip, Pill } from '@salt-ds/core';
import { getWidth } from './useWidth.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './InputPill.css.js';
import { CloseIcon } from '@salt-ds/icons';

const withBaseName = makePrefixer("saltInputPill");
const InputPill = memo(function InputPill2(props) {
  const {
    className,
    hidden,
    highlighted,
    index,
    label,
    onClose,
    pillsRef,
    ...rest
  } = props;
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-input-pill",
    css: css_248z,
    window: targetWindow
  });
  const ref = useRef(null);
  const [isEllipsisActive, setEllipsisActive] = useState(false);
  const isRemovable = Boolean(onClose);
  useIsomorphicLayoutEffect(() => {
    var _a;
    const text = (_a = ref == null ? void 0 : ref.current) == null ? void 0 : _a.firstElementChild;
    if (!isRemovable && pillsRef.current) {
      pillsRef.current[index] = getWidth(ref.current);
    }
    setEllipsisActive((text == null ? void 0 : text.offsetWidth) < (text == null ? void 0 : text.scrollWidth));
  }, [pillsRef, index, isRemovable]);
  useIsomorphicLayoutEffect(
    () => () => {
      pillsRef.current[index] = void 0;
    },
    [pillsRef, index]
  );
  const handleClose = (event) => {
    onClose == null ? void 0 : onClose(event, index);
  };
  return /* @__PURE__ */ jsx(Tooltip, {
    content: label,
    disabled: !isEllipsisActive,
    children: /* @__PURE__ */ jsxs(Pill, {
      className: clsx(
        withBaseName(),
        {
          [withBaseName("pillHighlighted")]: highlighted,
          [withBaseName("expanded")]: isRemovable,
          [withBaseName("hidden")]: hidden
        },
        className
      ),
      tabIndex: -1,
      onClick: isRemovable ? handleClose : void 0,
      ref,
      role: "option",
      ...rest,
      children: [
        /* @__PURE__ */ jsx("span", {
          className: withBaseName("label"),
          children: label
        }),
        isRemovable && /* @__PURE__ */ jsx(CloseIcon, {})
      ]
    })
  });
});

export { InputPill };
