// @salt-ds/lab/dist-cjs/carousel/CarouselSlide.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var core = require('@salt-ds/core');
var clsx = require('clsx');

const withBaseName = core.makePrefixer("saltCarouselSlide");
const CarouselSlide = React.forwardRef(
  function CarouselSlide2({ ButtonBar, Media, description, title, contentAlignment }, ref) {
    const buttonBarRef = React.useRef(null);
    return /* @__PURE__ */ jsxRuntime.jsxs("div", {
      ref,
      children: [
        Media && /* @__PURE__ */ jsxRuntime.jsx("div", {
          className: withBaseName("mediaContainer"),
          children: Media
        }),
        /* @__PURE__ */ jsxRuntime.jsxs("div", {
          className: withBaseName("fixedContainer"),
          ref: buttonBarRef,
          children: [
            /* @__PURE__ */ jsxRuntime.jsxs("div", {
              className: clsx.clsx({
                [withBaseName("textContainer")]: contentAlignment === "center",
                [withBaseName("textContainerLeft")]: contentAlignment === "left"
              }),
              children: [
                title && /* @__PURE__ */ jsxRuntime.jsx("div", {
                  "aria-level": 1,
                  className: withBaseName("titleContainer"),
                  role: "heading",
                  children: title
                }),
                description && /* @__PURE__ */ jsxRuntime.jsx("div", {
                  className: withBaseName("descriptionContainer"),
                  children: description
                })
              ]
            }),
            ButtonBar && /* @__PURE__ */ jsxRuntime.jsx("div", {
              className: clsx.clsx({
                [withBaseName("buttonBarOverride")]: contentAlignment === "center",
                [withBaseName("buttonBarOverrideLeft")]: contentAlignment === "left"
              }),
              children: /* @__PURE__ */ jsxRuntime.jsx(ButtonBar, {
                className: clsx.clsx({
                  [withBaseName("buttonBarContainer")]: contentAlignment === "center",
                  [withBaseName("buttonBarContainerLeft")]: contentAlignment === "left"
                }),
                stackAtBreakpoint: 0
              })
            })
          ]
        })
      ]
    });
  }
);

exports.CarouselSlide = CarouselSlide;
