// @salt-ds/lab/dist-es/tree/Tree.js
import { jsx } from 'react/jsx-runtime';
import { forwardRef, useRef, useCallback, isValidElement, createElement } from 'react';
import { makePrefixer, useIdMemo, useForkRef } from '@salt-ds/core';
import { clsx } from 'clsx';
import { useAutoSizer } from '../common-hooks/useAutoSizer.js';
import { calcPreferredHeight } from '../common-hooks/calcPreferredHeight.js';
import '../common-hooks/collectionProvider.js';
import '../common-hooks/keyUtils.js';
import { closestListItemIndex } from '../common-hooks/list-dom-utils.js';
import { useCollectionItems } from '../common-hooks/useCollectionItems.js';
import { GROUP_SELECTION_NONE } from '../common-hooks/useSelection.js';
import '../responsive/useResizeObserver.js';
import { isSelected } from '../common-hooks/utils/isSelected.js';
import { useTree } from './useTree.js';
import { TreeNode } from './TreeNode.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './Tree.css.js';

const withBaseName = makePrefixer("saltTree");
const Tree = forwardRef(function Tree2({
  className,
  defaultSelected,
  disabled,
  groupSelection = GROUP_SELECTION_NONE,
  height,
  id: idProp,
  onHighlight,
  onToggle,
  onSelect,
  onSelectionChange,
  revealSelected,
  selected: selectedProp,
  selectionStrategy,
  source,
  style: styleProp,
  width,
  ...htmlAttributes
}, forwardedRef) {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-tree",
    css: css_248z,
    window: targetWindow
  });
  const id = useIdMemo(idProp);
  const rootRef = useRef(null);
  const contentRef = useRef(null);
  const collectionHook = useCollectionItems({
    id,
    source,
    options: {
      noChildrenLabel: "No children available",
      revealSelected: revealSelected ? Boolean(selectedProp) || Boolean(defaultSelected) || false : void 0
    }
  });
  const preferredHeight = height != null ? height : calcPreferredHeight({
    displayedItemCount: 10,
    itemCount: collectionHook.data.length,
    itemHeight: 36
  });
  const autoSize = useAutoSizer({
    containerRef: rootRef,
    responsive: width === void 0 || height === void 0,
    height: preferredHeight,
    width
  });
  const handleSelect = useCallback(
    (evt, selectedItem) => {
      if (onSelect) {
        if (isValidElement(selectedItem.value)) {
          onSelect(evt, selectedItem.label);
        } else if (selectedItem.value !== null) {
          onSelect(evt, selectedItem.value);
        }
      }
    },
    [onSelect]
  );
  const handleSelectionChange = useCallback(
    (evt, selected2) => {
      if (onSelectionChange) {
        onSelectionChange(
          evt,
          Array.isArray(selected2) ? selected2.map((s) => s.value) : selected2 && selected2.value
        );
      }
    },
    [onSelectionChange]
  );
  const {
    focusVisible,
    highlightedIdx,
    highlightItemAtIndex,
    listHandlers,
    listProps,
    listItemHandlers,
    selected
  } = useTree({
    collectionHook,
    containerRef: rootRef,
    contentRef,
    defaultSelected: collectionHook.itemToCollectionItem(defaultSelected),
    disabled,
    groupSelection,
    onHighlight,
    onSelect: handleSelect,
    onSelectionChange: handleSelectionChange,
    onToggle,
    selected: collectionHook.itemToCollectionItem(selectedProp),
    selectionStrategy
  });
  const defaultItemHandlers = {
    onMouseEnter: (evt) => {
      const idx = closestListItemIndex(evt.target);
      highlightItemAtIndex(idx);
    }
  };
  const propsCommonToAllListItems = {
    ...defaultItemHandlers,
    ...listItemHandlers,
    isLeaf: true,
    role: "treeitem"
  };
  const allowGroupSelect = false;
  function addLeafNode(list, item, idx) {
    const itemProps = {
      "aria-disabled": disabled || item.disabled,
      "aria-level": item.level,
      "data-idx": idx.value,
      description: item.description,
      id: item.id,
      key: item.id,
      highlighted: idx.value === highlightedIdx || void 0,
      selected: isSelected(selected, item),
      className: clsx({
        focusVisible: focusVisible === idx.value
      })
    };
    list.push(
      /* @__PURE__ */ jsx(TreeNode, {
        ...propsCommonToAllListItems,
        ...itemProps,
        label: item.label
      })
    );
    idx.value += 1;
  }
  function addGroupNode(list, items, idx, id2, title) {
    const { value: i } = idx;
    const item = items[i];
    idx.value += 1;
    list.push(
      /* @__PURE__ */ createElement(TreeNode, {
        ...defaultItemHandlers,
        ...listItemHandlers,
        "aria-disabled": disabled || item.disabled,
        "aria-expanded": item.expanded,
        "aria-level": item.level,
        className: clsx({
          focusVisible: focusVisible === i,
          [withBaseName("toggle")]: !allowGroupSelect
        }),
        "data-idx": i,
        "data-selectable": true,
        description: item.description,
        highlighted: i === highlightedIdx,
        id: id2,
        key: `header-${i}`,
        label: title,
        selected: isSelected(selected, item)
      }, item.expanded ? /* @__PURE__ */ jsx("ul", {
        className: withBaseName("child-nodes"),
        role: "group",
        children: renderItems(items, idx, item.level + 1)
      }) : null)
    );
  }
  const renderItems = (items, idx = { value: 0 }, level = 1) => {
    const listItems = [];
    while (idx.value < items.length) {
      const item = items[idx.value];
      if (item.level < level) {
        break;
      }
      if (item.childNodes) {
        addGroupNode(listItems, items, idx, item.id, item.label);
      } else {
        addLeafNode(listItems, item, idx);
      }
    }
    return listItems;
  };
  const renderContent = () => {
    if (collectionHook.data.length) {
      return renderItems(collectionHook.data);
    }
  };
  return /* @__PURE__ */ jsx("div", {
    ...htmlAttributes,
    ...listHandlers,
    ...listProps,
    className: clsx(withBaseName(), className),
    id: `Tree-${id}`,
    ref: useForkRef(rootRef, forwardedRef),
    style: { ...styleProp, ...autoSize },
    tabIndex: 0,
    children: /* @__PURE__ */ jsx("ul", {
      className: withBaseName("scrollingContentContainer"),
      ref: contentRef,
      role: "tree",
      children: renderContent()
    })
  });
});

export { Tree };
