// @salt-ds/lab/dist-es/form-field-legacy/FormLabel.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { clsx } from 'clsx';
import { makePrefixer } from '@salt-ds/core';
import { NecessityIndicator } from './NecessityIndicator.js';
import { StatusIndicator } from './StatusIndicator.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './FormLabel.css.js';

const withBaseName = makePrefixer("saltFormLabel");
const FormLabel = ({
  label,
  NecessityIndicator: NecessityIndicator$1 = NecessityIndicator,
  required,
  displayedNecessity,
  hasStatusIndicator = false,
  StatusIndicatorProps: StatusIndicatorProps2,
  validationStatus,
  necessityText,
  className,
  disabled,
  readOnly,
  tooltipText,
  ...restProps
}) => {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-form-label",
    css: css_248z,
    window: targetWindow
  });
  return /* @__PURE__ */ jsxs("label", {
    className: clsx(withBaseName(), className, {
      [withBaseName("disabled")]: disabled
    }),
    ...restProps,
    children: [
      label,
      /* @__PURE__ */ jsx(NecessityIndicator$1, {
        required,
        displayedNecessity,
        necessityText,
        className: withBaseName("necessityIndicator")
      }),
      hasStatusIndicator && /* @__PURE__ */ jsx(StatusIndicator, {
        className: withBaseName("statusIndicator"),
        status: validationStatus,
        tooltipText,
        hasTooltip: true,
        ...StatusIndicatorProps2
      })
    ]
  });
};

export { FormLabel };
