// @salt-ds/lab/dist-es/form-field-legacy/FormActivationIndicator.js
import { jsx } from 'react/jsx-runtime';
import { makePrefixer } from '@salt-ds/core';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './FormActivationIndicator.css.js';

const ErrorIndicatorIcon = (props) => {
  return /* @__PURE__ */ jsx("svg", {
    viewBox: "0 0 16 16",
    focusable: false,
    "data-testid": "ErrorIndicatorIcon",
    ...props,
    children: /* @__PURE__ */ jsx("circle", {
      cx: 8,
      cy: 8,
      r: 8
    })
  });
};
const WarningIndicatorIcon = (props) => {
  return /* @__PURE__ */ jsx("svg", {
    viewBox: "0 0 16 16",
    focusable: false,
    "data-testid": "WarningIndicatorIcon",
    ...props,
    children: /* @__PURE__ */ jsx("polygon", {
      points: "0, 16 16, 16 16, 0"
    })
  });
};
const withBaseName = makePrefixer("saltFormActivationIndicator");
const ActivationIndicatorIcon = ({
  validationStatus,
  ...restSvgProps
}) => {
  if (validationStatus === "error") {
    return /* @__PURE__ */ jsx(ErrorIndicatorIcon, {
      ...restSvgProps
    });
  } else if (validationStatus === "warning") {
    return /* @__PURE__ */ jsx(WarningIndicatorIcon, {
      ...restSvgProps
    });
  } else {
    return null;
  }
};
const FormActivationIndicator = ({
  hasIcon,
  validationStatus
}) => {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-form-activation-indicator",
    css: css_248z,
    window: targetWindow
  });
  const rootClass = "saltFormActivationIndicator";
  return /* @__PURE__ */ jsx("div", {
    className: withBaseName(),
    children: /* @__PURE__ */ jsx("div", {
      className: withBaseName("bar"),
      children: hasIcon && validationStatus && /* @__PURE__ */ jsx(ActivationIndicatorIcon, {
        className: `${rootClass}-icon`,
        validationStatus
      })
    })
  });
};

export { FormActivationIndicator };
