// @salt-ds/lab/dist-cjs/contact-details/ContactMetadataItem.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var React = require('react');
var MailLinkComponent = require('./MailLinkComponent.js');

const withBaseName = core.makePrefixer("saltContactMetadataItem");
const ContactMetadataItem = React.forwardRef(function ContactMetadataItem2(props, ref) {
  const { label, value, icon: Icon, ValueComponent, ...restProps } = props;
  return /* @__PURE__ */ jsxRuntime.jsxs("div", {
    ...restProps,
    ref,
    className: withBaseName(),
    children: [
      !Icon ? /* @__PURE__ */ jsxRuntime.jsx(core.Text, {
        maxRows: 1,
        className: withBaseName("label"),
        variant: "secondary",
        children: label
      }) : null,
      Icon ? /* @__PURE__ */ jsxRuntime.jsx(Icon, {
        className: withBaseName("icon")
      }) : null,
      ValueComponent ? /* @__PURE__ */ jsxRuntime.jsx(ValueComponent, {
        value
      }) : /* @__PURE__ */ jsxRuntime.jsx(MailLinkComponent.MailLinkComponent, {
        value
      })
    ]
  });
});

exports.ContactMetadataItem = ContactMetadataItem;
