// @salt-ds/lab/dist-cjs/contact-details/ContactFavoriteToggle.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var React = require('react');
var FavoriteToggleWithTooltip = require('./internal/FavoriteToggleWithTooltip.js');
require('./internal/ContactDetailsContext.js');

const ContactFavoriteToggle = React.forwardRef(function ContactFavoriteToggle2(props, ref) {
  const {
    id: idProp,
    isFavorite: isFavoriteProp,
    defaultIsFavorite = false,
    onChange: onChangeProp,
    tooltipProps,
    ...restProps
  } = props;
  const id = core.useId(idProp);
  const [isFavorite, setIsFavorite] = core.useControlled({
    controlled: isFavoriteProp,
    default: defaultIsFavorite,
    name: "ContactDetails",
    state: "isFavorite"
  });
  const onChange = (isSelected) => {
    setIsFavorite(isSelected);
    if (onChangeProp) {
      onChangeProp(isSelected);
    }
  };
  const tooltipTitle = `${isFavorite ? "Remove" : "Mark"} as favorite`;
  return /* @__PURE__ */ jsxRuntime.jsx(FavoriteToggleWithTooltip.FavoriteToggleWithTooltip, {
    ...restProps,
    id,
    ref,
    tooltipProps,
    "aria-label": "Favorite",
    isSelected: isFavorite,
    onChange,
    tooltipTitle
  });
});

exports.ContactFavoriteToggle = ContactFavoriteToggle;
