// @salt-ds/lab/dist-cjs/calendar/internal/CalendarNavigation.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var core = require('@salt-ds/core');
var icons = require('@salt-ds/icons');
var clsx = require('clsx');
var CalendarContext = require('./CalendarContext.js');
var CalendarNavigation$1 = require('./CalendarNavigation.css.js');
var date = require('@internationalized/date');
var utils = require('./utils.js');
var window = require('@salt-ds/window');
var styles = require('@salt-ds/styles');

const withBaseName = core.makePrefixer("saltCalendarNavigation");
function useCalendarNavigation() {
  const {
    state: { visibleMonth, minDate, maxDate },
    helpers: {
      setVisibleMonth,
      isDayVisible,
      isOutsideAllowedYears,
      isOutsideAllowedMonths
    }
  } = CalendarContext.useCalendarContext();
  const moveToNextMonth = (event) => {
    setVisibleMonth(event, visibleMonth.add({ months: 1 }));
  };
  const moveToPreviousMonth = (event) => {
    setVisibleMonth(event, visibleMonth.subtract({ months: 1 }));
  };
  const moveToMonth = (event, month) => {
    let newMonth = month;
    if (!isOutsideAllowedYears(newMonth)) {
      if (isOutsideAllowedMonths(newMonth)) {
        const navigableMonths = utils.monthsForLocale(visibleMonth).filter(
          (n) => !isOutsideAllowedMonths(n)
        );
        newMonth = navigableMonths.reduce(
          (closestMonth, currentMonth) => Math.abs(utils.monthDiff(currentMonth, newMonth)) < Math.abs(utils.monthDiff(closestMonth, newMonth)) ? currentMonth : closestMonth
        );
      }
      setVisibleMonth(event, newMonth);
    }
  };
  const months = utils.monthsForLocale(visibleMonth);
  const years = [-2, -1, 0, 1, 2].map((delta) => visibleMonth.add({ years: delta })).filter((year) => !isOutsideAllowedYears(year));
  const selectedMonth = months.find(
    (month) => date.isSameMonth(month, visibleMonth)
  );
  const selectedYear = years.find(
    (year) => date.isSameYear(year, visibleMonth)
  );
  const canNavigatePrevious = !(minDate && isDayVisible(minDate));
  const canNavigateNext = !(maxDate && isDayVisible(maxDate));
  return {
    moveToNextMonth,
    moveToPreviousMonth,
    moveToMonth,
    visibleMonth,
    months,
    years,
    canNavigateNext,
    canNavigatePrevious,
    selectedMonth,
    selectedYear,
    isOutsideAllowedMonths
  };
}
const OptionWithTooltip = ({
  value,
  children,
  disabled,
  tooltipContent
}) => {
  const { activeState, openState } = core.useListControlContext();
  const open = (activeState == null ? void 0 : activeState.value) === value;
  return /* @__PURE__ */ jsxRuntime.jsx(core.Tooltip, {
    placement: "right",
    open: open && openState,
    disabled: !disabled,
    content: tooltipContent,
    enterDelay: 0,
    leaveDelay: 0,
    children: /* @__PURE__ */ jsxRuntime.jsx(core.Option, {
      value,
      disabled,
      children
    })
  });
};
const CalendarNavigation = React.forwardRef(function CalendarNavigation2(props, ref) {
  const {
    className,
    MonthDropdownProps,
    YearDropdownProps,
    hideYearDropdown,
    ...rest
  } = props;
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-calendar-navigation",
    css: CalendarNavigation$1,
    window: targetWindow
  });
  const {
    moveToPreviousMonth,
    moveToNextMonth,
    moveToMonth,
    months,
    years,
    canNavigateNext,
    canNavigatePrevious,
    visibleMonth,
    selectedMonth,
    selectedYear,
    isOutsideAllowedMonths
  } = useCalendarNavigation();
  const handleNavigatePrevious = (event) => {
    moveToPreviousMonth(event);
  };
  const handleNavigateNext = (event) => {
    moveToNextMonth(event);
  };
  const handleMonthSelect = (event, month) => {
    moveToMonth(event, month[0]);
  };
  const handleYearSelect = (event, year) => {
    moveToMonth(event, year[0]);
  };
  const formatMonth = (date) => {
    return !date ? "" : utils.formatDate(date, { month: hideYearDropdown ? "long" : "short" });
  };
  const formatYear = (date) => {
    return !date ? "" : utils.formatDate(date, { year: "numeric" });
  };
  return /* @__PURE__ */ jsxRuntime.jsxs("div", {
    className: clsx.clsx(
      withBaseName(),
      { [withBaseName("hideYearDropdown")]: hideYearDropdown },
      className
    ),
    ref,
    ...rest,
    children: [
      /* @__PURE__ */ jsxRuntime.jsx(core.Tooltip, {
        placement: "top",
        disabled: canNavigatePrevious,
        content: "Past dates are out of range",
        enterDelay: 0,
        leaveDelay: 0,
        children: /* @__PURE__ */ jsxRuntime.jsx(core.Button, {
          disabled: !canNavigatePrevious,
          variant: "secondary",
          onClick: handleNavigatePrevious,
          focusableWhenDisabled: true,
          children: /* @__PURE__ */ jsxRuntime.jsx(icons.ChevronLeftIcon, {
            "aria-label": `Previous Month, ${utils.formatDate(
              visibleMonth.subtract({ months: 1 })
            )}`
          })
        })
      }),
      /* @__PURE__ */ jsxRuntime.jsxs("div", {
        className: withBaseName("dropdowns"),
        children: [
          /* @__PURE__ */ jsxRuntime.jsx(core.Dropdown, {
            "aria-label": "Month Dropdown",
            selected: selectedMonth ? [selectedMonth] : [],
            value: formatMonth(selectedMonth),
            onSelectionChange: handleMonthSelect,
            ...MonthDropdownProps,
            children: months.map((month) => /* @__PURE__ */ jsxRuntime.jsx(OptionWithTooltip, {
              value: month,
              disabled: isOutsideAllowedMonths(month),
              tooltipContent: "This month is out of range",
              children: formatMonth(month)
            }, formatMonth(month)))
          }),
          !hideYearDropdown && /* @__PURE__ */ jsxRuntime.jsx(core.Dropdown, {
            "aria-label": "Year Dropdown",
            selected: selectedYear ? [selectedYear] : [],
            value: formatYear(selectedYear),
            onSelectionChange: handleYearSelect,
            ...YearDropdownProps,
            children: years.map((year) => /* @__PURE__ */ jsxRuntime.jsx(OptionWithTooltip, {
              value: year,
              tooltipContent: "This year is out of range",
              children: formatYear(year)
            }, formatYear(year)))
          })
        ]
      }),
      /* @__PURE__ */ jsxRuntime.jsx(core.Tooltip, {
        placement: "top",
        disabled: canNavigateNext,
        content: "Future dates are out of range",
        enterDelay: 0,
        leaveDelay: 0,
        children: /* @__PURE__ */ jsxRuntime.jsx(core.Button, {
          disabled: !canNavigateNext,
          variant: "secondary",
          onClick: handleNavigateNext,
          focusableWhenDisabled: true,
          children: /* @__PURE__ */ jsxRuntime.jsx(icons.ChevronRightIcon, {
            "aria-label": `Next Month, ${utils.formatDate(
              visibleMonth.add({ months: 1 })
            )}`
          })
        })
      })
    ]
  });
});

exports.CalendarNavigation = CalendarNavigation;
