// @salt-ds/lab/dist-cjs/calendar/internal/CalendarWeekHeader.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var clsx = require('clsx');
var core = require('@salt-ds/core');
var utils = require('./utils.js');
var CalendarWeekHeader$1 = require('./CalendarWeekHeader.css.js');
var window = require('@salt-ds/window');
var styles = require('@salt-ds/styles');

const withBaseName = core.makePrefixer("saltCalendarWeekHeader");
const CalendarWeekHeader = React.forwardRef(function CalendarWeekHeader2({ className, ...rest }, ref) {
  const weekdaysShort = utils.daysForLocale("narrow");
  const weekdaysLong = utils.daysForLocale("long");
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-calendar",
    css: CalendarWeekHeader$1,
    window: targetWindow
  });
  return /* @__PURE__ */ jsxRuntime.jsx("div", {
    "data-testid": "CalendarWeekHeader",
    className: clsx.clsx(withBaseName(), className),
    ref,
    ...rest,
    children: weekdaysShort.map((day, index) => /* @__PURE__ */ jsxRuntime.jsx("small", {
      "aria-hidden": "true",
      className: withBaseName("dayOfWeek"),
      children: day
    }, weekdaysLong[index]))
  });
});

exports.CalendarWeekHeader = CalendarWeekHeader;
