// @salt-ds/lab/dist-cjs/calendar/internal/utils.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var date = require('@internationalized/date');

const localTimezone = date.getLocalTimeZone();
function getCurrentLocale() {
  return navigator.languages[0];
}
function getDateFormatter(options) {
  return new date.DateFormatter(getCurrentLocale(), options);
}
function formatDate(date, options) {
  const formatter = getDateFormatter(options);
  return formatter.format(date.toDate(localTimezone));
}
function getCalender() {
  const calendarIdentifier = getDateFormatter().resolvedOptions().calendar;
  return date.createCalendar(calendarIdentifier);
}
function daysForLocale(weekday = "long") {
  return [...Array(7).keys()].map(
    (day) => formatDate(
      date.startOfWeek(date.today(date.getLocalTimeZone()), getCurrentLocale()).add({
        days: day
      }),
      { weekday }
    )
  );
}
function monthsForLocale(currentYear) {
  const calendar = getCalender();
  return [...Array(calendar.getMonthsInYear(currentYear)).keys()].map(
    (month) => date.startOfYear(currentYear).add({ months: month })
  );
}
function mapDate(currentDate, currentMonth) {
  return {
    date: currentDate,
    dateOfMonth: currentDate.month,
    isCurrentMonth: date.isSameMonth(currentDate, currentMonth)
  };
}
function generateVisibleDays(currentMonth) {
  const totalDays = 6 * 7;
  const currentLocale = getCurrentLocale();
  const startDate = date.startOfWeek(date.startOfMonth(currentMonth), currentLocale);
  return [...Array(totalDays).keys()].map((dayDelta) => {
    const day = startDate.add({ days: dayDelta });
    return mapDate(day, currentMonth);
  });
}
function monthDiff(a, b) {
  return b.month - a.month + 12 * (b.year - a.year);
}

exports.daysForLocale = daysForLocale;
exports.formatDate = formatDate;
exports.generateVisibleDays = generateVisibleDays;
exports.getCalender = getCalender;
exports.getCurrentLocale = getCurrentLocale;
exports.getDateFormatter = getDateFormatter;
exports.monthDiff = monthDiff;
exports.monthsForLocale = monthsForLocale;
