// @salt-ds/lab/dist-cjs/skip-link/SkipLink.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var core = require('@salt-ds/core');
var clsx = require('clsx');
var useManageFocusOnTarget = require('./internal/useManageFocusOnTarget.js');
var window = require('@salt-ds/window');
var styles = require('@salt-ds/styles');
var SkipLink$1 = require('./SkipLink.css.js');

const withBaseName = core.makePrefixer("saltSkipLink");
const SkipLink = React.forwardRef(
  function SkipLink2({ className, targetRef, ...rest }, ref) {
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-skip-link",
      css: SkipLink$1,
      window: targetWindow
    });
    const targetClass = clsx.clsx(withBaseName("target"), className);
    const eventHandlers = useManageFocusOnTarget.useManageFocusOnTarget({ targetRef, targetClass });
    return /* @__PURE__ */ jsxRuntime.jsx("li", {
      children: /* @__PURE__ */ jsxRuntime.jsx(core.Link, {
        ...eventHandlers,
        ...rest,
        className: clsx.clsx(withBaseName(), className),
        ref
      })
    });
  }
);

exports.SkipLink = SkipLink;
