// @salt-ds/lab/dist-es/combo-box-deprecated/internal/DefaultComboBox.js
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import { flip, shift, limitShift, size } from '@floating-ui/react';
import { useForkRef, useAriaAnnouncer, useFloatingUI } from '@salt-ds/core';
import { useRef, useEffect, useState } from 'react';
import { Portal } from '../../portal/Portal.js';
import '../../list-deprecated/List.js';
import { ListBase } from '../../list-deprecated/ListBase.js';
import '../../list-deprecated/ListItemBase.js';
import '../../list-deprecated/ListItem.js';
import '../../list-deprecated/ListItemContext.js';
import { ListStateContext } from '../../list-deprecated/ListStateContext.js';
import { getAnnouncement } from './getAnnouncement.js';
import { useComboBox } from './useComboBox.js';
import { isDesktop, Window } from '../../window/WindowContext.js';
import '../../window/ElectronWindow.js';
import { InputLegacy } from '../../input-legacy/InputLegacy.js';
import '../../input-legacy/StaticInputAdornment.js';

const DefaultComboBox = function DefaultComboBox2(props) {
  const {
    ListItem,
    WindowProps: WindowProps2 = {},
    rootRef,
    listRef: listRefProp,
    inputRef: inputRefProp,
    rootWidth,
    listWidth,
    ...restProps
  } = props;
  const inputRef = useRef(null);
  const listRef = useRef(null);
  const setInputRef = useForkRef(inputRef, inputRefProp);
  const setListRef = useForkRef(listRef, listRefProp);
  const { announce } = useAriaAnnouncer({ debounce: 1e3 });
  const {
    inputRef: setHookInputRef,
    listContext,
    inputProps,
    listProps
  } = useComboBox(restProps);
  const { allowAnnouncement, disabled, value, ...restInputProps } = inputProps;
  const { isListOpen, itemCount, itemToString, source, ...restListProps } = listProps;
  const firstItem = null;
  const allowAnnouncementRef = useRef(allowAnnouncement);
  useEffect(() => {
    allowAnnouncementRef.current = allowAnnouncement;
  }, [allowAnnouncement]);
  useEffect(() => {
    if (allowAnnouncementRef.current && value && firstItem) {
      announce(getAnnouncement(itemCount, firstItem));
    }
  }, [firstItem, value, itemCount, announce]);
  const [maxListHeight, setMaxListHeight] = useState(
    void 0
  );
  const middleware = isDesktop ? [] : [
    flip({
      fallbackPlacements: ["bottom-start", "top-start"]
    }),
    shift({ limiter: limitShift() }),
    size({
      apply({ availableHeight }) {
        setMaxListHeight(availableHeight);
      }
    })
  ];
  const { reference, floating, x, y, strategy } = useFloatingUI({
    placement: "bottom-start",
    middleware
  });
  useEffect(() => {
    if (rootRef.current) {
      reference(rootRef.current);
    }
  }, [rootRef, reference]);
  return /* @__PURE__ */ jsxs(Fragment, {
    children: [
      /* @__PURE__ */ jsx(InputLegacy, {
        disabled,
        ref: useForkRef(setInputRef, setHookInputRef),
        value,
        ...restInputProps
      }),
      rootRef.current && isListOpen && /* @__PURE__ */ jsx(Portal, {
        children: /* @__PURE__ */ jsx(Window, {
          style: {
            top: y != null ? y : 0,
            left: x != null ? x : 0,
            position: strategy,
            maxHeight: maxListHeight != null ? maxListHeight : ""
          },
          ...WindowProps2,
          ref: floating,
          children: /* @__PURE__ */ jsx(ListStateContext.Provider, {
            value: listContext,
            children: /* @__PURE__ */ jsx(ListBase, {
              ...{
                ListItem,
                disabled,
                itemCount,
                itemToString,
                width: listWidth || rootWidth,
                source,
                ...restListProps,
                listRef: setListRef
              },
              maxHeight: maxListHeight || listProps.maxHeight
            })
          })
        })
      })
    ]
  });
};

export { DefaultComboBox };
