// @salt-ds/lab/dist-es/breadcrumbs/Breadcrumbs.js
import { jsx } from 'react/jsx-runtime';
import { makePrefixer } from '@salt-ds/core';
import { clsx } from 'clsx';
import React, { useMemo, isValidElement } from 'react';
import { BreadcrumbsCollapsed } from './internal/BreadcrumbsCollapsed.js';
import { BreadcrumbsContext } from './internal/BreadcrumbsContext.js';
import { BreadcrumbsSeparator } from './internal/BreadcrumbsSeparator.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './Breadcrumbs.css.js';

const iconWidth = 12;
const withBaseName = makePrefixer("saltBreadcrumbs");
function insertSeparators(items, className, separator, hideCurrentLevel, shouldRenderAllItems, itemsAfterCollapse) {
  return items.reduce((acc, current, index) => {
    if (index < items.length - 1) {
      return acc.concat(
        current,
        /* @__PURE__ */ jsx("li", {
          "aria-hidden": true,
          className,
          children: separator
        }, `separator-${index}`)
      );
    } else {
      const skipLastElement = hideCurrentLevel && shouldRenderAllItems || hideCurrentLevel && !shouldRenderAllItems && itemsAfterCollapse !== 0;
      return skipLastElement ? acc : acc.concat(current);
    }
  }, []);
}
const Breadcrumbs = React.forwardRef(
  function Breadcrumbs2(props, ref) {
    const {
      children,
      className,
      itemsAfterCollapse = 1,
      itemsBeforeCollapse = 1,
      maxItems = 3,
      separator: separatorProp,
      hideCurrentLevel,
      itemsMinWidth: itemsMinWidthProp,
      itemsMaxWidth,
      wrap,
      Menu,
      SeparatorProps = {},
      ...other
    } = props;
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-breadcrumbs",
      css: css_248z,
      window: targetWindow
    });
    const itemsMinWidth = itemsMinWidthProp != null ? itemsMinWidthProp : iconWidth;
    const breadcrumbsContext = useMemo(
      () => ({
        wrap,
        itemsMinWidth,
        itemsMaxWidth,
        liClass: withBaseName("li")
      }),
      [wrap, itemsMaxWidth, itemsMinWidth]
    );
    const separator = separatorProp || /* @__PURE__ */ jsx(BreadcrumbsSeparator, {
      ...SeparatorProps
    });
    const childrenArray = React.Children.toArray(children);
    const shouldRenderAllItems = wrap || maxItems == null || childrenArray.length <= maxItems;
    const allItems = childrenArray.filter(isValidElement).map((child, index) => {
      const isLastChild = index === childrenArray.length - 1;
      return React.cloneElement(child, {
        isCurrentLevel: isLastChild
      });
    });
    const renderItemsBeforeAndAfter = () => {
      if (itemsBeforeCollapse + itemsAfterCollapse >= allItems.length) {
        console.warn(
          [
            "You have provided an invalid combination of properties to the Breadcrumbs.",
            `itemsAfterCollapse={${itemsAfterCollapse}} +itemsBeforeCollapse={${itemsBeforeCollapse}} >= maxItems={${maxItems}}`
          ].join("\n")
        );
        return allItems;
      }
      const hiddenItems = allItems.slice(
        itemsBeforeCollapse,
        allItems.length - itemsAfterCollapse
      );
      return [
        ...allItems.slice(0, itemsBeforeCollapse),
        /* @__PURE__ */ jsx("li", {
          className: withBaseName("li"),
          children: /* @__PURE__ */ jsx(BreadcrumbsCollapsed, {
            accessibleText: `Breadcrumb levels ${itemsBeforeCollapse + 1} to ${allItems.length - itemsAfterCollapse}`,
            className: withBaseName("overflowButton"),
            children: hiddenItems
          })
        }, "breadcrumbs-collapsed"),
        ...allItems.slice(
          allItems.length - itemsAfterCollapse,
          allItems.length
        )
      ];
    };
    const itemsToRender = shouldRenderAllItems ? allItems : renderItemsBeforeAndAfter();
    return /* @__PURE__ */ jsx(BreadcrumbsContext.Provider, {
      value: breadcrumbsContext,
      children: /* @__PURE__ */ jsx("nav", {
        "aria-label": "Breadcrumb",
        className: clsx(withBaseName(), className),
        "data-testid": "breadcrumbs",
        ref,
        ...other,
        children: /* @__PURE__ */ jsx("ol", {
          className: clsx(withBaseName("ol"), {
            [withBaseName("ol-wrap")]: wrap
          }),
          children: insertSeparators(
            itemsToRender,
            withBaseName("separator"),
            separator,
            hideCurrentLevel,
            shouldRenderAllItems,
            itemsAfterCollapse
          )
        })
      })
    });
  }
);

export { Breadcrumbs };
