// @salt-ds/lab/dist-es/slider/internal/SliderHandle.js
import { jsx } from 'react/jsx-runtime';
import { makePrefixer, Tooltip } from '@salt-ds/core';
import { clsx } from 'clsx';
import { getSliderAriaLabel } from './utils.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from '../Slider.css.js';

const withBaseName = makePrefixer("saltSliderHandle");
function SliderHandle(props) {
  const { min, max, value, disabled, valueLength, index, style } = props;
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-slider",
    css: css_248z,
    window: targetWindow
  });
  return /* @__PURE__ */ jsx(Tooltip, {
    content: value,
    placement: "top",
    children: /* @__PURE__ */ jsx("div", {
      className: clsx(withBaseName(), {
        [withBaseName("min")]: value === min,
        [withBaseName("max")]: value === max
      }),
      style,
      role: "slider",
      "aria-valuemin": min,
      "aria-valuemax": max,
      "aria-valuenow": value,
      "aria-disabled": disabled,
      "aria-label": getSliderAriaLabel(valueLength, index),
      tabIndex: 0,
      "data-handle-index": index
    })
  });
}

export { SliderHandle };
