// @salt-ds/lab/dist-es/slider/internal/SliderRailMarks.js
import { jsx } from 'react/jsx-runtime';
import { useMemo } from 'react';
import { makePrefixer } from '@salt-ds/core';
import { clsx } from 'clsx';
import { createSliderRailMarksStyle, createHandleStyles } from './styles.js';
import { isMarkAtMax } from './utils.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from '../Slider.css.js';

const withBaseName = makePrefixer("saltSliderRailMarks");
function SliderRailMarks(props) {
  const { min, max, marks } = props;
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-slider",
    css: css_248z,
    window: targetWindow
  });
  const style = useMemo(
    () => createSliderRailMarksStyle(min, max, marks),
    [min, max, marks]
  );
  const marksLength = marks.length;
  const markStyles = useMemo(
    () => createHandleStyles(marksLength),
    [marksLength]
  );
  return /* @__PURE__ */ jsx("div", {
    className: withBaseName(),
    style,
    children: marks.map((mark, i) => {
      return /* @__PURE__ */ jsx("div", {
        className: clsx(withBaseName("mark"), {
          [withBaseName("max")]: isMarkAtMax(max, mark)
        }),
        style: markStyles[i]
      }, `Mark${i}`);
    })
  });
}

export { SliderRailMarks };
