// @salt-ds/lab/dist-es/calendar/internal/CalendarWeekHeader.js
import { jsx } from 'react/jsx-runtime';
import { forwardRef } from 'react';
import { clsx } from 'clsx';
import { makePrefixer } from '@salt-ds/core';
import { daysForLocale } from './utils.js';
import css_248z from './CalendarWeekHeader.css.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';

const withBaseName = makePrefixer("saltCalendarWeekHeader");
const CalendarWeekHeader = forwardRef(function CalendarWeekHeader2({ className, ...rest }, ref) {
  const weekdaysShort = daysForLocale("narrow");
  const weekdaysLong = daysForLocale("long");
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-calendar",
    css: css_248z,
    window: targetWindow
  });
  return /* @__PURE__ */ jsx("div", {
    "data-testid": "CalendarWeekHeader",
    className: clsx(withBaseName(), className),
    ref,
    ...rest,
    children: weekdaysShort.map((day, index) => /* @__PURE__ */ jsx("small", {
      "aria-hidden": "true",
      className: withBaseName("dayOfWeek"),
      children: day
    }, weekdaysLong[index]))
  });
});

export { CalendarWeekHeader };
