// @salt-ds/lab/dist-es/app-header/AppHeader.js
import { jsx } from 'react/jsx-runtime';
import { clsx } from 'clsx';
import { Toolbar } from '../toolbar/Toolbar.js';
import { makePrefixer } from '@salt-ds/core';
import { forwardRef } from 'react';
import '../toolbar/overflow-panel/OverflowPanel.js';
import '../responsive/useResizeObserver.js';
import '../toolbar/toolbar-field/ToolbarField.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import '../toolbar/ToolbarButton.js';
import { MenuIcon } from '@salt-ds/icons';
import css_248z from './AppHeader.css.js';

const withBaseName = makePrefixer("saltAppHeader");
const AppHeader = forwardRef(
  ({ children, className, ...rest }, ref) => {
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-app-header",
      css: css_248z,
      window: targetWindow
    });
    return /* @__PURE__ */ jsx(Toolbar, {
      className: clsx(withBaseName(), className),
      ref,
      overflowButtonIcon: /* @__PURE__ */ jsx(MenuIcon, {}),
      children
    });
  }
);

export { AppHeader };
