// @salt-ds/lab/dist-es/toolbar/Tooltray.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { useIdMemo } from '@salt-ds/core';
import { clsx } from 'clsx';
import { cloneElement } from 'react';
import { OverflowPanel } from './overflow-panel/OverflowPanel.js';
import { useOverflowCollectionItems } from '../responsive/useOverflowCollectionItems.js';
import { useOverflowLayout } from '../responsive/useOverflowLayout.js';
import { renderTrayTools } from './internal/renderTrayTools.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './Tooltray.css.js';

const Tooltray = (props) => {
  const {
    "aria-label": ariaLabel,
    alignEnd,
    alignStart,
    children,
    className: classNameProp,
    collapse: collapseProp,
    collapsed: collapsedProp,
    "data-collapsible": collapse = collapseProp,
    "data-collapsed": collapsed = collapsedProp,
    id: idProp,
    isInsidePanel = false,
    overflowButtonIcon,
    overflowButtonLabel,
    orientation = "horizontal",
    ...rest
  } = props;
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-tooltray",
    css: css_248z,
    window: targetWindow
  });
  const tooltrayId = useIdMemo(idProp);
  const className = clsx(
    "saltTooltray",
    classNameProp,
    `saltTooltray-${orientation}`,
    { "saltTooltray-tooltrayOverflowed": isInsidePanel }
  );
  const collectionHook = useOverflowCollectionItems({
    children,
    id: tooltrayId,
    label: "Tooltray",
    orientation
  });
  const [innerContainerRef] = useOverflowLayout({
    collectionHook,
    id: tooltrayId,
    orientation,
    label: "Tooltray"
  });
  const overflowedItems = collapsed ? collectionHook.data.filter((item) => !item.isOverflowIndicator) : collectionHook.data.filter((item) => item.overflowed);
  const overflowMenuItems = overflowedItems.map(
    (i) => cloneElement(collectionHook.data[i.index].element, {
      "data-is-inside-panel": true,
      key: i.index
    })
  ).reverse();
  const overflowIndicator = collectionHook.data.find(
    (i) => i.isOverflowIndicator
  );
  const tooltrayProps = {
    className,
    "data-collapsed": collapsed,
    "data-collapsible": collapse
  };
  return /* @__PURE__ */ jsx("div", {
    ...rest,
    ...tooltrayProps,
    id: tooltrayId,
    children: /* @__PURE__ */ jsxs("div", {
      className: clsx("Responsive-inner"),
      ref: innerContainerRef,
      children: [
        renderTrayTools(
          collectionHook,
          overflowedItems,
          orientation,
          collapsed
        ),
        overflowIndicator || collapsed ? /* @__PURE__ */ jsx(OverflowPanel, {
          className: clsx("saltToolbarField"),
          "data-index": collectionHook.data.length,
          "data-overflow-indicator": true,
          "data-priority": 1,
          id: overflowIndicator == null ? void 0 : overflowIndicator.id,
          triggerButtonIcon: overflowButtonIcon,
          triggerButtonLabel: overflowButtonLabel,
          children: overflowMenuItems
        }) : null
      ]
    })
  });
};

export { Tooltray };
