// @salt-ds/lab/dist-es/slider/Slider.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { makePrefixer, useControlled } from '@salt-ds/core';
import { clsx } from 'clsx';
import { forwardRef, useRef, useMemo } from 'react';
import { SliderHandle } from './internal/SliderHandle.js';
import { SliderMarkLabels } from './internal/SliderMarkLabels.js';
import { SliderRail } from './internal/SliderRail.js';
import { SliderRailMarks } from './internal/SliderRailMarks.js';
import { SliderSelection } from './internal/SliderSelection.js';
import { createTrackStyle, createHandleStyles } from './internal/styles.js';
import { useSliderKeyDown } from './internal/useSliderKeyDown.js';
import { useSliderMouseDown } from './internal/useSliderMouseDown.js';
import { useValueUpdater } from './internal/utils.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './Slider.css.js';

const withBaseName = makePrefixer("saltSlider");
const defaultMin = 0;
const defaultMax = 100;
const defaultStep = 1;
const Slider = forwardRef(function Slider2({
  min = defaultMin,
  max = defaultMax,
  step = defaultStep,
  pageStep = step,
  value: valueProp,
  defaultValue = defaultMin,
  onChange,
  label,
  className,
  pushable,
  pushDistance = 0,
  disabled,
  marks,
  hideMarks,
  hideMarkLabels,
  ...restProps
}, ref) {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-slider",
    css: css_248z,
    window: targetWindow
  });
  const trackRef = useRef(null);
  const [value, setValue] = useControlled({
    controlled: valueProp,
    default: defaultValue,
    name: "Slider",
    state: "Value"
  });
  const updateValueItem = useValueUpdater(pushable, pushDistance, min, max);
  const trackStyle = useMemo(
    () => createTrackStyle(min, max, value),
    [min, max, value]
  );
  const valueLength = Array.isArray(value) ? value.length : 1;
  const handleStyles = useMemo(
    () => createHandleStyles(valueLength),
    [valueLength]
  );
  const onMouseDown = useSliderMouseDown(
    trackRef,
    value,
    min,
    max,
    step,
    updateValueItem,
    setValue,
    onChange
  );
  const onKeyDown = useSliderKeyDown(
    value,
    min,
    max,
    pageStep,
    step,
    updateValueItem,
    setValue,
    onChange
  );
  return /* @__PURE__ */ jsxs("div", {
    className: clsx(
      withBaseName(),
      {
        [withBaseName("disabled")]: disabled
      },
      className
    ),
    ref,
    onKeyDown: disabled ? void 0 : onKeyDown,
    "aria-label": `${label} slider from ${min} to ${max}`,
    role: "group",
    children: [
      label !== void 0 ? /* @__PURE__ */ jsx("div", {
        className: withBaseName("label"),
        children: label
      }) : null,
      /* @__PURE__ */ jsx("div", {
        className: withBaseName("clickable"),
        onMouseDown: disabled ? void 0 : onMouseDown,
        children: /* @__PURE__ */ jsxs("div", {
          className: withBaseName("track"),
          style: trackStyle,
          ref: trackRef,
          children: [
            /* @__PURE__ */ jsx(SliderRail, {}),
            marks && !hideMarks ? /* @__PURE__ */ jsx(SliderRailMarks, {
              min,
              max,
              marks
            }) : null,
            marks && !hideMarkLabels ? /* @__PURE__ */ jsx(SliderMarkLabels, {
              min,
              max,
              marks
            }) : null,
            /* @__PURE__ */ jsx(SliderSelection, {
              valueLength
            }),
            (Array.isArray(value) ? value : [value]).map((v, i) => /* @__PURE__ */ jsx(SliderHandle, {
              min,
              max,
              value: v,
              index: i,
              disabled: !!disabled,
              valueLength,
              style: handleStyles[i]
            }, i))
          ]
        })
      })
    ]
  });
});

export { Slider };
