// @salt-ds/lab/dist-es/query-input/internal/CategoryList.js
import { jsx } from 'react/jsx-runtime';
import { useRef, useState, useCallback } from 'react';
import { makePrefixer, useIsomorphicLayoutEffect } from '@salt-ds/core';
import '../../list/Highlighter.js';
import '../../list/ListItem.js';
import { List } from '../../list/List.js';
import '../../common-hooks/collectionProvider.js';
import '../../common-hooks/keyUtils.js';
import '../../responsive/useResizeObserver.js';
import '../../list/VirtualizedList.js';
import { CategoryListItem } from './CategoryListItem.js';
import { CategoryListContext } from './CategoryListContext.js';

const withBaseName = makePrefixer("saltCategoryList");
const getCategoryLabel = (category) => category.name;
function CategoryList(props) {
  const {
    categories,
    rootWidth,
    onCategorySelect,
    highlightedCategoryIndex,
    setHighlightedCategoryIndex
  } = props;
  const measureRef = useRef(null);
  const [measuredCategories, setMeasuredCategories] = useState([]);
  const [contextValue, setContextValue] = useState({
    width: 0
  });
  useIsomorphicLayoutEffect(() => {
    setTimeout(() => {
      if (measureRef.current) {
        const width = measureRef.current.offsetWidth;
        setMeasuredCategories(categories);
        setContextValue({
          width
        });
      }
    }, 0);
  }, [categories, setMeasuredCategories, setContextValue]);
  const onSelect = useCallback(
    (_, item) => {
      onCategorySelect(item);
    },
    [onCategorySelect]
  );
  const isMeasuring = measuredCategories !== categories;
  if (isMeasuring) {
    return /* @__PURE__ */ jsx("div", {
      ref: measureRef,
      className: withBaseName("categoryWidthMeasure"),
      children: categories.map((c, i) => {
        return /* @__PURE__ */ jsx("div", {
          className: withBaseName("category"),
          children: c.name
        }, i);
      })
    });
  }
  return /* @__PURE__ */ jsx(CategoryListContext.Provider, {
    value: contextValue,
    children: /* @__PURE__ */ jsx(List, {
      ListItem: CategoryListItem,
      "data-testid": "category-list",
      highlightedIndex: highlightedCategoryIndex,
      itemToString: getCategoryLabel,
      onHighlight: setHighlightedCategoryIndex,
      onSelect,
      source: categories,
      width: rootWidth
    })
  });
}

export { CategoryList };
