// @salt-ds/lab/dist-es/query-input/internal/QueryInputBody.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { makePrefixer, ToggleButtonGroup, ToggleButton } from '@salt-ds/core';
import { FilterIcon } from '@salt-ds/icons';
import { forwardRef, useMemo } from 'react';
import { TokenizedInput } from '../../tokenized-input/TokenizedInput.js';
import '../../tokenized-input/TokenizedInputBase.js';
import 'clipboard-copy';
import '../../form-field-context-legacy/FormFieldLegacyContext.js';

const withBaseName = makePrefixer("saltQueryInputBody");
const QueryInputBody = forwardRef(
  function QueryInputBody2(props, ref) {
    const {
      inputRef,
      disabled,
      selectedItems,
      onInputFocus,
      onInputBlur,
      showCategory,
      inputValue,
      onInputChange,
      isFocused,
      onInputClear,
      onSelectedItemsChange,
      onKeyDown,
      onInputClick,
      booleanOperator,
      onBooleanOperatorChange
    } = props;
    const itemToString = useMemo(() => {
      if (showCategory) {
        return (item) => [item.category, item.value].join(": ");
      }
      return (item) => item.value;
    }, [showCategory]);
    const stringToItem = (selectedItems2, value) => {
      return {
        category: null,
        value
      };
    };
    const onChange = (event) => {
      const newBooleanOperator = event.currentTarget.value;
      onBooleanOperatorChange(newBooleanOperator);
    };
    return /* @__PURE__ */ jsxs("div", {
      ref,
      className: withBaseName(),
      children: [
        /* @__PURE__ */ jsx(FilterIcon, {}),
        /* @__PURE__ */ jsx(TokenizedInput, {
          inputRef,
          disabled,
          className: withBaseName("input"),
          selectedItems,
          itemToString,
          stringToItem,
          onInputFocus,
          onInputBlur,
          value: inputValue,
          onClick: onInputClick,
          onInputChange,
          expanded: isFocused,
          onClear: onInputClear,
          onChange: onSelectedItemsChange,
          onKeyDown
        }),
        /* @__PURE__ */ jsx("div", {
          className: withBaseName("separator")
        }),
        /* @__PURE__ */ jsxs(ToggleButtonGroup, {
          className: withBaseName("buttonGroup"),
          "data-testid": "boolean-selector",
          value: booleanOperator,
          onChange,
          children: [
            /* @__PURE__ */ jsx(ToggleButton, {
              value: "and",
              children: "And"
            }),
            /* @__PURE__ */ jsx(ToggleButton, {
              value: "or",
              children: "Or"
            })
          ]
        })
      ]
    });
  }
);

export { QueryInputBody };
