// @salt-ds/lab/dist-es/query-input/internal/ValueList.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { makePrefixer } from '@salt-ds/core';
import { ChevronLeftIcon } from '@salt-ds/icons';
import '../../list/Highlighter.js';
import { ListItem } from '../../list/ListItem.js';
import { List } from '../../list/List.js';
import 'react';
import '../../common-hooks/collectionProvider.js';
import '../../common-hooks/keyUtils.js';
import '../../responsive/useResizeObserver.js';
import '../../list/VirtualizedList.js';

const withBaseName = makePrefixer("saltQueryInputValueList");
function ValueList(props) {
  const {
    category,
    rootWidth,
    onBack,
    selectedValues,
    onValueToggle,
    highlightedValueIndex,
    setHighlightedValueIndex
  } = props;
  const handleSelect = (_, value) => {
    onValueToggle(category, value);
  };
  const items = category ? category.values.map((value) => /* @__PURE__ */ jsx(ListItem, {
    item: value,
    children: value
  }, value)) : [];
  return /* @__PURE__ */ jsx("div", {
    className: withBaseName(),
    tabIndex: 0,
    "data-testid": "value-list",
    children: /* @__PURE__ */ jsxs(List, {
      borderless: true,
      checkable: false,
      highlightedIndex: highlightedValueIndex,
      onHighlight: setHighlightedValueIndex,
      onSelect: handleSelect,
      selectionStrategy: "multiple",
      selected: selectedValues,
      width: rootWidth,
      children: [
        /* @__PURE__ */ jsxs(ListItem, {
          className: withBaseName("back"),
          onClick: onBack,
          selectable: false,
          children: [
            /* @__PURE__ */ jsx(ChevronLeftIcon, {}),
            /* @__PURE__ */ jsx("div", {
              className: withBaseName("category"),
              children: category ? category.name : ""
            })
          ]
        }),
        items
      ]
    })
  });
}

export { ValueList };
