// @salt-ds/lab/dist-es/calendar/Calendar.js
import { jsx, jsxs } from 'react/jsx-runtime';
import { forwardRef, useCallback } from 'react';
import { clsx } from 'clsx';
import { makePrefixer } from '@salt-ds/core';
import { CalendarNavigation } from './internal/CalendarNavigation.js';
import { CalendarWeekHeader } from './internal/CalendarWeekHeader.js';
import { CalendarCarousel } from './internal/CalendarCarousel.js';
import { CalendarContext } from './internal/CalendarContext.js';
import { useCalendar } from './useCalendar.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './Calendar.css.js';

const withBaseName = makePrefixer("saltCalendar");
const Calendar = forwardRef(
  function Calendar2(props, ref) {
    const {
      className,
      renderDayContents,
      hideYearDropdown,
      TooltipProps,
      ...rest
    } = props;
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-calendar",
      css: css_248z,
      window: targetWindow
    });
    const { state, helpers } = useCalendar({ hideYearDropdown, ...rest });
    const { setCalendarFocused } = helpers;
    const handleFocus = useCallback(() => {
      setCalendarFocused(true);
    }, [setCalendarFocused]);
    const handleBlur = useCallback(() => {
      setCalendarFocused(false);
    }, [setCalendarFocused]);
    return /* @__PURE__ */ jsx(CalendarContext.Provider, {
      value: {
        state,
        helpers
      },
      children: /* @__PURE__ */ jsxs("div", {
        className: clsx(withBaseName(), className),
        role: "application",
        ref,
        children: [
          /* @__PURE__ */ jsx(CalendarNavigation, {
            hideYearDropdown
          }),
          /* @__PURE__ */ jsx(CalendarWeekHeader, {}),
          /* @__PURE__ */ jsx(CalendarCarousel, {
            onFocus: handleFocus,
            onBlur: handleBlur,
            renderDayContents,
            TooltipProps
          })
        ]
      })
    });
  }
);

export { Calendar };
