// @salt-ds/lab/dist-cjs/form-field-legacy/FormHelperText.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var window = require('@salt-ds/window');
var styles = require('@salt-ds/styles');
var FormHelperText$1 = require('./FormHelperText.css.js');

const FormHelperText = ({
  helperText,
  helperTextPlacement,
  ...restProps
}) => {
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-form-helper-text",
    css: FormHelperText$1,
    window: targetWindow
  });
  if (helperText) {
    if (helperTextPlacement === "bottom") {
      return /* @__PURE__ */ jsxRuntime.jsx("p", {
        className: "saltFormHelperText",
        ...restProps,
        children: helperText
      });
    } else if (helperTextPlacement === "tooltip") {
      console.warn("helperTextPlacement tooltip has not yet implemented");
      return null;
    } else {
      return null;
    }
  } else {
    return null;
  }
};

exports.FormHelperText = FormHelperText;
