// @salt-ds/lab/dist-cjs/form-field-legacy/FormFieldLegacy.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var clsx = require('clsx');
var React = require('react');
var core = require('@salt-ds/core');
var FormFieldLegacyContext = require('../form-field-context-legacy/FormFieldLegacyContext.js');
var constant = require('./constant.js');
var FormActivationIndicator = require('./FormActivationIndicator.js');
var FormHelperText = require('./FormHelperText.js');
var FormLabel = require('./FormLabel.js');
var window = require('@salt-ds/window');
var styles = require('@salt-ds/styles');
var FormFieldLegacy$1 = require('./FormFieldLegacy.css.js');

const useA11yValue = ({
  required,
  disabled,
  readOnly,
  labelId,
  helperTextId,
  renderHelperText
}) => {
  return React.useMemo(
    () => ({
      "aria-labelledby": labelId,
      "aria-required": required,
      "aria-describedby": renderHelperText ? helperTextId : void 0,
      disabled,
      readOnly
    }),
    [labelId, disabled, readOnly, required, renderHelperText, helperTextId]
  );
};
const useFormFieldLegacy = ({
  onBlur,
  onFocus
}) => {
  const [focused, setFocused] = React.useState(false);
  const handleBlur = (event) => {
    setFocused(false);
    onBlur == null ? void 0 : onBlur(event);
  };
  const handleFocus = (event) => {
    setFocused(true);
    onFocus == null ? void 0 : onFocus(event);
  };
  return [
    {
      focused
    },
    {
      setFocused
    },
    {
      onBlur: handleBlur,
      onFocus: handleFocus
    }
  ];
};
const withBaseName = core.makePrefixer(constant.classBase);
const FormFieldLegacy = React.forwardRef(
  ({
    ActivationIndicatorComponent = FormActivationIndicator.FormActivationIndicator,
    children,
    className,
    disabled,
    disableFocusRing = false,
    fullWidth = true,
    hasStatusIndicator,
    HelperTextComponent = FormHelperText.FormHelperText,
    HelperTextProps,
    helperText,
    helperTextPlacement = "bottom",
    label,
    LabelComponent = FormLabel.FormLabel,
    labelPlacement = "top",
    LabelProps = { displayedNecessity: "required" },
    onBlur,
    onFocus,
    readOnly,
    required,
    StatusIndicatorProps: StatusIndicatorProps2,
    validationStatus,
    variant = "primary",
    ...restProps
  }, ref) => {
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-form-field-legacy",
      css: FormFieldLegacy$1,
      window: targetWindow
    });
    const labelId = core.useId(LabelProps == null ? void 0 : LabelProps.id);
    const helperTextId = core.useId(HelperTextProps == null ? void 0 : HelperTextProps.id);
    const rootRef = React.useRef(null);
    const renderHelperText = !!helperText;
    const a11yValue = useA11yValue({
      required,
      disabled,
      readOnly,
      labelId,
      helperTextId,
      renderHelperText
    });
    const [states, dispatchers, eventHandlers] = useFormFieldLegacy({
      onBlur,
      onFocus
    });
    const hasLabel = label !== void 0;
    const labelTop = hasLabel && labelPlacement === "top";
    const labelLeft = hasLabel && labelPlacement === "left";
    const isWarning = validationStatus === "warning";
    const isError = validationStatus === "error";
    const focusClass = disableFocusRing ? "lowFocused" : "focused";
    const inlineHelperText = renderHelperText && helperTextPlacement === "bottom";
    const tooltipHelperText = renderHelperText && helperTextPlacement === "tooltip" && !hasStatusIndicator;
    const handleTriggerRef = core.useForkRef(rootRef, ref);
    return /* @__PURE__ */ jsxRuntime.jsx(core.Tooltip, {
      disabled: !tooltipHelperText,
      content: helperText,
      children: /* @__PURE__ */ jsxRuntime.jsx("div", {
        ref: handleTriggerRef,
        className: clsx.clsx(
          withBaseName(),
          {
            [withBaseName("disabled")]: disabled,
            [withBaseName("readOnly")]: readOnly,
            [withBaseName("warning")]: isWarning,
            [withBaseName("error")]: isError,
            [withBaseName("fullWidth")]: fullWidth,
            [withBaseName(focusClass)]: states.focused,
            [withBaseName("labelTop")]: labelTop,
            [withBaseName("labelLeft")]: labelLeft,
            [withBaseName(`withHelperText`)]: inlineHelperText,
            [withBaseName(variant)]: variant
          },
          className
        ),
        ...eventHandlers,
        ...restProps,
        children: /* @__PURE__ */ jsxRuntime.jsxs(FormFieldLegacyContext.FormFieldLegacyContext.Provider, {
          value: {
            ...states,
            ...dispatchers,
            ...eventHandlers,
            a11yProps: a11yValue,
            inFormField: true,
            ref: rootRef
          },
          children: [
            hasLabel && /* @__PURE__ */ jsxRuntime.jsx(LabelComponent, {
              ...LabelProps,
              validationStatus,
              hasStatusIndicator,
              StatusIndicatorProps: StatusIndicatorProps2,
              className: LabelProps.className,
              label,
              disabled,
              readOnly,
              required,
              tooltipText: helperText,
              id: labelId
            }),
            children,
            /* @__PURE__ */ jsxRuntime.jsx(ActivationIndicatorComponent, {
              hasIcon: !hasStatusIndicator,
              validationStatus
            }),
            renderHelperText && /* @__PURE__ */ jsxRuntime.jsx(HelperTextComponent, {
              helperText,
              helperTextPlacement,
              ...HelperTextProps,
              id: helperTextId
            })
          ]
        })
      })
    });
  }
);

exports.FormFieldLegacy = FormFieldLegacy;
exports.useFormFieldLegacy = useFormFieldLegacy;
