// @salt-ds/lab/dist-cjs/form-field-legacy/FormActivationIndicator.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var window = require('@salt-ds/window');
var styles = require('@salt-ds/styles');
var FormActivationIndicator$1 = require('./FormActivationIndicator.css.js');

const ErrorIndicatorIcon = (props) => {
  return /* @__PURE__ */ jsxRuntime.jsx("svg", {
    viewBox: "0 0 16 16",
    focusable: false,
    "data-testid": "ErrorIndicatorIcon",
    ...props,
    children: /* @__PURE__ */ jsxRuntime.jsx("circle", {
      cx: 8,
      cy: 8,
      r: 8
    })
  });
};
const WarningIndicatorIcon = (props) => {
  return /* @__PURE__ */ jsxRuntime.jsx("svg", {
    viewBox: "0 0 16 16",
    focusable: false,
    "data-testid": "WarningIndicatorIcon",
    ...props,
    children: /* @__PURE__ */ jsxRuntime.jsx("polygon", {
      points: "0, 16 16, 16 16, 0"
    })
  });
};
const withBaseName = core.makePrefixer("saltFormActivationIndicator");
const ActivationIndicatorIcon = ({
  validationStatus,
  ...restSvgProps
}) => {
  if (validationStatus === "error") {
    return /* @__PURE__ */ jsxRuntime.jsx(ErrorIndicatorIcon, {
      ...restSvgProps
    });
  } else if (validationStatus === "warning") {
    return /* @__PURE__ */ jsxRuntime.jsx(WarningIndicatorIcon, {
      ...restSvgProps
    });
  } else {
    return null;
  }
};
const FormActivationIndicator = ({
  hasIcon,
  validationStatus
}) => {
  const targetWindow = window.useWindow();
  styles.useComponentCssInjection({
    testId: "salt-form-activation-indicator",
    css: FormActivationIndicator$1,
    window: targetWindow
  });
  const rootClass = "saltFormActivationIndicator";
  return /* @__PURE__ */ jsxRuntime.jsx("div", {
    className: withBaseName(),
    children: /* @__PURE__ */ jsxRuntime.jsx("div", {
      className: withBaseName("bar"),
      children: hasIcon && validationStatus && /* @__PURE__ */ jsxRuntime.jsx(ActivationIndicatorIcon, {
        className: `${rootClass}-icon`,
        validationStatus
      })
    })
  });
};

exports.FormActivationIndicator = FormActivationIndicator;
