// @salt-ds/lab/dist-cjs/form-field-legacy/StatusIndicator.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var clsx = require('clsx');
var core = require('@salt-ds/core');

const withBaseName = core.makePrefixer("saltFormField");
const StatusIndicator = (props) => {
  const {
    TooltipComponent = core.Tooltip,
    hasTooltip = false,
    TooltipProps: TooltipProps2,
    tooltipText,
    IconProps: IconProps2,
    className,
    status = "info",
    ...restProps
  } = props;
  return /* @__PURE__ */ jsxRuntime.jsx(TooltipComponent, {
    hideIcon: true,
    status,
    content: tooltipText,
    placement: "top",
    disabled: !hasTooltip,
    ...TooltipProps2,
    children: /* @__PURE__ */ jsxRuntime.jsx(core.StatusIndicator, {
      status,
      ...restProps,
      ...IconProps2,
      className: clsx.clsx(
        withBaseName("statusIndicator"),
        className,
        IconProps2 == null ? void 0 : IconProps2.className
      )
    })
  });
};

exports.StatusIndicator = StatusIndicator;
