// @salt-ds/lab/dist-cjs/toolbar/Toolbar.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var core = require('@salt-ds/core');
var clsx = require('clsx');
var React = require('react');
var OverflowPanel = require('./overflow-panel/OverflowPanel.js');
var OverflowSeparator = require('./overflow-panel/OverflowSeparator.js');
var overflowUtils = require('../responsive/overflowUtils.js');
require('../responsive/useResizeObserver.js');
var useOverflowCollectionItems = require('../responsive/useOverflowCollectionItems.js');
var useOverflowLayout = require('../responsive/useOverflowLayout.js');
var renderToolbarItems = require('./internal/renderToolbarItems.js');
var Tooltray = require('./Tooltray.js');
var window = require('@salt-ds/window');
var styles = require('@salt-ds/styles');
var Toolbar$1 = require('./Toolbar.css.js');

const classBase = "saltToolbar";
const Toolbar = React.forwardRef(
  function Toolbar2(props, ref) {
    const {
      TooltipComponent,
      "aria-label": ariaLabel,
      "aria-labelledby": ariaLabelledBy,
      children,
      className,
      id: idProp,
      overflowButtonIcon,
      overflowButtonLabel,
      overflowButtonPlacement = "end",
      responsive = true,
      disabled = false,
      orientation = "horizontal",
      ...restProp
    } = props;
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-toolbar",
      css: Toolbar$1,
      window: targetWindow
    });
    const toolbarId = core.useIdMemo(idProp);
    const containerRef = React.useRef(null);
    const setContainerRef = core.useForkRef(ref, containerRef);
    const collectionHook = useOverflowCollectionItems.useOverflowCollectionItems({
      children,
      id: toolbarId,
      orientation,
      label: "Toolbar"
    });
    const [innerContainerRef] = useOverflowLayout.useOverflowLayout({
      collectionHook,
      id: toolbarId,
      orientation,
      label: "Toolbar"
    });
    const overflowedItems = collectionHook.data.filter(overflowUtils.isOverflowed);
    const overflowMenuItems = overflowedItems.reduce(
      (items, item, i, arr) => {
        const { element, id } = item;
        if (element.type === Tooltray.Tooltray) {
          const tooltrayProps = element.props;
          const nestedElements = tooltrayProps.children;
          items.push(
            ...React.Children.map(
              nestedElements,
              (el) => React.cloneElement(el, {
                "data-is-inside-panel": true,
                key: id
              })
            )
          );
          if (i < arr.length - 1) {
            items.push(/* @__PURE__ */ jsxRuntime.jsx(OverflowSeparator.OverflowSeparator, {}, `separator-${i}`));
          }
        } else {
          items.push(
            React.cloneElement(element, {
              key: id,
              "data-is-inside-panel": true
            })
          );
        }
        return items;
      },
      []
    );
    const overflowIndicator = collectionHook.data.find(
      (i) => i.isOverflowIndicator
    );
    const overflowPanel = overflowIndicator ? /* @__PURE__ */ jsxRuntime.jsx(OverflowPanel.OverflowPanel, {
      className: clsx.clsx("saltToolbarField"),
      "data-index": collectionHook.data.length,
      "data-overflow-indicator": true,
      "data-priority": 1,
      id: overflowIndicator.id,
      triggerButtonIcon: overflowButtonIcon,
      triggerButtonLabel: overflowButtonLabel,
      children: overflowMenuItems
    }) : null;
    return /* @__PURE__ */ jsxRuntime.jsx("div", {
      "aria-label": ariaLabel,
      "aria-labelledby": clsx.clsx(toolbarId, ariaLabelledBy),
      "aria-orientation": orientation,
      className: clsx.clsx(classBase, className, {
        [`${classBase}-disabled`]: disabled,
        [`${classBase}-nonResponsive`]: !responsive
      }),
      id: toolbarId,
      ref: setContainerRef,
      role: "toolbar",
      ...restProp,
      children: /* @__PURE__ */ jsxRuntime.jsxs("div", {
        className: "Responsive-inner",
        ref: innerContainerRef,
        "data-collapsing": collectionHook.data.some((item) => item.collapsing),
        children: [
          overflowButtonPlacement === "start" && overflowPanel,
          renderToolbarItems.renderToolbarItems(collectionHook, overflowedItems, orientation),
          overflowButtonPlacement === "end" && overflowPanel
        ]
      })
    });
  }
);

exports.Toolbar = Toolbar;
