// @salt-ds/lab/dist-cjs/app-header/AppHeader.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var clsx = require('clsx');
var Toolbar = require('../toolbar/Toolbar.js');
var core = require('@salt-ds/core');
var React = require('react');
require('../toolbar/overflow-panel/OverflowPanel.js');
require('../responsive/useResizeObserver.js');
require('../toolbar/toolbar-field/ToolbarField.js');
var window = require('@salt-ds/window');
var styles = require('@salt-ds/styles');
require('../toolbar/ToolbarButton.js');
var icons = require('@salt-ds/icons');
var AppHeader$1 = require('./AppHeader.css.js');

const withBaseName = core.makePrefixer("saltAppHeader");
const AppHeader = React.forwardRef(
  ({ children, className, ...rest }, ref) => {
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-app-header",
      css: AppHeader$1,
      window: targetWindow
    });
    return /* @__PURE__ */ jsxRuntime.jsx(Toolbar.Toolbar, {
      className: clsx.clsx(withBaseName(), className),
      ref,
      overflowButtonIcon: /* @__PURE__ */ jsxRuntime.jsx(icons.MenuIcon, {}),
      children
    });
  }
);

exports.AppHeader = AppHeader;
