// @salt-ds/core/dist-es/pagination/GoToInput.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { clsx } from 'clsx';
import { forwardRef, useState } from 'react';
import { FormField } from '../form-field/FormField.js';
import '../status-indicator/StatusIndicator.js';
import '../text/Text.js';
import '../text/Code.js';
import '../text/Display.js';
import '../text/Headings.js';
import '../text/Label.js';
import '../text/TextAction.js';
import '../text/TextNotation.js';
import { makePrefixer } from '../utils/makePrefixer.js';
import '../utils/useFloatingUI/useFloatingUI.js';
import '../utils/useId.js';
import '../salt-provider/SaltProvider.js';
import '../viewport/ViewportProvider.js';
import '../accordion/AccordionGroup.js';
import '../accordion/AccordionPanel.js';
import '../accordion/Accordion.js';
import '../accordion/AccordionHeader.js';
import '../aria-announcer/AriaAnnouncerContext.js';
import '../avatar/Avatar.js';
import '../badge/Badge.js';
import '../banner/Banner.js';
import '../banner/BannerActions.js';
import '../banner/BannerContent.js';
import '../border-item/BorderItem.js';
import '../border-layout/BorderLayout.js';
import '../button/Button.js';
import '../card/Card.js';
import '../checkbox/Checkbox.js';
import '../checkbox/CheckboxGroup.js';
import { useComponentCssInjection } from '@salt-ds/styles';
import { useWindow } from '@salt-ds/window';
import '@salt-ds/icons';
import '../combo-box/ComboBox.js';
import '../dialog/Dialog.js';
import '../dialog/DialogHeader.js';
import '../dialog/DialogActions.js';
import '../dialog/DialogContent.js';
import '../dialog/DialogContext.js';
import '../dialog/DialogCloseButton.js';
import '../drawer/Drawer.js';
import '../drawer/DrawerCloseButton.js';
import '../dropdown/Dropdown.js';
import '../file-drop-zone/FileDropZone.js';
import '../file-drop-zone/FileDropZoneIcon.js';
import '../file-drop-zone/FileDropZoneTrigger.js';
import '../flex-item/FlexItem.js';
import '../flex-layout/FlexLayout.js';
import '../flow-layout/FlowLayout.js';
import '../form-field-context/FormFieldContext.js';
import '../grid-item/GridItem.js';
import '../grid-layout/GridLayout.js';
import { Input } from '../input/Input.js';
import '../interactable-card/InteractableCard.js';
import '../interactable-card/InteractableCardGroup.js';
import '../interactable-card/InteractableCardGroupContext.js';
import '../link/Link.js';
import '../link-card/LinkCard.js';
import '../list-control/ListControlContext.js';
import '../multiline-input/MultilineInput.js';
import '../navigation-item/NavigationItem.js';
import '../option/Option.js';
import '../option/OptionGroup.js';
import './Pagination.js';
import './Paginator.js';
import './CompactInput.js';
import './CompactPaginator.js';
import '../panel/Panel.js';
import '../pill/Pill.js';
import '../progress/CircularProgress/CircularProgress.js';
import '../progress/LinearProgress/LinearProgress.js';
import '../radio-button/RadioButton.js';
import '../radio-button/RadioButtonGroup.js';
import '../scrim/Scrim.js';
import '../segmented-button-group/SegmentedButtonGroup.js';
import '../spinner/Spinner.js';
import '../stack-layout/StackLayout.js';
import '../status-adornment/StatusAdornment.js';
import '../toast/Toast.js';
import '../toast/ToastContent.js';
import '../split-layout/SplitLayout.js';
import '../switch/Switch.js';
import '../toggle-button/ToggleButton.js';
import '../toggle-button-group/ToggleButtonGroup.js';
import '../toggle-button-group/ToggleButtonGroupContext.js';
import '../tooltip/Tooltip.js';
import '@floating-ui/react';
import { FormFieldLabel } from '../form-field/FormFieldLabel.js';
import { usePaginationContext } from './usePaginationContext.js';
import css_248z from './GoToInput.css.js';

const withBaseName = makePrefixer("saltGoToInput");
const GoToInput = forwardRef(
  function GoToInput2({
    className,
    inputRef,
    inputVariant = "primary",
    label = "Go to",
    ...restProps
  }, ref) {
    const targetWindow = useWindow();
    useComponentCssInjection({
      testId: "salt-go-to-input",
      css: css_248z,
      window: targetWindow
    });
    const { count, onPageChange } = usePaginationContext();
    const [inputValue, setInputValue] = useState("");
    const onChange = (event) => {
      setInputValue(event.target.value);
    };
    const onKeyDown = (event) => {
      if (event.key === "Enter") {
        const pageValue = Number(inputValue);
        if (!inputValue.startsWith("0") && !isNaN(pageValue) && pageValue > 0 && pageValue <= count) {
          onPageChange(event, pageValue);
        }
        setInputValue("");
      }
    };
    const onBlur = () => {
      setInputValue("");
    };
    const widthCh = `${`${count}`.length}ch`;
    return /* @__PURE__ */ jsxs(FormField, {
      labelPlacement: "left",
      className: clsx(withBaseName(), className),
      ref,
      ...restProps,
      children: [
        /* @__PURE__ */ jsx(FormFieldLabel, {
          children: label
        }),
        /* @__PURE__ */ jsx(Input, {
          className: clsx(withBaseName("input"), {
            [withBaseName("inputDefaultSize")]: count < 100
          }),
          ref: inputRef,
          inputProps: {
            style: { width: widthCh }
          },
          onBlur,
          onChange,
          onKeyDown,
          value: inputValue,
          textAlign: "center",
          variant: inputVariant
        })
      ]
    });
  }
);

export { GoToInput };
