// @salt-ds/lab/dist-cjs/search-input/SearchInput.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var clsx = require('clsx');
var core = require('@salt-ds/core');
var InputLegacy = require('../input-legacy/InputLegacy.js');
var StaticInputAdornment = require('../input-legacy/StaticInputAdornment.js');
var icons = require('@salt-ds/icons');
var window = require('@salt-ds/window');
var styles = require('@salt-ds/styles');
var SearchInput$1 = require('./SearchInput.css.js');

const baseName = "saltSearchInput";
const SearchInput = React.forwardRef(
  function SearchInput2({
    IconComponent = icons.SearchIcon,
    className,
    onChange,
    onClear,
    onKeyUp,
    onSubmit,
    value: valueProp,
    defaultValue,
    ...other
  }, ref) {
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-search-input",
      css: SearchInput$1,
      window: targetWindow
    });
    const inputRef = React.useRef(null);
    const handleRef = core.useForkRef(inputRef, ref);
    const [value, setValue, isControlled] = core.useControlled({
      controlled: valueProp,
      default: defaultValue,
      name: "SearchInput",
      state: "value"
    });
    const handleChange = (event, newValue) => {
      setValue(newValue);
      onChange == null ? void 0 : onChange(event, newValue);
    };
    const handleClear = (event) => {
      var _a;
      setValue("");
      if (!isControlled) {
        onChange == null ? void 0 : onChange(event, "");
      }
      onClear == null ? void 0 : onClear();
      (_a = inputRef.current) == null ? void 0 : _a.focus();
    };
    const handleKeyUp = (event) => {
      if (event.key === "Enter" && onSubmit && value) {
        onSubmit == null ? void 0 : onSubmit(value);
      }
      onKeyUp == null ? void 0 : onKeyUp(event);
    };
    return /* @__PURE__ */ jsxRuntime.jsx(InputLegacy.InputLegacy, {
      className: clsx.clsx(baseName, className),
      ref: handleRef,
      endAdornment: !!value && /* @__PURE__ */ jsxRuntime.jsx(core.Button, {
        className: clsx.clsx(`${baseName}-clearButton`),
        onClick: handleClear,
        variant: "secondary",
        children: /* @__PURE__ */ jsxRuntime.jsx(icons.CloseIcon, {
          "aria-label": "clear input",
          className: `${baseName}-clearIcon`
        })
      }),
      onChange: handleChange,
      onKeyUp: handleKeyUp,
      startAdornment: IconComponent && /* @__PURE__ */ jsxRuntime.jsx(StaticInputAdornment.StaticInputAdornment, {
        children: /* @__PURE__ */ jsxRuntime.jsx(IconComponent, {
          className: `${baseName}-icon`
        })
      }),
      value: value || "",
      ...other
    });
  }
);

exports.SearchInput = SearchInput;
