// @salt-ds/lab/dist-cjs/color-chooser/Color.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tinycolor = require('tinycolor2');
var ColorHelpers = require('./ColorHelpers.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var tinycolor__default = /*#__PURE__*/_interopDefaultLegacy(tinycolor);

class Color {
  constructor() {
    this.color = tinycolor__default["default"]();
  }
  get hex() {
    return this.color.getAlpha() === 1 ? this.color.toHexString() : this.color.toHex8String();
  }
  get rgba() {
    return {
      r: this.color.toRgb().r,
      g: this.color.toRgb().g,
      b: this.color.toRgb().b,
      a: this.color.toRgb().a
    };
  }
  get colorName() {
    return ColorHelpers.getColorNameByHexValue(
      this.color.toHex8String(),
      false,
      void 0,
      true
    );
  }
  static makeColorFromHex(hexValue) {
    const colorObj = new Color();
    colorObj.color = tinycolor__default["default"](hexValue);
    if (colorObj.color.isValid()) {
      return colorObj;
    } else {
      return void 0;
    }
  }
  static makeColorFromRGB(r, g, b, a) {
    const colorObj = new Color();
    colorObj.color = tinycolor__default["default"]({ r, g, b, a });
    return colorObj;
  }
  setAlpha(alpha) {
    const colorObj = new Color();
    colorObj.color = this.color.setAlpha(alpha);
    return colorObj;
  }
}

exports.Color = Color;
