// @salt-ds/lab/dist-cjs/dropdown/useDropdown.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@salt-ds/core');
var React = require('react');
require('../list/Highlighter.js');
require('../list/ListItem.js');
require('../list/List.js');
var useList = require('../list/useList.js');
require('../list/VirtualizedList.js');
require('../common-hooks/collectionProvider.js');
var itemToString = require('../common-hooks/itemToString.js');
require('../common-hooks/keyUtils.js');
require('react/jsx-runtime');
require('../responsive/useResizeObserver.js');

const NULL_REF = { current: null };
const useDropdown = ({
  collectionHook,
  defaultHighlightedIndex: defaultHighlightedIndexProp,
  defaultIsOpen,
  defaultSelected,
  highlightedIndex: highlightedIndexProp,
  isOpen: isOpenProp,
  itemToString: itemToString$1 = itemToString.itemToString,
  onHighlight,
  onOpenChange,
  onSelectionChange,
  onSelect,
  selected,
  selectionStrategy
}) => {
  const isMultiSelect = selectionStrategy === "multiple" || selectionStrategy === "extended";
  const [isOpen, setIsOpen] = core.useControlled({
    controlled: isOpenProp,
    default: defaultIsOpen != null ? defaultIsOpen : false,
    name: "useDropdownList"
  });
  const handleSelectionChange = React.useCallback(
    (evt, selected2) => {
      if (!isMultiSelect) {
        setIsOpen(false);
        onOpenChange == null ? void 0 : onOpenChange(false);
      }
      onSelectionChange == null ? void 0 : onSelectionChange(evt, selected2);
    },
    [isMultiSelect, onOpenChange, onSelectionChange, setIsOpen]
  );
  const handleSelect = React.useCallback(
    (evt, selected2) => {
      if (!isMultiSelect) {
        setIsOpen(false);
        onOpenChange == null ? void 0 : onOpenChange(false);
      }
      onSelect == null ? void 0 : onSelect(evt, selected2);
    },
    [isMultiSelect, onOpenChange, onSelect, setIsOpen]
  );
  const listHook = useList.useList({
    collectionHook,
    defaultHighlightedIndex: (defaultHighlightedIndexProp != null ? defaultHighlightedIndexProp : highlightedIndexProp === void 0) ? 0 : void 0,
    defaultSelected,
    label: "useDropDownList",
    onSelectionChange: handleSelectionChange,
    onSelect: handleSelect,
    containerRef: NULL_REF,
    highlightedIndex: highlightedIndexProp,
    onHighlight,
    selected,
    selectionStrategy,
    tabToSelect: !isMultiSelect
  });
  const handleOpenChange = React.useCallback(
    (open) => {
      setIsOpen(open);
      onOpenChange == null ? void 0 : onOpenChange(open);
    },
    [onOpenChange, setIsOpen]
  );
  const triggerLabel = React.useMemo(() => {
    if (isMultiSelect && Array.isArray(listHook.selected)) {
      const selectedItems = listHook.selected;
      if (selectedItems.length === 0) {
        return void 0;
      } else if (selectedItems.length === 1) {
        const { value } = selectedItems[0];
        return value === null ? void 0 : itemToString$1(value);
      } else {
        return `${selectedItems.length} items selected`;
      }
    } else {
      const selectedItem = listHook.selected;
      return selectedItem == null || selectedItem.value === null ? void 0 : itemToString$1(selectedItem.value);
    }
  }, [isMultiSelect, itemToString$1, listHook.selected]);
  return {
    isOpen,
    onOpenChange: handleOpenChange,
    triggerLabel,
    ...listHook
  };
};

exports.useDropdown = useDropdown;
