// @salt-ds/lab/dist-cjs/stepped-tracker/SteppedTrackerContext.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');

const defaultSteppedTrackerContext = {
  activeStep: 0,
  totalSteps: 1,
  isWithinSteppedTracker: false
};
const SteppedTrackerContext = React.createContext(
  defaultSteppedTrackerContext
);
const SteppedTrackerProvider = ({
  activeStep,
  totalSteps,
  children
}) => {
  const steppedTrackerContextValue = React.useMemo(
    () => ({
      activeStep,
      totalSteps,
      isWithinSteppedTracker: true
    }),
    [activeStep, totalSteps]
  );
  return /* @__PURE__ */ jsxRuntime.jsx(SteppedTrackerContext.Provider, {
    value: steppedTrackerContextValue,
    children
  });
};
const useSteppedTrackerContext = () => React.useContext(SteppedTrackerContext);
const TrackerStepContext = React.createContext(0);
const useTrackerStepContext = () => React.useContext(TrackerStepContext);
const TrackerStepProvider = ({
  children,
  stepNumber
}) => {
  return /* @__PURE__ */ jsxRuntime.jsx(TrackerStepContext.Provider, {
    value: stepNumber,
    children
  });
};

exports.SteppedTrackerProvider = SteppedTrackerProvider;
exports.TrackerStepProvider = TrackerStepProvider;
exports.useSteppedTrackerContext = useSteppedTrackerContext;
exports.useTrackerStepContext = useTrackerStepContext;
