// @salt-ds/lab/dist-cjs/menu-button/MenuButton.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var CascadingMenu = require('../cascading-menu/CascadingMenu.js');
var MenuButtonTrigger = require('./MenuButtonTrigger.js');
var window = require('@salt-ds/window');
var styles = require('@salt-ds/styles');
var MenuButton$1 = require('./MenuButton.css.js');

const MenuButton = React.forwardRef(
  function MenuButton2({ variant = "secondary", CascadingMenuProps: CascadingMenuProps2 = {}, children, ...restProps }, ref) {
    const {
      initialSource = { menuItems: [] },
      onOpen,
      onClose,
      ...restMenuProps
    } = CascadingMenuProps2;
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-menu-button",
      css: MenuButton$1,
      window: targetWindow
    });
    const [isMenuOpen, setMenuOpen] = React.useState(false);
    const handleOpen = React.useCallback(() => {
      setMenuOpen(true);
      onOpen && onOpen();
    }, [onOpen]);
    const handleClose = React.useCallback(() => {
      setMenuOpen(false);
      onClose && onClose();
    }, [onClose]);
    return /* @__PURE__ */ jsxRuntime.jsx(CascadingMenu.CascadingMenu, {
      initialSource,
      onClose: handleClose,
      onOpen: handleOpen,
      ...restMenuProps,
      children: /* @__PURE__ */ jsxRuntime.jsx(MenuButtonTrigger.MenuButtonTrigger, {
        "aria-expanded": isMenuOpen,
        "aria-haspopup": true,
        isMenuOpen,
        ref,
        variant,
        ...restProps,
        children
      })
    });
  }
);

exports.MenuButton = MenuButton;
