// @salt-ds/lab/dist-es/tabs/TabActivationIndicator.js
import { jsx } from 'react/jsx-runtime';
import { makePrefixer } from '@salt-ds/core';
import { useRef } from 'react';
import { useActivationIndicator } from './useActivationIndicator.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './TabActivationIndicator.css.js';

const withBaseName = makePrefixer("saltTabActivationIndicator");
const TabActivationIndicator = ({
  hideThumb = false,
  orientation = "horizontal",
  tabId
}) => {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-tab-activation-indicator",
    css: css_248z,
    window: targetWindow
  });
  const rootRef = useRef(null);
  const style = useActivationIndicator({
    rootRef,
    tabId,
    orientation
  });
  return /* @__PURE__ */ jsx("div", {
    className: withBaseName(),
    ref: rootRef,
    children: hideThumb === false && tabId !== null ? /* @__PURE__ */ jsx("div", {
      className: withBaseName("thumb"),
      style
    }) : null
  });
};

export { TabActivationIndicator };
