// @salt-ds/lab/dist-es/tabs/Tabs.js
import { jsxs, jsx } from 'react/jsx-runtime';
import { forwardRef, useRef, useImperativeHandle, useCallback } from 'react';
import { clsx } from 'clsx';
import { makePrefixer } from '@salt-ds/core';
import { useTabs } from './useTabs.js';
import { Tabstrip } from './Tabstrip.js';
import { useWindow } from '@salt-ds/window';
import { useComponentCssInjection } from '@salt-ds/styles';
import css_248z from './Tabs.css.js';

const withBaseName = makePrefixer("saltTabs");
const Tabs = forwardRef(function Tabs2({
  activeTabIndex: activeTabIndexProp,
  allowDragDrop,
  centered,
  children,
  className,
  defaultActiveTabIndex,
  editing,
  enableAddTab,
  enableCloseTab,
  enableRenameTab,
  onActiveChange: onActiveChangeProp,
  onAddTab,
  onCloseTab,
  onEnterEditMode,
  onExitEditMode,
  onMoveTab,
  overflowMenu,
  variant,
  ...htmlAttributes
}, forwardedRef) {
  const targetWindow = useWindow();
  useComponentCssInjection({
    testId: "salt-tabs",
    css: css_248z,
    window: targetWindow
  });
  const tabstripRef = useRef(null);
  useImperativeHandle(
    forwardedRef,
    () => ({
      focus: () => {
        var _a;
        return (_a = tabstripRef.current) == null ? void 0 : _a.focus();
      }
    }),
    []
  );
  const { id, onActiveChange, activeTabIndex, tabPanel, tabs } = useTabs({
    activeTabIndex: activeTabIndexProp,
    children,
    defaultActiveTabIndex,
    onActiveChange: onActiveChangeProp
  });
  const tabstripProps = {
    activeTabIndex,
    allowDragDrop,
    centered,
    editing,
    variant,
    enableAddTab,
    enableCloseTab,
    enableRenameTab,
    onAddTab,
    onCloseTab,
    onEnterEditMode,
    onExitEditMode,
    onMoveTab,
    overflowMenu
  };
  useCallback(
    (tabIndex) => {
      if (tabIndex === activeTabIndex) {
        return `${id}-${tabIndex}-panel`;
      }
    },
    [id, activeTabIndex]
  );
  return /* @__PURE__ */ jsxs("div", {
    ...htmlAttributes,
    className: clsx(withBaseName(), className),
    id: `${id}-tabs`,
    children: [
      /* @__PURE__ */ jsx(Tabstrip, {
        ...tabstripProps,
        id,
        onActiveChange,
        ref: tabstripRef,
        source: tabs
      }),
      tabPanel
    ]
  });
});

export { Tabs };
