// @salt-ds/lab/dist-es/tabs/useTabs.js
import { useIdMemo, useControlled } from '@salt-ds/core';
import { useCallback } from 'react';
import { useItemsWithIds } from './useItemsWithIds.js';

const useTabs = ({
  activeTabIndex: activeTabIndexProp,
  children,
  defaultActiveTabIndex,
  id: idProp,
  onActiveChange
}) => {
  var _a, _b;
  const id = useIdMemo(idProp);
  const [itemsWithIds] = useItemsWithIds(children, id);
  const [activeTabIndex, setActiveTabIndex] = useControlled({
    controlled: activeTabIndexProp,
    default: defaultActiveTabIndex != null ? defaultActiveTabIndex : 0,
    name: "useTabs",
    state: "activeTabIndex"
  });
  const handleTabActivated = useCallback(
    (tabIndex) => {
      setActiveTabIndex(tabIndex);
      onActiveChange == null ? void 0 : onActiveChange(tabIndex);
    },
    [onActiveChange, setActiveTabIndex]
  );
  return {
    id,
    onActiveChange: handleTabActivated,
    activeTabIndex,
    tabs: itemsWithIds,
    tabPanel: activeTabIndex === null ? null : (_b = (_a = itemsWithIds[activeTabIndex]) == null ? void 0 : _a.element) != null ? _b : null
  };
};

export { useTabs };
