// @salt-ds/lab/dist-cjs/stepped-tracker/TrackerStep/TrackerStep.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var clsx = require('clsx');
var styles = require('@salt-ds/styles');
var window = require('@salt-ds/window');
var core = require('@salt-ds/core');
var icons = require('@salt-ds/icons');
var TrackerConnector = require('../TrackerConnector/TrackerConnector.js');
var SteppedTrackerContext = require('../SteppedTrackerContext.js');
var TrackerStep$1 = require('./TrackerStep.css.js');

const withBaseName = core.makePrefixer("saltTrackerStep");
const iconMap = {
  default: icons.StepDefaultIcon,
  completed: icons.StepSuccessIcon
};
const getStateIcon = ({
  isActive,
  state
}) => {
  if (state === "default" && isActive) {
    return icons.StepActiveIcon;
  }
  return iconMap[state];
};
const getState = ({
  isActive,
  state
}) => {
  if (state === "default" && isActive) {
    return "active";
  }
  return state;
};
const useCheckWithinSteppedTracker = (isWithinSteppedTracker) => {
  React.useEffect(() => {
    if (process.env.NODE_ENV !== "production") {
      if (!isWithinSteppedTracker) {
        console.error(
          "The TrackerStep component must be placed within a SteppedTracker component"
        );
      }
    }
  }, [isWithinSteppedTracker]);
};
const TrackerStep = React.forwardRef(
  function TrackerStep2(props, ref) {
    const {
      state = "default",
      style,
      className,
      children,
      ...restProps
    } = props;
    const targetWindow = window.useWindow();
    styles.useComponentCssInjection({
      testId: "salt-tracker-step",
      css: TrackerStep$1,
      window: targetWindow
    });
    const { activeStep, totalSteps, isWithinSteppedTracker } = SteppedTrackerContext.useSteppedTrackerContext();
    const stepNumber = SteppedTrackerContext.useTrackerStepContext();
    useCheckWithinSteppedTracker(isWithinSteppedTracker);
    const isActive = activeStep === stepNumber;
    const Icon = getStateIcon({ isActive, state });
    const resolvedState = getState({ isActive, state });
    const connectorState = activeStep > stepNumber ? "active" : "default";
    const hasConnector = stepNumber < totalSteps - 1;
    const innerStyle = {
      ...style,
      "--saltTrackerStep-width": `${100 / totalSteps}%`
    };
    return /* @__PURE__ */ jsxRuntime.jsxs("li", {
      className: clsx.clsx(withBaseName(), withBaseName(resolvedState), className),
      style: innerStyle,
      "aria-current": isActive ? "step" : void 0,
      "data-state": state,
      ref,
      ...restProps,
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(Icon, {}),
        hasConnector && /* @__PURE__ */ jsxRuntime.jsx(TrackerConnector.TrackerConnector, {
          state: connectorState
        }),
        /* @__PURE__ */ jsxRuntime.jsx("div", {
          className: withBaseName("body"),
          children
        })
      ]
    });
  }
);

exports.TrackerStep = TrackerStep;
